
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../graph_relationships_loader/">
      
      
        <link rel="next" href="../about/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.30">
    
    
      
        <title>Documentation - ChemGraphBuilder</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#documentation-of-the-main-classes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="ChemGraphBuilder" class="md-header__button md-logo" aria-label="ChemGraphBuilder" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            ChemGraphBuilder
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Documentation
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-hidden="true"  type="radio" name="__palette" id="__palette_0">
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-hidden="true"  type="radio" name="__palette" id="__palette_1">
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../1.node_collector_processor/" class="md-tabs__link">
          
  
    
  
  Usage Examples

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
    
  
  Documentation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../about/" class="md-tabs__link">
        
  
    
  
  About

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="ChemGraphBuilder" class="md-nav__button md-logo" aria-label="ChemGraphBuilder" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    ChemGraphBuilder
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Usage Examples
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Usage Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.node_collector_processor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Step 1: Nodes Collector Processor
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../node_collector_processor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Class Documentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.relationship_collector_processor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Step 2: Relationship Collector Processor
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../relationship_collector_processor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Class Documentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.graph_nodes_loader/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Step 3: Graph Nodes Loader
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../graph_nodes_loader/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Class Documentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.graph_relationships_loader/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Step 4: Graph Relationships Loader
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../graph_relationships_loader/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Class Documentation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Documentation
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Documentation
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-setup-data-folder" class="md-nav__link">
    <span class="md-ellipsis">
      1. Setup Data Folder
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder" class="md-nav__link">
    <span class="md-ellipsis">
      setup_data_folder
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder" class="md-nav__link">
    <span class="md-ellipsis">
      SetupDataFolder
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SetupDataFolder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.create_folder" class="md-nav__link">
    <span class="md-ellipsis">
      create_folder
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.setup" class="md-nav__link">
    <span class="md-ellipsis">
      setup
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-neo4j-driver" class="md-nav__link">
    <span class="md-ellipsis">
      2. Neo4j Driver
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver" class="md-nav__link">
    <span class="md-ellipsis">
      neo4jdriver
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver.Neo4jBase" class="md-nav__link">
    <span class="md-ellipsis">
      Neo4jBase
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Neo4jBase">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver.Neo4jBase.close" class="md-nav__link">
    <span class="md-ellipsis">
      close
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver.Neo4jBase.connect_to_neo4j" class="md-nav__link">
    <span class="md-ellipsis">
      connect_to_neo4j
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver.Neo4jConnectionError" class="md-nav__link">
    <span class="md-ellipsis">
      Neo4jConnectionError
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-node-properties-extractor" class="md-nav__link">
    <span class="md-ellipsis">
      3. Node Properties Extractor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor" class="md-nav__link">
    <span class="md-ellipsis">
      node_properties_extractor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor" class="md-nav__link">
    <span class="md-ellipsis">
      NodePropertiesExtractor
    </span>
  </a>
  
    <nav class="md-nav" aria-label="NodePropertiesExtractor">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_assay_properties" class="md-nav__link">
    <span class="md-ellipsis">
      extract_assay_properties
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_compound_properties" class="md-nav__link">
    <span class="md-ellipsis">
      extract_compound_properties
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_gene_properties" class="md-nav__link">
    <span class="md-ellipsis">
      extract_gene_properties
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_protein_properties" class="md-nav__link">
    <span class="md-ellipsis">
      extract_protein_properties
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      fetch_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.get_enzyme_assays" class="md-nav__link">
    <span class="md-ellipsis">
      get_enzyme_assays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.run" class="md-nav__link">
    <span class="md-ellipsis">
      run
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-node-data-processor" class="md-nav__link">
    <span class="md-ellipsis">
      4. Node Data Processor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor" class="md-nav__link">
    <span class="md-ellipsis">
      node_data_processor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor" class="md-nav__link">
    <span class="md-ellipsis">
      NodeDataProcessor
    </span>
  </a>
  
    <nav class="md-nav" aria-label="NodeDataProcessor">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_assays" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess_assays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_compounds" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess_compounds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_genes" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess_genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_proteins" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess_proteins
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-add-graph-nodes" class="md-nav__link">
    <span class="md-ellipsis">
      5. Add Graph Nodes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes" class="md-nav__link">
    <span class="md-ellipsis">
      add_graph_nodes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes" class="md-nav__link">
    <span class="md-ellipsis">
      AddGraphNodes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="AddGraphNodes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes--methods" class="md-nav__link">
    <span class="md-ellipsis">
      Methods:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
    <nav class="md-nav" aria-label="__init__">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.__init__--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files" class="md-nav__link">
    <span class="md-ellipsis">
      combine_csv_files
    </span>
  </a>
  
    <nav class="md-nav" aria-label="combine_csv_files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files--returns" class="md-nav__link">
    <span class="md-ellipsis">
      Returns:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.create_uniqueness_constraint" class="md-nav__link">
    <span class="md-ellipsis">
      create_uniqueness_constraint
    </span>
  </a>
  
    <nav class="md-nav" aria-label="create_uniqueness_constraint">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.create_uniqueness_constraint--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.execute_queries" class="md-nav__link">
    <span class="md-ellipsis">
      execute_queries
    </span>
  </a>
  
    <nav class="md-nav" aria-label="execute_queries">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.execute_queries--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries" class="md-nav__link">
    <span class="md-ellipsis">
      generate_cypher_queries
    </span>
  </a>
  
    <nav class="md-nav" aria-label="generate_cypher_queries">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries--yields" class="md-nav__link">
    <span class="md-ellipsis">
      Yields:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes" class="md-nav__link">
    <span class="md-ellipsis">
      process_and_add_nodes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="process_and_add_nodes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes_from_directory" class="md-nav__link">
    <span class="md-ellipsis">
      process_and_add_nodes_from_directory
    </span>
  </a>
  
    <nav class="md-nav" aria-label="process_and_add_nodes_from_directory">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes_from_directory--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string" class="md-nav__link">
    <span class="md-ellipsis">
      public_generate_property_string
    </span>
  </a>
  
    <nav class="md-nav" aria-label="public_generate_property_string">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string--returns" class="md-nav__link">
    <span class="md-ellipsis">
      Returns:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file" class="md-nav__link">
    <span class="md-ellipsis">
      read_csv_file
    </span>
  </a>
  
    <nav class="md-nav" aria-label="read_csv_file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file--returns" class="md-nav__link">
    <span class="md-ellipsis">
      Returns:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-relationship-properties-extractor" class="md-nav__link">
    <span class="md-ellipsis">
      6. Relationship Properties Extractor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor" class="md-nav__link">
    <span class="md-ellipsis">
      relationship_properties_extractor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor" class="md-nav__link">
    <span class="md-ellipsis">
      RelationshipPropertiesExtractor
    </span>
  </a>
  
    <nav class="md-nav" aria-label="RelationshipPropertiesExtractor">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.assay_compound_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      assay_compound_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.assay_enzyme_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      assay_enzyme_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_cooccurrence" class="md-nav__link">
    <span class="md-ellipsis">
      compound_cooccurrence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_gene_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      compound_gene_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_similarity_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      compound_similarity_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_transformation" class="md-nav__link">
    <span class="md-ellipsis">
      compound_transformation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.fetch_data_for_aid" class="md-nav__link">
    <span class="md-ellipsis">
      fetch_data_for_aid
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.fetch_similar_cids" class="md-nav__link">
    <span class="md-ellipsis">
      fetch_similar_cids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.gene_enzyme_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      gene_enzyme_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.process_chunk" class="md-nav__link">
    <span class="md-ellipsis">
      process_chunk
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-relationship-data-processor" class="md-nav__link">
    <span class="md-ellipsis">
      7. Relationship Data Processor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor" class="md-nav__link">
    <span class="md-ellipsis">
      relationship_data_processor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor" class="md-nav__link">
    <span class="md-ellipsis">
      RelationshipDataProcessor
    </span>
  </a>
  
    <nav class="md-nav" aria-label="RelationshipDataProcessor">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.most_frequent" class="md-nav__link">
    <span class="md-ellipsis">
      most_frequent
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.process_files" class="md-nav__link">
    <span class="md-ellipsis">
      process_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.propagate_phenotype" class="md-nav__link">
    <span class="md-ellipsis">
      propagate_phenotype
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-add-graph-relationships" class="md-nav__link">
    <span class="md-ellipsis">
      8. Add Graph Relationships
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder" class="md-nav__link">
    <span class="md-ellipsis">
      setup_data_folder
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder" class="md-nav__link">
    <span class="md-ellipsis">
      SetupDataFolder
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SetupDataFolder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.create_folder" class="md-nav__link">
    <span class="md-ellipsis">
      create_folder
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.setup" class="md-nav__link">
    <span class="md-ellipsis">
      setup
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    About
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-setup-data-folder" class="md-nav__link">
    <span class="md-ellipsis">
      1. Setup Data Folder
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder" class="md-nav__link">
    <span class="md-ellipsis">
      setup_data_folder
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder" class="md-nav__link">
    <span class="md-ellipsis">
      SetupDataFolder
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SetupDataFolder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.create_folder" class="md-nav__link">
    <span class="md-ellipsis">
      create_folder
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.setup" class="md-nav__link">
    <span class="md-ellipsis">
      setup
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-neo4j-driver" class="md-nav__link">
    <span class="md-ellipsis">
      2. Neo4j Driver
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver" class="md-nav__link">
    <span class="md-ellipsis">
      neo4jdriver
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver.Neo4jBase" class="md-nav__link">
    <span class="md-ellipsis">
      Neo4jBase
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Neo4jBase">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver.Neo4jBase.close" class="md-nav__link">
    <span class="md-ellipsis">
      close
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver.Neo4jBase.connect_to_neo4j" class="md-nav__link">
    <span class="md-ellipsis">
      connect_to_neo4j
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.neo4jdriver.Neo4jConnectionError" class="md-nav__link">
    <span class="md-ellipsis">
      Neo4jConnectionError
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-node-properties-extractor" class="md-nav__link">
    <span class="md-ellipsis">
      3. Node Properties Extractor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor" class="md-nav__link">
    <span class="md-ellipsis">
      node_properties_extractor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor" class="md-nav__link">
    <span class="md-ellipsis">
      NodePropertiesExtractor
    </span>
  </a>
  
    <nav class="md-nav" aria-label="NodePropertiesExtractor">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_assay_properties" class="md-nav__link">
    <span class="md-ellipsis">
      extract_assay_properties
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_compound_properties" class="md-nav__link">
    <span class="md-ellipsis">
      extract_compound_properties
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_gene_properties" class="md-nav__link">
    <span class="md-ellipsis">
      extract_gene_properties
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_protein_properties" class="md-nav__link">
    <span class="md-ellipsis">
      extract_protein_properties
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      fetch_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.get_enzyme_assays" class="md-nav__link">
    <span class="md-ellipsis">
      get_enzyme_assays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.run" class="md-nav__link">
    <span class="md-ellipsis">
      run
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-node-data-processor" class="md-nav__link">
    <span class="md-ellipsis">
      4. Node Data Processor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor" class="md-nav__link">
    <span class="md-ellipsis">
      node_data_processor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor" class="md-nav__link">
    <span class="md-ellipsis">
      NodeDataProcessor
    </span>
  </a>
  
    <nav class="md-nav" aria-label="NodeDataProcessor">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_assays" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess_assays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_compounds" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess_compounds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_genes" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess_genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_proteins" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess_proteins
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-add-graph-nodes" class="md-nav__link">
    <span class="md-ellipsis">
      5. Add Graph Nodes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes" class="md-nav__link">
    <span class="md-ellipsis">
      add_graph_nodes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes" class="md-nav__link">
    <span class="md-ellipsis">
      AddGraphNodes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="AddGraphNodes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes--methods" class="md-nav__link">
    <span class="md-ellipsis">
      Methods:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
    <nav class="md-nav" aria-label="__init__">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.__init__--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files" class="md-nav__link">
    <span class="md-ellipsis">
      combine_csv_files
    </span>
  </a>
  
    <nav class="md-nav" aria-label="combine_csv_files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files--returns" class="md-nav__link">
    <span class="md-ellipsis">
      Returns:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.create_uniqueness_constraint" class="md-nav__link">
    <span class="md-ellipsis">
      create_uniqueness_constraint
    </span>
  </a>
  
    <nav class="md-nav" aria-label="create_uniqueness_constraint">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.create_uniqueness_constraint--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.execute_queries" class="md-nav__link">
    <span class="md-ellipsis">
      execute_queries
    </span>
  </a>
  
    <nav class="md-nav" aria-label="execute_queries">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.execute_queries--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries" class="md-nav__link">
    <span class="md-ellipsis">
      generate_cypher_queries
    </span>
  </a>
  
    <nav class="md-nav" aria-label="generate_cypher_queries">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries--yields" class="md-nav__link">
    <span class="md-ellipsis">
      Yields:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes" class="md-nav__link">
    <span class="md-ellipsis">
      process_and_add_nodes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="process_and_add_nodes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes_from_directory" class="md-nav__link">
    <span class="md-ellipsis">
      process_and_add_nodes_from_directory
    </span>
  </a>
  
    <nav class="md-nav" aria-label="process_and_add_nodes_from_directory">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes_from_directory--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string" class="md-nav__link">
    <span class="md-ellipsis">
      public_generate_property_string
    </span>
  </a>
  
    <nav class="md-nav" aria-label="public_generate_property_string">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string--returns" class="md-nav__link">
    <span class="md-ellipsis">
      Returns:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file" class="md-nav__link">
    <span class="md-ellipsis">
      read_csv_file
    </span>
  </a>
  
    <nav class="md-nav" aria-label="read_csv_file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file--returns" class="md-nav__link">
    <span class="md-ellipsis">
      Returns:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-relationship-properties-extractor" class="md-nav__link">
    <span class="md-ellipsis">
      6. Relationship Properties Extractor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor" class="md-nav__link">
    <span class="md-ellipsis">
      relationship_properties_extractor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor" class="md-nav__link">
    <span class="md-ellipsis">
      RelationshipPropertiesExtractor
    </span>
  </a>
  
    <nav class="md-nav" aria-label="RelationshipPropertiesExtractor">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.assay_compound_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      assay_compound_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.assay_enzyme_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      assay_enzyme_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_cooccurrence" class="md-nav__link">
    <span class="md-ellipsis">
      compound_cooccurrence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_gene_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      compound_gene_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_similarity_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      compound_similarity_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_transformation" class="md-nav__link">
    <span class="md-ellipsis">
      compound_transformation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.fetch_data_for_aid" class="md-nav__link">
    <span class="md-ellipsis">
      fetch_data_for_aid
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.fetch_similar_cids" class="md-nav__link">
    <span class="md-ellipsis">
      fetch_similar_cids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.gene_enzyme_relationship" class="md-nav__link">
    <span class="md-ellipsis">
      gene_enzyme_relationship
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.process_chunk" class="md-nav__link">
    <span class="md-ellipsis">
      process_chunk
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-relationship-data-processor" class="md-nav__link">
    <span class="md-ellipsis">
      7. Relationship Data Processor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor" class="md-nav__link">
    <span class="md-ellipsis">
      relationship_data_processor
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor" class="md-nav__link">
    <span class="md-ellipsis">
      RelationshipDataProcessor
    </span>
  </a>
  
    <nav class="md-nav" aria-label="RelationshipDataProcessor">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.most_frequent" class="md-nav__link">
    <span class="md-ellipsis">
      most_frequent
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.process_files" class="md-nav__link">
    <span class="md-ellipsis">
      process_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.propagate_phenotype" class="md-nav__link">
    <span class="md-ellipsis">
      propagate_phenotype
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-add-graph-relationships" class="md-nav__link">
    <span class="md-ellipsis">
      8. Add Graph Relationships
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder" class="md-nav__link">
    <span class="md-ellipsis">
      setup_data_folder
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder" class="md-nav__link">
    <span class="md-ellipsis">
      SetupDataFolder
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SetupDataFolder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.create_folder" class="md-nav__link">
    <span class="md-ellipsis">
      create_folder
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.SetupDataFolder.setup" class="md-nav__link">
    <span class="md-ellipsis">
      setup
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemgraphbuilder.setup_data_folder.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="documentation-of-the-main-classes">Documentation of the main classes</h1>
<h2 id="1-setup-data-folder">1. Setup Data Folder</h2>


<div class="doc doc-object doc-module">



<a id="chemgraphbuilder.setup_data_folder"></a>
    <div class="doc doc-contents first">

      <p>Module to set up a data directory with a predefined structure.</p>
<p>This module provides the DataFolderSetup class, which creates a directory
structure for a data folder. The structure includes nodes and
relationships folders with specified subfolders.</p>


<p><span class="doc-section-title">Classes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.setup_data_folder.DataFolderSetup">DataFolderSetup</span></code></td>
            <td>
              <div class="doc-md-description">
                <p>Class to set up a data directory with a predefined structure.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Functions:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="chemgraphbuilder.setup_data_folder.main" href="#chemgraphbuilder.setup_data_folder.main">main</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Main function to set up the data directory.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.setup_data_folder.SetupDataFolder" class="doc doc-heading">
            <code>SetupDataFolder</code>


</h2>


    <div class="doc doc-contents ">


      <p>Class to set up a data directory with a predefined structure.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.setup_data_folder.SetupDataFolder.data_folder">data_folder</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the data folder.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.setup_data_folder.SetupDataFolder.base_path">base_path</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The base path for the data directory.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.setup_data_folder.SetupDataFolder.structure">structure</span></code></td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The structure of directories to create.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">SetupDataFolder</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class to set up a data directory with a predefined structure.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data_folder (str): The name of the data folder.</span>
<span class="sd">        base_path (str): The base path for the data directory.</span>
<span class="sd">        structure (dict): The structure of directories to create.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes the DataFolderSetup with the data folder name and directory structure.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_folder</span> <span class="o">=</span> <span class="s2">&quot;Data&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">base_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_folder</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Nodes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Compound_Properties&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Relationships&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;Assay_Compound_Relationship&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Compound_Similarities&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Cpd_Cpd_CoOccurrence&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Cpd_Gene_CoOccurrence&quot;</span>
            <span class="p">]</span>
        <span class="p">}</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">create_folder</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a folder if it does not already exist.</span>

<span class="sd">        Args:</span>
<span class="sd">            path (str): The path of the folder to create.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Created folder: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Folder already exists: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">setup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets up the data directory structure based on the predefined structure.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Create the base data directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">)</span>

        <span class="c1"># Create the &#39;Nodes&#39; directory and its subdirectories</span>
        <span class="n">nodes_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">,</span> <span class="s2">&quot;Nodes&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">nodes_path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="s2">&quot;Nodes&quot;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nodes_path</span><span class="p">,</span> <span class="n">folder</span><span class="p">))</span>

        <span class="c1"># Create the &#39;Relationships&#39; directory and its subdirectories</span>
        <span class="n">relationships_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">,</span> <span class="s2">&quot;Relationships&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">relationships_path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="s2">&quot;Relationships&quot;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">relationships_path</span><span class="p">,</span> <span class="n">folder</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.setup_data_folder.SetupDataFolder.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Initializes the DataFolderSetup with the data folder name and directory structure.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the DataFolderSetup with the data folder name and directory structure.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data_folder</span> <span class="o">=</span> <span class="s2">&quot;Data&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">base_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_folder</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;Nodes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Compound_Properties&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Relationships&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;Assay_Compound_Relationship&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Compound_Similarities&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Cpd_Cpd_CoOccurrence&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Cpd_Gene_CoOccurrence&quot;</span>
        <span class="p">]</span>
    <span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.setup_data_folder.SetupDataFolder.create_folder" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_folder</span><span class="p">(</span><span class="n">path</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

      <p>Creates a folder if it does not already exist.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>path</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path of the folder to create.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">create_folder</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a folder if it does not already exist.</span>

<span class="sd">    Args:</span>
<span class="sd">        path (str): The path of the folder to create.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Created folder: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Folder already exists: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.setup_data_folder.SetupDataFolder.setup" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">setup</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Sets up the data directory structure based on the predefined structure.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">setup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sets up the data directory structure based on the predefined structure.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create the base data directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">)</span>

    <span class="c1"># Create the &#39;Nodes&#39; directory and its subdirectories</span>
    <span class="n">nodes_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">,</span> <span class="s2">&quot;Nodes&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">nodes_path</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="s2">&quot;Nodes&quot;</span><span class="p">]:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nodes_path</span><span class="p">,</span> <span class="n">folder</span><span class="p">))</span>

    <span class="c1"># Create the &#39;Relationships&#39; directory and its subdirectories</span>
    <span class="n">relationships_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">,</span> <span class="s2">&quot;Relationships&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">relationships_path</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="s2">&quot;Relationships&quot;</span><span class="p">]:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">relationships_path</span><span class="p">,</span> <span class="n">folder</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h2 id="chemgraphbuilder.setup_data_folder.main" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Main function to set up the data directory.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function to set up the data directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_folder_setup</span> <span class="o">=</span> <span class="n">SetupDataFolder</span><span class="p">()</span>
    <span class="n">data_folder_setup</span><span class="o">.</span><span class="n">setup</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="2-neo4j-driver">2. Neo4j Driver</h2>


<div class="doc doc-object doc-module">



<a id="chemgraphbuilder.neo4jdriver"></a>
    <div class="doc doc-contents first">

      <p>Module for managing connections to a Neo4j database.</p>
<p>This module provides classes and methods to establish and manage connections
with a Neo4j database, including custom error handling.</p>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.neo4jdriver.Neo4jBase" class="doc doc-heading">
            <code>Neo4jBase</code>


</h2>


    <div class="doc doc-contents ">


      <p>Base class to manage connections with the Neo4j database.</p>
<p>Attributes:
- uri: The connection URI for the Neo4j database.
- user: The username to use for authentication.
- driver: The driver object used to interact with the Neo4j database.</p>
<p>Methods:
- connect_to_neo4j: Establish a connection to the Neo4j database.
- close: Close the connection to the Neo4j database.</p>

              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/neo4jdriver.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Neo4jBase</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Base class to manage connections with the Neo4j database.</span>

<span class="sd">    Attributes:</span>
<span class="sd">    - uri: The connection URI for the Neo4j database.</span>
<span class="sd">    - user: The username to use for authentication.</span>
<span class="sd">    - driver: The driver object used to interact with the Neo4j database.</span>

<span class="sd">    Methods:</span>
<span class="sd">    - connect_to_neo4j: Establish a connection to the Neo4j database.</span>
<span class="sd">    - close: Close the connection to the Neo4j database.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">uri</span><span class="o">=</span><span class="s2">&quot;tcp://5.tcp.eu.ngrok.io:12445&quot;</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="s2">&quot;neo4j&quot;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">uri</span> <span class="o">=</span> <span class="n">uri</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">user</span> <span class="o">=</span> <span class="n">user</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">driver</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Set up logging configuration</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> - </span><span class="si">%(levelname)s</span><span class="s1"> - </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span> <span class="o">=</span> <span class="n">logger</span> <span class="ow">or</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">connect_to_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Establish a connection to the Neo4j database using provided URI and username.&quot;&quot;&quot;</span>
        <span class="n">password</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s2">&quot;NEO4J_PASSWORD&quot;</span><span class="p">)</span>  <span class="c1"># Check if password is set in environment variables</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">password</span><span class="p">:</span>
            <span class="n">password</span> <span class="o">=</span> <span class="n">getpass</span><span class="o">.</span><span class="n">getpass</span><span class="p">(</span><span class="n">prompt</span><span class="o">=</span><span class="s2">&quot;Enter Neo4j password: &quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">driver</span> <span class="o">=</span> <span class="n">GraphDatabase</span><span class="o">.</span><span class="n">driver</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uri</span><span class="p">,</span> <span class="n">auth</span><span class="o">=</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully connected to the Neo4j database.&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed to connect to the Neo4j database: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">Neo4jConnectionError</span><span class="p">(</span><span class="s2">&quot;Failed to connect to the Neo4j database.&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>

    <span class="k">def</span> <span class="nf">close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Close the connection to the Neo4j database.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">driver</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">driver</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Neo4j connection closed successfully.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.neo4jdriver.Neo4jBase.close" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">close</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Close the connection to the Neo4j database.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/neo4jdriver.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Close the connection to the Neo4j database.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">driver</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">driver</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Neo4j connection closed successfully.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.neo4jdriver.Neo4jBase.connect_to_neo4j" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">connect_to_neo4j</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Establish a connection to the Neo4j database using provided URI and username.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/neo4jdriver.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">connect_to_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Establish a connection to the Neo4j database using provided URI and username.&quot;&quot;&quot;</span>
    <span class="n">password</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s2">&quot;NEO4J_PASSWORD&quot;</span><span class="p">)</span>  <span class="c1"># Check if password is set in environment variables</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">password</span><span class="p">:</span>
        <span class="n">password</span> <span class="o">=</span> <span class="n">getpass</span><span class="o">.</span><span class="n">getpass</span><span class="p">(</span><span class="n">prompt</span><span class="o">=</span><span class="s2">&quot;Enter Neo4j password: &quot;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">driver</span> <span class="o">=</span> <span class="n">GraphDatabase</span><span class="o">.</span><span class="n">driver</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uri</span><span class="p">,</span> <span class="n">auth</span><span class="o">=</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully connected to the Neo4j database.&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed to connect to the Neo4j database: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">Neo4jConnectionError</span><span class="p">(</span><span class="s2">&quot;Failed to connect to the Neo4j database.&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.neo4jdriver.Neo4jConnectionError" class="doc doc-heading">
            <code>Neo4jConnectionError</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>Exception</code></p>


      <p>Custom exception for Neo4j connection errors.</p>

              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/neo4jdriver.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">14</span>
<span class="normal">15</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Neo4jConnectionError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Custom exception for Neo4j connection errors.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
              </details>

    </div>

</div>




  </div>

    </div>

</div><h2 id="3-node-properties-extractor">3. Node Properties Extractor</h2>


<div class="doc doc-object doc-module">



<a id="chemgraphbuilder.node_properties_extractor"></a>
    <div class="doc doc-contents first">

      <p>This module defines the <code>NodePropertiesExtractor</code> class, responsible for
extracting data from the PubChem database to build knowledge graphs in Neo4j.
The class focuses on nodes representing chemical entities and their relationships,
allowing users to query chemical data and construct a graph-based representation
of chemical compounds, their assays, related genes, and proteins.</p>
<p>The primary functionality revolves around fetching detailed information about
specified enzymes from PubChem, including assay data, gene properties, protein
properties, and compound properties. It processes this data into a structured
format suitable for knowledge graph construction, specifically tailored for use
with Neo4j databases.</p>


<p><span class="doc-section-title">Classes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.- NodePropertiesExtractor">- NodePropertiesExtractor</span></code></td>
            <td>
              <div class="doc-md-description">
                <p>A class to extract data from PubChem to build
knowledge graphs in Neo4j.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="usage-example" open>
  <summary>Usage Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>enzyme_list = ['CYP2D6', 'CYP3A4']
extractor = NodePropertiesExtractor(enzyme_list)
df = extractor.run()
This example initiates the extractor with a list of enzymes, fetches their
data from PubChem, processes it, and potentially prepares it for knowledge
graph construction in Neo4j.</p>
</blockquote>
</blockquote>
</blockquote>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>To fully utilize this class, ensure you have network access to the PubChem
API for data retrieval and a Neo4j database instance for knowledge graph
construction. The class methods facilitate data extraction and processing,
but integrating the output into Neo4j requires additional steps outside the
scope of this class.</p>
</details>


  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor" class="doc doc-heading">
            <code>NodePropertiesExtractor</code>


</h2>


    <div class="doc doc-contents ">


      <p>Extracts data from PubChem to build knowledge graphs in Neo4j,
focusing on nodes representing chemical entities and their relationships.
This class serves as a bridge between the PubChem database and Neo4j,
allowing users to query chemical data and construct a graph-based
representation of chemical compounds, their assays, related genes, and proteins.</p>
<p>The primary functionality revolves around fetching detailed information
about specified enzymes from PubChem, including assay data, gene properties,
protein properties, and compound properties. It processes this data into
a structured format suitable for knowledge graph construction, specifically
tailored for use with Neo4j databases.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.enzyme_list">enzyme_list</span></code></td>
            <td>
                  <code>list of str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Enzymes to query in the PubChem database.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor._base_url">_base_url</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Base URL for the PubChem API requests.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor._sep">_sep</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Delimiter for parsing CSV data from PubChem.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor._enzyme_count">_enzyme_count</span></code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of enzymes in the enzyme_list, calculated at</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>enzyme_list</code></td>
            <td>
                  <code>list of str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of enzyme names for which assay data</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>base_url</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Base URL for PubChem API requests. Defaults to</p>
              </div>
            </td>
            <td>
                  <code>&#39;https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/target/genesymbol&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>sep</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Separator used for parsing CSV data returned</p>
              </div>
            </td>
            <td>
                  <code>&#39;,&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


<details class="usage-example" open>
  <summary>Usage Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>enzyme_list = ['CYP2D6', 'CYP3A4']
extractor = NodePropertiesExtractor(enzyme_list)
df = extractor.run()
This example initiates the extractor with a list of enzymes, fetches
their data from PubChem, processes it, and potentially prepares it for
knowledge graph construction in Neo4j.</p>
</blockquote>
</blockquote>
</blockquote>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>To fully utilize this class, ensure you have network access to the
PubChem API for data retrieval and a Neo4j database instance for
knowledge graph construction. The class methods facilitate data extraction
and processing, but integrating the output into Neo4j requires additional
steps outside the scope of this class.</p>
</details>
              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">NodePropertiesExtractor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts data from PubChem to build knowledge graphs in Neo4j,</span>
<span class="sd">    focusing on nodes representing chemical entities and their relationships.</span>
<span class="sd">    This class serves as a bridge between the PubChem database and Neo4j,</span>
<span class="sd">    allowing users to query chemical data and construct a graph-based</span>
<span class="sd">    representation of chemical compounds, their assays, related genes, and proteins.</span>

<span class="sd">    The primary functionality revolves around fetching detailed information</span>
<span class="sd">    about specified enzymes from PubChem, including assay data, gene properties,</span>
<span class="sd">    protein properties, and compound properties. It processes this data into</span>
<span class="sd">    a structured format suitable for knowledge graph construction, specifically</span>
<span class="sd">    tailored for use with Neo4j databases.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        enzyme_list (list of str): Enzymes to query in the PubChem database.</span>
<span class="sd">        _base_url (str): Base URL for the PubChem API requests.</span>
<span class="sd">        _sep (str): Delimiter for parsing CSV data from PubChem.</span>
<span class="sd">        _enzyme_count (int): Number of enzymes in the enzyme_list, calculated at</span>
<span class="sd">        initialization.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        enzyme_list (list of str): List of enzyme names for which assay data</span>
<span class="sd">        will be fetched from PubChem.</span>
<span class="sd">        base_url (str, optional): Base URL for PubChem API requests. Defaults to</span>
<span class="sd">        the assay target genesymbol endpoint.</span>
<span class="sd">        sep (str, optional): Separator used for parsing CSV data returned</span>
<span class="sd">        by PubChem. Defaults to &#39;,&#39;.</span>

<span class="sd">    Usage Example:</span>
<span class="sd">        &gt;&gt;&gt; enzyme_list = [&#39;CYP2D6&#39;, &#39;CYP3A4&#39;]</span>
<span class="sd">        &gt;&gt;&gt; extractor = NodePropertiesExtractor(enzyme_list)</span>
<span class="sd">        &gt;&gt;&gt; df = extractor.run()</span>
<span class="sd">        This example initiates the extractor with a list of enzymes, fetches</span>
<span class="sd">        their data from PubChem, processes it, and potentially prepares it for</span>
<span class="sd">        knowledge graph construction in Neo4j.</span>

<span class="sd">    Note:</span>
<span class="sd">        To fully utilize this class, ensure you have network access to the</span>
<span class="sd">        PubChem API for data retrieval and a Neo4j database instance for</span>
<span class="sd">        knowledge graph construction. The class methods facilitate data extraction</span>
<span class="sd">        and processing, but integrating the output into Neo4j requires additional</span>
<span class="sd">        steps outside the scope of this class.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">_REQUEST_TIMEOUT</span> <span class="o">=</span> <span class="mi">30</span>  <span class="c1"># in seconds</span>
    <span class="n">_CONCURRENT_REQUEST_LIMIT</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">_RETRY_ATTEMPTS</span> <span class="o">=</span> <span class="mi">3</span>  <span class="c1"># number of times to retry a failed request</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">enzyme_list</span><span class="p">,</span>
                 <span class="n">base_url</span><span class="o">=</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/target/genesymbol&quot;</span><span class="p">,</span>
                 <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes a NodePropertiesExtractor instance, setting up the base URL</span>
<span class="sd">        for API requests, the separator for CSV parsing, and the list of enzymes</span>
<span class="sd">        to query from the PubChem database.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            enzyme_list (list of str): A list of enzyme names for which to fetch</span>
<span class="sd">            assay data.</span>
<span class="sd">            base_url (str, optional): The base URL for PubChem API requests.</span>
<span class="sd">            Default is set to the assay target genesymbol endpoint.</span>
<span class="sd">            sep (str, optional): The delimiter to use for parsing CSV files</span>
<span class="sd">            returned by PubChem. Defaults to &#39;,&#39;.</span>

<span class="sd">        Attributes:</span>
<span class="sd">            _base_url (str): Stores the base URL for API requests.</span>
<span class="sd">            _sep (str): Stores the delimiter for parsing CSV data.</span>
<span class="sd">            enzyme_list (list of str): Stores the list of enzyme names provided</span>
<span class="sd">            during initialization.</span>
<span class="sd">            _enzyme_count (int): The number of enzymes in the enzyme_list.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_base_url</span> <span class="o">=</span> <span class="n">base_url</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sep</span> <span class="o">=</span> <span class="n">sep</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span> <span class="o">=</span> <span class="n">enzyme_list</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_enzyme_count</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">enzyme_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_make_request</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">url</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sends an HTTP GET request to a specified URL with built-in retry logic.</span>
<span class="sd">        If the request fails, it retries the request up to a predefined number</span>
<span class="sd">        of attempts with exponential backoff to handle potential temporary network</span>
<span class="sd">        or server issues.</span>

<span class="sd">        The method attempts to gracefully handle server-side errors</span>
<span class="sd">        (HTTP 4XX/5XX responses) by raising an exception if the response status</span>
<span class="sd">        code indicates an error. For client-side errors (e.g., connectivity issues),</span>
<span class="sd">        it logs a warning and retries the request.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            url (str): The complete URL to which the HTTP GET request is sent.</span>

<span class="sd">        Returns:</span>
<span class="sd">            requests.Response: The response object from the server if the request</span>
<span class="sd">            is successfully completed.</span>

<span class="sd">        Raises:</span>
<span class="sd">            requests.RequestException: If the request fails to complete</span>
<span class="sd">            successfully after the maximum number of retry attempts.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">attempt</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_RETRY_ATTEMPTS</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_REQUEST_TIMEOUT</span><span class="p">)</span>
                <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>  <span class="c1"># Checks for HTTP errors</span>
                <span class="k">return</span> <span class="n">response</span>
            <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Attempt </span><span class="si">%s</span><span class="s2"> of </span><span class="si">%s</span><span class="s2"> failed for URL: </span><span class="si">%s</span><span class="s2">. Error: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                                <span class="n">attempt</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_RETRY_ATTEMPTS</span><span class="p">,</span> <span class="n">url</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">attempt</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_RETRY_ATTEMPTS</span><span class="p">:</span>
                    <span class="k">raise</span>  <span class="c1"># All attempts failed; re-raise the last exception</span>
                <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">2</span> <span class="o">**</span> <span class="n">attempt</span><span class="p">)</span>  <span class="c1"># Exponential backoff</span>

    <span class="k">def</span> <span class="nf">get_enzyme_assays</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches assay data for a specified enzyme from the PubChem database and</span>
<span class="sd">        returns it as a pandas DataFrame.</span>

<span class="sd">        This method constructs a URL to query the PubChem database for concise</span>
<span class="sd">        assay data related to the given enzyme. It processes the CSV response</span>
<span class="sd">        into a DataFrame, which includes various assay data points provided by PubChem.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            enzyme (str): The name of the enzyme for which assay data is</span>
<span class="sd">            requested. This name is used in the API query.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: A DataFrame containing the assay data fetched from</span>
<span class="sd">            PubChem for the specified enzyme. The DataFrame includes columns</span>
<span class="sd">            based on the CSV response from PubChem, such as assay ID, results,</span>
<span class="sd">            and conditions. Returns None if no data is available or if an error</span>
<span class="sd">            occurs during data fetching or processing.</span>

<span class="sd">        Raises:</span>
<span class="sd">            requests.RequestException: If an error occurs during the HTTP</span>
<span class="sd">            request to the PubChem API.</span>
<span class="sd">            pd.errors.EmptyDataError: If the response from PubChem contains no data.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;enzyme&#39;])</span>
<span class="sd">            &gt;&gt;&gt; enzyme_assays_df = extractor.get_enzyme_assays(&#39;enzyme&#39;)</span>
<span class="sd">            &gt;&gt;&gt; print(enzyme_assays_df.head())</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">assays_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_base_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">enzyme</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="si">}</span><span class="s2">/concise/CSV&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fetching assays for enzyme: </span><span class="si">{</span><span class="n">enzyme</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_make_request</span><span class="p">(</span><span class="n">assays_url</span><span class="p">)</span>

        <span class="n">assays_csv_string</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">text</span>
        <span class="n">assays_csv_string_io</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">assays_csv_string</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">assays_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">assays_csv_string_io</span><span class="p">,</span>
                                    <span class="n">sep</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_sep</span><span class="p">,</span>
                                    <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Assays DataFrame for enzyme </span><span class="si">%s</span><span class="s2"> has shape: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                         <span class="n">enzyme</span><span class="p">,</span> <span class="n">assays_df</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">assays_df</span>
        <span class="k">except</span> <span class="n">pd</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">EmptyDataError</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No data available for enzyme </span><span class="si">%s</span><span class="s2">.&quot;</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_process_enzymes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">enzyme_list</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Iterates over a list of enzyme names, fetching assay data for each enzyme</span>
<span class="sd">        and aggregating the results into a list of DataFrames.</span>

<span class="sd">        This method calls `get_enzyme_assays` for each enzyme in the provided</span>
<span class="sd">        list, collecting the assay data (if available) into a list of pandas</span>
<span class="sd">        DataFrames. This list can then be used for further processing or analysis.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            enzyme_list (list of str): A list containing the names of enzymes</span>
<span class="sd">            for which to fetch assay data.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list of pd.DataFrame: A list containing a pandas DataFrame for each</span>
<span class="sd">            enzyme for which assay data was successfully fetched and processed.</span>
<span class="sd">            Each DataFrame includes the assay data from PubChem for that enzyme.</span>
<span class="sd">            If no data is available for an enzyme, it is omitted from the list.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df_list</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">get_enzyme_assays</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span> <span class="k">for</span> <span class="n">enzyme</span> <span class="ow">in</span> <span class="n">enzyme_list</span><span class="p">]</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">df</span> <span class="k">for</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">df_list</span> <span class="k">if</span> <span class="n">df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">_concatenate_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df_list</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Concatenates a list of pandas DataFrames into a single DataFrame.</span>
<span class="sd">        This method is useful for aggregating</span>
<span class="sd">        data fetched from multiple sources or APIs into a unified structure.</span>
<span class="sd">        If the list is empty, it returns None to indicate that no data was</span>
<span class="sd">        aggregated.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            df_list (List[pd.DataFrame]): A list of pandas DataFrames to</span>
<span class="sd">            concatenate. These DataFrames should have the same structure</span>
<span class="sd">            (columns) to ensure proper concatenation.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame or None: A single concatenated DataFrame comprising all</span>
<span class="sd">            rows from the input DataFrames, indexed continuously. Returns None</span>
<span class="sd">            if the input list is empty, indicating there is no data to concatenate.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">df_list</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_list</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Orchestrates the process of fetching, filtering, and aggregating assay</span>
<span class="sd">        data from PubChem for a predefined list of enzymes.</span>

<span class="sd">        This method iteratively queries PubChem for assay data corresponding</span>
<span class="sd">        to each enzyme specified in the `enzyme_list` attribute during class</span>
<span class="sd">        initialization. It performs the following steps for each enzyme:</span>
<span class="sd">        1. Constructs a query URL and fetches assay data from PubChem.</span>
<span class="sd">        2. Filters the fetched data based on predefined criteria</span>
<span class="sd">        (e.g., containing specific substrings in the assay name).</span>
<span class="sd">        3. Aggregates the filtered data into a single pandas DataFrame.</span>
<span class="sd">        4. Identifies enzymes for which data could not be fetched or were</span>
<span class="sd">        excluded based on filtering criteria, logging their names.</span>

<span class="sd">        The final aggregated DataFrame, containing assay data for all successfully</span>
<span class="sd">        processed enzymes, is then saved to a CSV file. This method facilitates</span>
<span class="sd">        the extraction and preprocessing of chemical assay data for further</span>
<span class="sd">        analysis or integration into knowledge graphs.</span>

<span class="sd">        Note:</span>
<span class="sd">            - This method relies on the successful response from PubChem</span>
<span class="sd">            for each enzyme query.</span>
<span class="sd">            - Enzymes with no available data or failing to meet the filtering</span>
<span class="sd">            criteria are excluded from the final DataFrame.</span>
<span class="sd">            - The output CSV file is saved in the current working directory</span>
<span class="sd">            with the name &#39;Data/AllDataConnected.csv&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: A DataFrame containing the aggregated and filtered</span>
<span class="sd">            assay data for the specified enzymes. Columns in the DataFrame</span>
<span class="sd">            correspond to the assay data fields returned by PubChem, subject to</span>
<span class="sd">            the filtering criteria applied within this method.</span>

<span class="sd">        Raises:</span>
<span class="sd">            requests.RequestException: If there is an issue with fetching data</span>
<span class="sd">            from PubChem, such as a network problem or an invalid response.</span>

<span class="sd">        Example:</span>
<span class="sd">            Assuming `enzyme_list` was set to [&#39;CYP2D6&#39;, &#39;CYP3A4&#39;] during</span>
<span class="sd">            class initialization:</span>

<span class="sd">            &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">            &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">            &gt;&gt;&gt; result_df = extractor.run()</span>
<span class="sd">            &gt;&gt;&gt; print(result_df.head())</span>

<span class="sd">            This will fetch and process assay data for &#39;CYP2D6&#39; and &#39;CYP3A4&#39;,</span>
<span class="sd">            returning a DataFrame with the processed data.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Initialize an empty list to store enzymes with successful responses</span>
        <span class="n">enzymes_with_response</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Keep a copy of the original list to identify removed enzymes later</span>
        <span class="n">original_enzyme_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">enzyme</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span><span class="p">:</span>
            <span class="c1"># Formulate the URL</span>
            <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_base_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">enzyme</span><span class="si">}</span><span class="s2">/concise/CSV&quot;</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
                <span class="c1"># Check for a successful response (status code 200)</span>
                <span class="k">if</span> <span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
                    <span class="n">enzymes_with_response</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span>  <span class="c1"># Keep the enzyme in the new list</span>
            <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span><span class="p">:</span>
                <span class="c1"># If there&#39;s an exception, skip adding the enzyme to the new list</span>
                <span class="k">pass</span>

        <span class="c1"># Update the enzyme list with only the enzymes that had a successful response</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span> <span class="o">=</span> <span class="n">enzymes_with_response</span>

        <span class="c1"># Identify and print the removed enzymes</span>
        <span class="n">removed_enzymes</span> <span class="o">=</span> <span class="p">[</span><span class="n">enzyme</span> <span class="k">for</span> <span class="n">enzyme</span> <span class="ow">in</span> <span class="n">original_enzyme_list</span> <span class="k">if</span> <span class="n">enzyme</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">enzymes_with_response</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">removed_enzymes</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;These enzymes were removed because their names aren&#39;t correct: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                         <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">removed_enzymes</span><span class="p">))</span>

        <span class="n">df_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process_enzymes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_concatenate_data</span><span class="p">(</span><span class="n">df_list</span><span class="p">)</span>
        <span class="n">substrings_to_filter</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CYP&#39;</span><span class="p">,</span> <span class="s1">&#39;Cytochrome&#39;</span><span class="p">]</span>
        <span class="n">pattern</span> <span class="o">=</span> <span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">substrings_to_filter</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Assay Name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">pattern</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">)]</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Data/AllDataConnected.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>

    <span class="k">def</span> <span class="nf">_fetch_gene_details</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches gene details in parallel using the PubChem API.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_id (int): The gene ID for fetching details.</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: Contains gene ID, symbol, taxonomy, taxonomy ID, and synonyms.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">BASE_URL</span> <span class="o">=</span> <span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug&quot;</span>
        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">BASE_URL</span><span class="si">}</span><span class="s2">/gene/geneid/</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span><span class="si">}</span><span class="s2">/summary/JSON&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_make_request</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>

            <span class="c1"># Extracting the necessary details</span>
            <span class="n">symbol</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;GeneSummaries&#39;</span><span class="p">][</span><span class="s1">&#39;GeneSummary&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;Symbol&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="n">taxonomy</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;GeneSummaries&#39;</span><span class="p">][</span><span class="s1">&#39;GeneSummary&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;Taxonomy&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="n">taxonomy_id</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;GeneSummaries&#39;</span><span class="p">][</span><span class="s1">&#39;GeneSummary&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;TaxonomyID&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="n">synonyms</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;GeneSummaries&#39;</span><span class="p">][</span><span class="s1">&#39;GeneSummary&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;Synonym&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="c1"># print(type(synonyms))</span>
            <span class="k">return</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">symbol</span><span class="p">,</span> <span class="n">taxonomy</span><span class="p">,</span> <span class="n">taxonomy_id</span><span class="p">,</span> <span class="n">synonyms</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error fetching details for gene_id </span><span class="si">{</span><span class="n">gene_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">gene_id</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">extract_gene_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts and processes gene properties from a given data source,</span>
<span class="sd">        specifically targeting genes relevant to the study (e.g., CYP enzymes)</span>
<span class="sd">        and records their details in a structured DataFrame.</span>

<span class="sd">        This method reads gene data from a CSV file specified by `main_data`,</span>
<span class="sd">        queries the PubChem database for additional properties of each unique</span>
<span class="sd">        gene ID found in the file, and compiles these properties into a new</span>
<span class="sd">        DataFrame. It focuses on fetching details like gene symbols, taxonomy,</span>
<span class="sd">        taxonomy IDs, and synonyms for each gene. The final DataFrame is filtered</span>
<span class="sd">        to include only genes of particular interest (e.g., certain CYP enzymes)</span>
<span class="sd">        and saved to a separate CSV file for further analysis or use.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to a CSV file containing main data was which</span>
<span class="sd">            generated after running `extractor.run()`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: A DataFrame containing the compiled gene properties,</span>
<span class="sd">            including GeneID, Symbol, Taxonomy, Taxonomy ID, and Synonyms,</span>
<span class="sd">            filtered to include only specified genes of interest. This DataFrame</span>
<span class="sd">            is also saved to &#39;Data/Nodes/Gene_Properties.csv&#39;.</span>

<span class="sd">        Raises:</span>
<span class="sd">            Exception: If there&#39;s an issue reading the initial CSV file or</span>
<span class="sd">            fetching gene details from PubChem, details of the exception are</span>
<span class="sd">            logged, and the method proceeds to process the next gene ID.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">            &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">            &gt;&gt;&gt; extractor.run()</span>
<span class="sd">            &gt;&gt;&gt; gene_properties_df = extractor.extract_gene_properties(&#39;Data/AllDataConnected.csv&#39;)</span>
<span class="sd">            &gt;&gt;&gt; print(gene_properties_df.head())</span>

<span class="sd">            This would read gene IDs from &#39;Data/AllDataConnected.csv&#39;, fetch</span>
<span class="sd">            their properties from PubChem, and compile the details into a</span>
<span class="sd">            DataFrame, filtering for specified genes of interest and saving</span>
<span class="sd">            the results to &#39;Data/Nodes/Gene_Properties.csv&#39;.</span>

<span class="sd">        Note:</span>
<span class="sd">            The method filters the resulting DataFrame to include only genes with</span>
<span class="sd">            symbols in the predefined enzyme_list. Adjust this list as necessary</span>
<span class="sd">            to match the focus of your study or application.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
        <span class="n">df_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Symbol&#39;</span><span class="p">,</span> <span class="s1">&#39;Taxonomy&#39;</span><span class="p">,</span>
                                        <span class="s1">&#39;Taxonomy ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Synonyms&#39;</span><span class="p">])</span>

        <span class="n">unique_gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="n">gene_details</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">unique_gene_ids</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">gene_id</span><span class="p">,</span> <span class="n">symbol</span><span class="p">,</span> <span class="n">taxonomy</span><span class="p">,</span> <span class="n">taxonomy_id</span><span class="p">,</span> <span class="n">synonyms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fetch_gene_details</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>
                <span class="n">gene_details</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                    <span class="s1">&#39;GeneID&#39;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">,</span>
                    <span class="s1">&#39;Symbol&#39;</span><span class="p">:</span> <span class="n">symbol</span><span class="p">,</span>
                    <span class="s1">&#39;Taxonomy&#39;</span><span class="p">:</span> <span class="n">taxonomy</span><span class="p">,</span>
                    <span class="s1">&#39;Taxonomy ID&#39;</span><span class="p">:</span> <span class="n">taxonomy_id</span><span class="p">,</span>
                    <span class="s1">&#39;Synonyms&#39;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">synonyms</span><span class="p">)</span>
                <span class="p">})</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Error occurred while processing gene_id </span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                              <span class="n">gene_id</span><span class="p">,</span> <span class="n">exc</span><span class="p">)</span>
                <span class="n">gene_details</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                    <span class="s1">&#39;GeneID&#39;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">,</span>
                    <span class="s1">&#39;Symbol&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="s1">&#39;Taxonomy&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="s1">&#39;Taxonomy ID&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="s1">&#39;Synonyms&#39;</span><span class="p">:</span> <span class="kc">None</span>
                <span class="p">})</span>

        <span class="c1"># Now create the DataFrame from the list of dictionaries</span>
        <span class="n">df_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_details</span><span class="p">)</span>
        <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_enzyme_count</span>
        <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">df_gene</span> <span class="o">=</span> <span class="n">df_gene</span><span class="p">[</span><span class="n">df_gene</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">])]</span>
        <span class="n">df_gene</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Data/Nodes/Gene_Properties.csv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df_gene</span>

    <span class="k">def</span> <span class="nf">_fetch_assay_details</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">aid</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches assay details from the PubChem API for a given assay ID.</span>

<span class="sd">        Args:</span>
<span class="sd">            aid (int): The assay ID to fetch details for.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary containing assay details like AID, SourceName,</span>
<span class="sd">            SourceID, Name, and Description. Returns None if an error occurs</span>
<span class="sd">            during fetching or parsing.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">BASE_URL</span> <span class="o">=</span> <span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug&quot;</span>
        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">BASE_URL</span><span class="si">}</span><span class="s2">/assay/aid/</span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">/summary/XML&quot;</span>  <span class="c1"># Constructing the API URL</span>
        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_make_request</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>  <span class="c1"># Making the API request</span>
        <span class="n">xml_data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">text</span>  <span class="c1"># Getting the response text</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Parsing the XML response</span>
            <span class="n">data_dict</span> <span class="o">=</span> <span class="n">xmltodict</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">xml_data</span><span class="p">)</span>
            <span class="n">properties</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">,</span> <span class="s1">&#39;SourceName&#39;</span><span class="p">,</span> <span class="s1">&#39;SourceID&#39;</span><span class="p">,</span> <span class="s1">&#39;Name&#39;</span><span class="p">,</span> <span class="s1">&#39;Description&#39;</span><span class="p">,</span>
                          <span class="s1">&#39;Protocol&#39;</span><span class="p">,</span> <span class="s1">&#39;Comment&#39;</span><span class="p">,</span> <span class="s1">&#39;Method&#39;</span><span class="p">,</span> <span class="s1">&#39;Target&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountAll&#39;</span><span class="p">,</span>
                          <span class="s1">&#39;CIDCountActive&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountInactive&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountInconclusive&#39;</span><span class="p">,</span>
                          <span class="s1">&#39;CIDCountUnspecified&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountProbe&#39;</span><span class="p">]</span>

            <span class="n">assay_data</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="c1"># Extracting required properties from the parsed XML</span>
            <span class="k">for</span> <span class="n">prop</span> <span class="ow">in</span> <span class="n">properties</span><span class="p">:</span>
                <span class="n">assay_data</span><span class="p">[</span><span class="n">prop</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;AssaySummaries&#39;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;AssaySummary&#39;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">prop</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">assay_data</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error parsing XML for AID </span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">extract_assay_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts detailed properties of assays from PubChem for each unique assay</span>
<span class="sd">        ID found in the input data file.</span>

<span class="sd">        This method processes an input CSV file containing assay IDs (AID) and</span>
<span class="sd">        performs concurrent HTTP requests to fetch detailed assay properties</span>
<span class="sd">        from the PubChem database. The retrieved details include assay type,</span>
<span class="sd">        activity name, source name, source ID, name, and description. These</span>
<span class="sd">        properties are compiled into a new DataFrame, which is then</span>
<span class="sd">        saved to a CSV file for further analysis or use.</span>

<span class="sd">        The method employs a ThreadPoolExecutor to manage concurrent requests</span>
<span class="sd">        efficiently, improving the performance when dealing with a large number</span>
<span class="sd">        of assay IDs. Errors encountered during data fetching are logged, and the</span>
<span class="sd">        process continues with the next assay ID, ensuring the method&#39;s robustness.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to a CSV file containing main data was which</span>
<span class="sd">            generated after running `extractor.run()`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: A DataFrame containing the fetched assay properties,</span>
<span class="sd">            including columns for AID, Assay Type, Activity Name, SourceName,</span>
<span class="sd">            SourceID, Name, and Description. This DataFrame is saved to</span>
<span class="sd">            &#39;Data/Nodes/Assay_Properties.csv&#39; in the current working directory.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If the input CSV file is empty or does not contain the &#39;AID&#39; column.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">            &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">            &gt;&gt;&gt; extractor.run()</span>
<span class="sd">            &gt;&gt;&gt; assay_properties_df = extractor.extract_assay_properties(&#39;Data/AllDataConnected.csv&#39;)</span>
<span class="sd">            &gt;&gt;&gt; print(assay_properties_df.head())</span>

<span class="sd">            This example reads assay IDs from &#39;Data/AllDataConnected.csv&#39;,</span>
<span class="sd">            queries PubChem for their detailed properties, and compiles the</span>
<span class="sd">            results into a DataFrame, which is also saved to &#39;Data/Nodes/Assay_Properties.csv&#39;.</span>

<span class="sd">        Note:</span>
<span class="sd">            This method requires network access to the PubChem API and assumes</span>
<span class="sd">            the availability of a valid &#39;AID&#39; column in the input CSV file.</span>
<span class="sd">            Ensure the input file path is correct and accessible to avoid errors during processing.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>

        <span class="c1"># Check if the DataFrame is valid</span>
        <span class="k">if</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span> <span class="ow">or</span> <span class="s1">&#39;AID&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;DataFrame is empty or does not contain &#39;AID&#39; column.&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

        <span class="n">unique_aids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># Extracting unique assay IDs</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">,</span> <span class="s1">&#39;Assay Type&#39;</span><span class="p">,</span> <span class="s1">&#39;Activity Name&#39;</span><span class="p">,</span> <span class="s1">&#39;SourceName&#39;</span><span class="p">,</span>
                   <span class="s1">&#39;SourceID&#39;</span><span class="p">,</span> <span class="s1">&#39;Name&#39;</span><span class="p">,</span> <span class="s1">&#39;Description&#39;</span><span class="p">,</span> <span class="s1">&#39;Protocol&#39;</span><span class="p">,</span>
                   <span class="s1">&#39;Comment&#39;</span><span class="p">,</span> <span class="s1">&#39;Method&#39;</span><span class="p">,</span> <span class="s1">&#39;Target&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountAll&#39;</span><span class="p">,</span>
                   <span class="s1">&#39;CIDCountActive&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountInactive&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountInconclusive&#39;</span><span class="p">,</span>
                   <span class="s1">&#39;CIDCountUnspecified&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountProbe&#39;</span><span class="p">]</span>
        <span class="n">assay_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>  <span class="c1"># Initializing a DataFrame to store assay properties</span>

        <span class="c1"># Using ThreadPoolExecutor for concurrent fetching of assay details</span>
        <span class="k">with</span> <span class="n">concurrent</span><span class="o">.</span><span class="n">futures</span><span class="o">.</span><span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_CONCURRENT_REQUEST_LIMIT</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
            <span class="n">future_to_aid</span> <span class="o">=</span> <span class="p">{</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_fetch_assay_details</span><span class="p">,</span> <span class="n">aid</span><span class="p">):</span> <span class="n">aid</span> <span class="k">for</span> <span class="n">aid</span> <span class="ow">in</span> <span class="n">unique_aids</span><span class="p">}</span>

            <span class="c1"># Iterating over completed futures</span>
            <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">concurrent</span><span class="o">.</span><span class="n">futures</span><span class="o">.</span><span class="n">as_completed</span><span class="p">(</span><span class="n">future_to_aid</span><span class="p">):</span>
                <span class="n">aid</span> <span class="o">=</span> <span class="n">future_to_aid</span><span class="p">[</span><span class="n">future</span><span class="p">]</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">assay_data</span> <span class="o">=</span> <span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span>  <span class="c1"># Fetching the result from the future</span>
                    <span class="k">if</span> <span class="n">assay_data</span><span class="p">:</span>
                        <span class="c1"># Preparing a new row with the fetched data</span>
                        <span class="n">new_row</span> <span class="o">=</span> <span class="p">{</span>
                            <span class="s1">&#39;AID&#39;</span><span class="p">:</span> <span class="n">aid</span><span class="p">,</span>
                            <span class="s1">&#39;Assay Type&#39;</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">aid</span><span class="p">,</span> <span class="s1">&#39;Assay Type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="s1">&#39;Activity Name&#39;</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">aid</span><span class="p">,</span> <span class="s1">&#39;Activity Name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="o">**</span><span class="n">assay_data</span>
                        <span class="p">}</span>
                        <span class="c1"># Adding the new row to the DataFrame</span>
                        <span class="n">assay_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">assay_df</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">new_row</span><span class="p">])],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
                    <span class="c1"># Logging any errors encountered during the fetch</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error occurred while processing AID </span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">exc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Saving the updated DataFrame to a CSV file</span>
        <span class="n">assay_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Data/Nodes/Assay_Properties.csv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">assay_df</span>

    <span class="k">def</span> <span class="nf">extract_protein_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts and compiles protein properties from the NCBI protein database</span>
<span class="sd">        based on accession numbers.</span>

<span class="sd">        Given a CSV file specified by `main_data`, this method reads protein</span>
<span class="sd">        accession numbers and performs web scraping on the NCBI protein database</span>
<span class="sd">        pages to extract protein titles. The method constructs a URL for</span>
<span class="sd">        each accession number, sends a request to retrieve the page content,</span>
<span class="sd">        and parses the HTML to find the protein title. The extracted titles,</span>
<span class="sd">        along with their corresponding accession numbers and URLs, are</span>
<span class="sd">        compiled into a DataFrame. This DataFrame is saved to a CSV file,</span>
<span class="sd">        providing a structured summary of protein properties for further analysis or use.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to a CSV file containing main data was which</span>
<span class="sd">            generated after running `extractor.run()`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: A DataFrame with columns &#39;RefSeq Accession&#39;, &#39;URL&#39;,</span>
<span class="sd">            and &#39;Description&#39;, where &#39;Description&#39; contains the title of the</span>
<span class="sd">            protein extracted from its NCBI page. This DataFrame is saved to</span>
<span class="sd">            &#39;Data/Nodes/Protein_Properties.csv&#39; in the current working directory.</span>

<span class="sd">        Raises:</span>
<span class="sd">            Exception: If there&#39;s an issue reading the initial CSV file or</span>
<span class="sd">            querying the NCBI database, details of the exception are logged.</span>
<span class="sd">            The method continues processing the next accession number,</span>
<span class="sd">            ensuring robustness against individual failures.</span>

<span class="sd">        Example:</span>
<span class="sd">            Assuming &#39;protein_data.csv&#39; contains a column &#39;Target Accession&#39;</span>
<span class="sd">            with accession numbers:</span>

<span class="sd">            &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">            &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">            &gt;&gt;&gt; extractor.run() # you need to run this only once</span>
<span class="sd">            &gt;&gt;&gt; protein_properties_df = extractor.extract_protein_properties(&#39;Data/AllDataConnected.csv&#39;)</span>
<span class="sd">            &gt;&gt;&gt; print(protein_properties_df.head())</span>

<span class="sd">            This would read accession numbers from &#39;Data/AllDataConnected.csv&#39;,</span>
<span class="sd">            scrape their titles from the NCBI protein database, and compile the</span>
<span class="sd">            results into a DataFrame, which is also saved to</span>
<span class="sd">            &#39;Data/Nodes/Protein_Properties.csv&#39;.</span>

<span class="sd">        Note:</span>
<span class="sd">            This method requires internet access to query the NCBI protein</span>
<span class="sd">            database. Ensure the input file path is correct and accessible to</span>
<span class="sd">            avoid errors during processing. Web scraping is dependent on the</span>
<span class="sd">            structure of the web page; changes to the NCBI protein database</span>
<span class="sd">            pages may require updates to the scraping logic.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Initialize a list to store the extracted data</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_enzyme_count</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
        <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">])]</span>
        <span class="n">Accessions</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target Accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="c1"># Iterate over each protein accession number in the DataFrame</span>
        <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">Accessions</span><span class="p">:</span>
            <span class="c1"># Construct the URL to query the NCBI protein database</span>
            <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ncbi.nlm.nih.gov/protein/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># Send an HTTP request to the URL</span>
                <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>

                <span class="c1"># Parse the HTML content of the response</span>
                <span class="n">soup</span> <span class="o">=</span> <span class="n">BeautifulSoup</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">,</span> <span class="s1">&#39;html.parser&#39;</span><span class="p">)</span>

                <span class="c1"># Extract the title from the parsed HTML</span>
                <span class="n">title</span> <span class="o">=</span> <span class="n">soup</span><span class="o">.</span><span class="n">title</span><span class="o">.</span><span class="n">string</span> <span class="k">if</span> <span class="n">soup</span><span class="o">.</span><span class="n">title</span> <span class="k">else</span> <span class="s1">&#39;Title Not Found&#39;</span>

                <span class="c1"># Append the extracted data to the list</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;RefSeq Accession&#39;</span><span class="p">:</span> <span class="n">accession</span><span class="p">,</span>
                             <span class="s1">&#39;URL&#39;</span><span class="p">:</span> <span class="n">url</span><span class="p">,</span> <span class="s1">&#39;Description&#39;</span><span class="p">:</span> <span class="n">title</span><span class="p">})</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="c1"># In case of an error, log the error message</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error fetching data for accession </span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;RefSeq Accession&#39;</span><span class="p">:</span> <span class="n">accession</span><span class="p">,</span> <span class="s1">&#39;URL&#39;</span><span class="p">:</span> <span class="n">url</span><span class="p">,</span>
                             <span class="s1">&#39;Description&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;Error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">})</span>

        <span class="c1"># Convert the list of data into a DataFrame</span>
        <span class="n">protein_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="c1"># Save the DataFrame to a CSV file</span>
        <span class="n">protein_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Data/Nodes/Protein_Properties.csv&#39;</span><span class="p">,</span>
                          <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Return the DataFrame</span>
        <span class="k">return</span> <span class="n">protein_df</span>

    <span class="k">def</span> <span class="nf">fetch_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves detailed chemical compound properties for a specified</span>
<span class="sd">        Compound ID (CID) from the PubChem database.</span>

<span class="sd">        This method constructs a query URL to fetch a wide range of properties</span>
<span class="sd">        for the given CID from PubChem, including molecular formula,</span>
<span class="sd">        molecular weight, canonical and isomeric SMILES, InChI codes,</span>
<span class="sd">        physicochemical properties, and more. If the CID is valid and data is</span>
<span class="sd">        available, it returns a pandas DataFrame containing these properties. This</span>
<span class="sd">        method also generates a URL to retrieve the structure image of the</span>
<span class="sd">        compound as a 2D PNG image, adding it as a column in the DataFrame.</span>
<span class="sd">        In cases where the CID is NaN or an error occurs during data retrieval,</span>
<span class="sd">        an empty DataFrame is returned.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            cid (int or float): The Compound ID for which to fetch data.</span>
<span class="sd">            Can be an integer or NaN.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: A DataFrame containing the fetched properties for the</span>
<span class="sd">            given CID. The DataFrame includes columns for each property fetched</span>
<span class="sd">            from PubChem, along with a &#39;StructureImage2DURL&#39; column containing</span>
<span class="sd">            the URL to the compound&#39;s structure image. Returns an empty DataFrame</span>
<span class="sd">            if the CID is NaN or if any error occurs during the fetch operation.</span>

<span class="sd">        Raises:</span>
<span class="sd">            Exception: Logs an error message if the request to PubChem fails or</span>
<span class="sd">            if the response cannot be processed into a DataFrame.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">            &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">            &gt;&gt;&gt; compound_data_df = extractor.fetch_data(2244)</span>
<span class="sd">            &gt;&gt;&gt; print(compound_data_df.head())</span>

<span class="sd">            This example fetches the properties for the compound with CID 2244</span>
<span class="sd">            from PubChem and prints the first few rows</span>
<span class="sd">            of the resulting DataFrame.</span>

<span class="sd">        Note:</span>
<span class="sd">            This method requires an active internet connection to access the</span>
<span class="sd">            PubChem database. Ensure that the CID provided is valid and not NaN</span>
<span class="sd">            to avoid fetching errors. The structure and availability of data</span>
<span class="sd">            fields are subject to the current state of the PubChem database</span>
<span class="sd">            and may vary.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">cid</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># Return an empty DataFrame for NaN CIDs</span>

        <span class="n">cid</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span>  <span class="c1"># Convert CID to integer</span>
        <span class="n">url</span> <span class="o">=</span> <span class="p">(</span><span class="sa">f</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">/property/&quot;</span>
               <span class="s2">&quot;MolecularFormula,MolecularWeight,CanonicalSMILES,IsomericSMILES,InChI,&quot;</span>
               <span class="s2">&quot;InChIKey,IUPACName,Title,XLogP,ExactMass,MonoisotopicMass,TPSA,Complexity,&quot;</span>
               <span class="s2">&quot;Charge,HBondDonorCount,HBondAcceptorCount,RotatableBondCount,HeavyAtomCount,&quot;</span>
               <span class="s2">&quot;IsotopeAtomCount,AtomStereoCount,DefinedAtomStereoCount,UndefinedAtomStereoCount,&quot;</span>
               <span class="s2">&quot;BondStereoCount,DefinedBondStereoCount,UndefinedBondStereoCount,CovalentUnitCount,&quot;</span>
               <span class="s2">&quot;PatentCount,PatentFamilyCount,LiteratureCount,Volume3D,XStericQuadrupole3D,&quot;</span>
               <span class="s2">&quot;YStericQuadrupole3D,ZStericQuadrupole3D,FeatureCount3D,FeatureAcceptorCount3D,&quot;</span>
               <span class="s2">&quot;FeatureDonorCount3D,FeatureAnionCount3D,FeatureCationCount3D,FeatureRingCount3D,&quot;</span>
               <span class="s2">&quot;FeatureHydrophobeCount3D,ConformerModelRMSD3D,EffectiveRotorCount3D,ConformerCount3D,&quot;</span>
               <span class="s2">&quot;Fingerprint2D/CSV&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
            <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
            <span class="n">compound_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">),</span>
                                        <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">compound_data</span><span class="p">[</span><span class="s1">&#39;StructureImage2DURL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">/PNG&quot;</span>
            <span class="k">return</span> <span class="n">compound_data</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing CID </span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># Return an empty DataFrame in case of error</span>

    <span class="k">def</span> <span class="nf">extract_compound_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts and aggregates compound properties from PubChem for a list of</span>
<span class="sd">        compounds associated with specific genes.</span>

<span class="sd">        This method processes a CSV file specified by `main_data`, which contains</span>
<span class="sd">        gene identifiers and their associated compound IDs (CIDs). It selects</span>
<span class="sd">        compounds related to the top `n` most frequently occurring genes in the</span>
<span class="sd">        dataset, where `n` is determined by the instance&#39;s `_enzyme_count`</span>
<span class="sd">        attribute. The method then fetches detailed compound properties from</span>
<span class="sd">        PubChem in chunks, using concurrent requests to improve efficiency and</span>
<span class="sd">        manage the load on the PubChem API. The fetched compound properties are</span>
<span class="sd">        aggregated into a single DataFrame and saved to multiple CSV files,</span>
<span class="sd">        one for each chunk of compound IDs processed.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to a CSV file containing main data was which</span>
<span class="sd">            generated after running `extractor.run()`.</span>

<span class="sd">        Side Effects:</span>
<span class="sd">            - Saves the aggregated compound properties to CSV files in the current</span>
<span class="sd">            working directory. The files are named</span>
<span class="sd">            &#39;Data/Nodes/Compound_Properties/Chunk_{i}.csv&#39;, where `{i}` is</span>
<span class="sd">            the chunk index.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None: This method does not return a value. Instead, it saves the</span>
<span class="sd">            fetched compound data directly to CSV files.</span>

<span class="sd">        Raises:</span>
<span class="sd">            Exception: Logs an error and continues processing the next CID if</span>
<span class="sd">            an error occurs while fetching data for a specific CID.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">            &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">            &gt;&gt;&gt; extractor.extract_compound_properties(&#39;Data/AllDataConnected.csv&#39;)</span>
<span class="sd">            This will read &#39;Data/AllDataConnected.csv&#39;, filter for compounds</span>
<span class="sd">            associated with the top n genes, fetch their properties from PubChem,</span>
<span class="sd">            and save the results into multiple CSV files for each chunk</span>
<span class="sd">            of compounds processed.</span>

<span class="sd">        Note:</span>
<span class="sd">            - Ensure that the &#39;main_data&#39; CSV file exists and is accessible at</span>
<span class="sd">            the specified path.</span>
<span class="sd">            - The method automatically handles NaN values in the &#39;CID&#39; column</span>
<span class="sd">            and excludes them from processing.</span>
<span class="sd">            - The `enzyme_count` attribute determines the number of top genes</span>
<span class="sd">            for which compound properties will be fetched.</span>
<span class="sd">            - Internet access is required to fetch compound data from the PubChem API.</span>
<span class="sd">            - The method employs a `ThreadPoolExecutor` with a configurable</span>
<span class="sd">            number of workers (default is len(enzyme_list)) to parallelize</span>
<span class="sd">            requests, which can be adjusted based on system capabilities and</span>
<span class="sd">            API rate limits.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_enzyme_count</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
        <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">])]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">])</span>
        <span class="n">IDs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="c1"># Define chunk size and calculate number of chunks</span>
        <span class="n">chunk_size</span> <span class="o">=</span> <span class="mi">10000</span>
        <span class="n">num_chunks</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">IDs</span><span class="p">)</span> <span class="o">/</span> <span class="n">chunk_size</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">start_chunk</span> <span class="o">&lt;</span> <span class="n">num_chunks</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start_chunk</span><span class="p">,</span> <span class="n">num_chunks</span><span class="p">):</span>
                <span class="c1"># Calculate start and end indices for each chunk</span>
                <span class="n">start_index</span> <span class="o">=</span> <span class="n">i</span> <span class="o">*</span> <span class="n">chunk_size</span>
                <span class="n">end_index</span> <span class="o">=</span> <span class="n">start_index</span> <span class="o">+</span> <span class="n">chunk_size</span>

                <span class="c1"># Extract chunk of CIDs</span>
                <span class="n">chunk_cids</span> <span class="o">=</span> <span class="n">IDs</span><span class="p">[</span><span class="n">start_index</span><span class="p">:</span><span class="n">end_index</span><span class="p">]</span>
                <span class="c1"># chunk_cids = [x for x in chunk_cids if not np.isnan(x)]</span>

                <span class="c1"># Use ThreadPoolExecutor to parallelize requests for the chunk</span>
                <span class="k">with</span> <span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
                    <span class="n">future_to_cid</span> <span class="o">=</span> <span class="p">{</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fetch_data</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span> <span class="n">cid</span> <span class="k">for</span> <span class="n">cid</span> <span class="ow">in</span> <span class="n">chunk_cids</span><span class="p">}</span>
                    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>

                    <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">as_completed</span><span class="p">(</span><span class="n">future_to_cid</span><span class="p">):</span>
                        <span class="n">cid</span> <span class="o">=</span> <span class="n">future_to_cid</span><span class="p">[</span><span class="n">future</span><span class="p">]</span>
                        <span class="k">try</span><span class="p">:</span>
                            <span class="n">data</span> <span class="o">=</span> <span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span>
                            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
                        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing CID </span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="c1"># Concatenate results for the current chunk</span>
                <span class="n">chunk_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

                <span class="c1"># Save the concatenated DataFrame to a CSV file for the chunk</span>
                <span class="n">chunk_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Nodes/Compound_Properties/Chunk_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span>
                                <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No more chunks to process.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">enzyme_list</span><span class="p">,</span> <span class="n">base_url</span><span class="o">=</span><span class="s1">&#39;https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/target/genesymbol&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Initializes a NodePropertiesExtractor instance, setting up the base URL
for API requests, the separator for CSV parsing, and the list of enzymes
to query from the PubChem database.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>enzyme_list</code></td>
            <td>
                  <code>list of str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of enzyme names for which to fetch</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>base_url</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The base URL for PubChem API requests.</p>
              </div>
            </td>
            <td>
                  <code>&#39;https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/target/genesymbol&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>sep</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The delimiter to use for parsing CSV files</p>
              </div>
            </td>
            <td>
                  <code>&#39;,&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.__init__._base_url">_base_url</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Stores the base URL for API requests.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.__init__._sep">_sep</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Stores the delimiter for parsing CSV data.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.__init__.enzyme_list">enzyme_list</span></code></td>
            <td>
                  <code>list of str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Stores the list of enzyme names provided</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.__init__._enzyme_count">_enzyme_count</span></code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of enzymes in the enzyme_list.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">enzyme_list</span><span class="p">,</span>
             <span class="n">base_url</span><span class="o">=</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/target/genesymbol&quot;</span><span class="p">,</span>
             <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes a NodePropertiesExtractor instance, setting up the base URL</span>
<span class="sd">    for API requests, the separator for CSV parsing, and the list of enzymes</span>
<span class="sd">    to query from the PubChem database.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        enzyme_list (list of str): A list of enzyme names for which to fetch</span>
<span class="sd">        assay data.</span>
<span class="sd">        base_url (str, optional): The base URL for PubChem API requests.</span>
<span class="sd">        Default is set to the assay target genesymbol endpoint.</span>
<span class="sd">        sep (str, optional): The delimiter to use for parsing CSV files</span>
<span class="sd">        returned by PubChem. Defaults to &#39;,&#39;.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        _base_url (str): Stores the base URL for API requests.</span>
<span class="sd">        _sep (str): Stores the delimiter for parsing CSV data.</span>
<span class="sd">        enzyme_list (list of str): Stores the list of enzyme names provided</span>
<span class="sd">        during initialization.</span>
<span class="sd">        _enzyme_count (int): The number of enzymes in the enzyme_list.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_base_url</span> <span class="o">=</span> <span class="n">base_url</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_sep</span> <span class="o">=</span> <span class="n">sep</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span> <span class="o">=</span> <span class="n">enzyme_list</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_enzyme_count</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">enzyme_list</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_assay_properties" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_assay_properties</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Extracts detailed properties of assays from PubChem for each unique assay
ID found in the input data file.</p>
<p>This method processes an input CSV file containing assay IDs (AID) and
performs concurrent HTTP requests to fetch detailed assay properties
from the PubChem database. The retrieved details include assay type,
activity name, source name, source ID, name, and description. These
properties are compiled into a new DataFrame, which is then
saved to a CSV file for further analysis or use.</p>
<p>The method employs a ThreadPoolExecutor to manage concurrent requests
efficiently, improving the performance when dealing with a large number
of assay IDs. Errors encountered during data fetching are logged, and the
process continues with the next assay ID, ensuring the method's robustness.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to a CSV file containing main data was which</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the fetched assay properties,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>including columns for AID, Assay Type, Activity Name, SourceName,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>SourceID, Name, and Description. This DataFrame is saved to</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>'Data/Nodes/Assay_Properties.csv' in the current working directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the input CSV file is empty or does not contain the 'AID' column.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = NodePropertiesExtractor(['CYP2D6', 'CYP3A4'])
extractor.create_data_directories()
extractor.run()
assay_properties_df = extractor.extract_assay_properties('Data/AllDataConnected.csv')
print(assay_properties_df.head())</p>
</blockquote>
</blockquote>
</blockquote>
<p>This example reads assay IDs from 'Data/AllDataConnected.csv',
queries PubChem for their detailed properties, and compiles the
results into a DataFrame, which is also saved to 'Data/Nodes/Assay_Properties.csv'.</p>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>This method requires network access to the PubChem API and assumes
the availability of a valid 'AID' column in the input CSV file.
Ensure the input file path is correct and accessible to avoid errors during processing.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_assay_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts detailed properties of assays from PubChem for each unique assay</span>
<span class="sd">    ID found in the input data file.</span>

<span class="sd">    This method processes an input CSV file containing assay IDs (AID) and</span>
<span class="sd">    performs concurrent HTTP requests to fetch detailed assay properties</span>
<span class="sd">    from the PubChem database. The retrieved details include assay type,</span>
<span class="sd">    activity name, source name, source ID, name, and description. These</span>
<span class="sd">    properties are compiled into a new DataFrame, which is then</span>
<span class="sd">    saved to a CSV file for further analysis or use.</span>

<span class="sd">    The method employs a ThreadPoolExecutor to manage concurrent requests</span>
<span class="sd">    efficiently, improving the performance when dealing with a large number</span>
<span class="sd">    of assay IDs. Errors encountered during data fetching are logged, and the</span>
<span class="sd">    process continues with the next assay ID, ensuring the method&#39;s robustness.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to a CSV file containing main data was which</span>
<span class="sd">        generated after running `extractor.run()`.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the fetched assay properties,</span>
<span class="sd">        including columns for AID, Assay Type, Activity Name, SourceName,</span>
<span class="sd">        SourceID, Name, and Description. This DataFrame is saved to</span>
<span class="sd">        &#39;Data/Nodes/Assay_Properties.csv&#39; in the current working directory.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the input CSV file is empty or does not contain the &#39;AID&#39; column.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">        &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">        &gt;&gt;&gt; extractor.run()</span>
<span class="sd">        &gt;&gt;&gt; assay_properties_df = extractor.extract_assay_properties(&#39;Data/AllDataConnected.csv&#39;)</span>
<span class="sd">        &gt;&gt;&gt; print(assay_properties_df.head())</span>

<span class="sd">        This example reads assay IDs from &#39;Data/AllDataConnected.csv&#39;,</span>
<span class="sd">        queries PubChem for their detailed properties, and compiles the</span>
<span class="sd">        results into a DataFrame, which is also saved to &#39;Data/Nodes/Assay_Properties.csv&#39;.</span>

<span class="sd">    Note:</span>
<span class="sd">        This method requires network access to the PubChem API and assumes</span>
<span class="sd">        the availability of a valid &#39;AID&#39; column in the input CSV file.</span>
<span class="sd">        Ensure the input file path is correct and accessible to avoid errors during processing.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>

    <span class="c1"># Check if the DataFrame is valid</span>
    <span class="k">if</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span> <span class="ow">or</span> <span class="s1">&#39;AID&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;DataFrame is empty or does not contain &#39;AID&#39; column.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="n">unique_aids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># Extracting unique assay IDs</span>

    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">,</span> <span class="s1">&#39;Assay Type&#39;</span><span class="p">,</span> <span class="s1">&#39;Activity Name&#39;</span><span class="p">,</span> <span class="s1">&#39;SourceName&#39;</span><span class="p">,</span>
               <span class="s1">&#39;SourceID&#39;</span><span class="p">,</span> <span class="s1">&#39;Name&#39;</span><span class="p">,</span> <span class="s1">&#39;Description&#39;</span><span class="p">,</span> <span class="s1">&#39;Protocol&#39;</span><span class="p">,</span>
               <span class="s1">&#39;Comment&#39;</span><span class="p">,</span> <span class="s1">&#39;Method&#39;</span><span class="p">,</span> <span class="s1">&#39;Target&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountAll&#39;</span><span class="p">,</span>
               <span class="s1">&#39;CIDCountActive&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountInactive&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountInconclusive&#39;</span><span class="p">,</span>
               <span class="s1">&#39;CIDCountUnspecified&#39;</span><span class="p">,</span> <span class="s1">&#39;CIDCountProbe&#39;</span><span class="p">]</span>
    <span class="n">assay_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>  <span class="c1"># Initializing a DataFrame to store assay properties</span>

    <span class="c1"># Using ThreadPoolExecutor for concurrent fetching of assay details</span>
    <span class="k">with</span> <span class="n">concurrent</span><span class="o">.</span><span class="n">futures</span><span class="o">.</span><span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_CONCURRENT_REQUEST_LIMIT</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
        <span class="n">future_to_aid</span> <span class="o">=</span> <span class="p">{</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_fetch_assay_details</span><span class="p">,</span> <span class="n">aid</span><span class="p">):</span> <span class="n">aid</span> <span class="k">for</span> <span class="n">aid</span> <span class="ow">in</span> <span class="n">unique_aids</span><span class="p">}</span>

        <span class="c1"># Iterating over completed futures</span>
        <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">concurrent</span><span class="o">.</span><span class="n">futures</span><span class="o">.</span><span class="n">as_completed</span><span class="p">(</span><span class="n">future_to_aid</span><span class="p">):</span>
            <span class="n">aid</span> <span class="o">=</span> <span class="n">future_to_aid</span><span class="p">[</span><span class="n">future</span><span class="p">]</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">assay_data</span> <span class="o">=</span> <span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span>  <span class="c1"># Fetching the result from the future</span>
                <span class="k">if</span> <span class="n">assay_data</span><span class="p">:</span>
                    <span class="c1"># Preparing a new row with the fetched data</span>
                    <span class="n">new_row</span> <span class="o">=</span> <span class="p">{</span>
                        <span class="s1">&#39;AID&#39;</span><span class="p">:</span> <span class="n">aid</span><span class="p">,</span>
                        <span class="s1">&#39;Assay Type&#39;</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">aid</span><span class="p">,</span> <span class="s1">&#39;Assay Type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                        <span class="s1">&#39;Activity Name&#39;</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">aid</span><span class="p">,</span> <span class="s1">&#39;Activity Name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                        <span class="o">**</span><span class="n">assay_data</span>
                    <span class="p">}</span>
                    <span class="c1"># Adding the new row to the DataFrame</span>
                    <span class="n">assay_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">assay_df</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">new_row</span><span class="p">])],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
                <span class="c1"># Logging any errors encountered during the fetch</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error occurred while processing AID </span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">exc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Saving the updated DataFrame to a CSV file</span>
    <span class="n">assay_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Data/Nodes/Assay_Properties.csv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">assay_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_compound_properties" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_compound_properties</span><span class="p">(</span><span class="n">main_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Extracts and aggregates compound properties from PubChem for a list of
compounds associated with specific genes.</p>
<p>This method processes a CSV file specified by <code>main_data</code>, which contains
gene identifiers and their associated compound IDs (CIDs). It selects
compounds related to the top <code>n</code> most frequently occurring genes in the
dataset, where <code>n</code> is determined by the instance's <code>_enzyme_count</code>
attribute. The method then fetches detailed compound properties from
PubChem in chunks, using concurrent requests to improve efficiency and
manage the load on the PubChem API. The fetched compound properties are
aggregated into a single DataFrame and saved to multiple CSV files,
one for each chunk of compound IDs processed.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to a CSV file containing main data was which</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<details class="side-effects" open>
  <summary>Side Effects</summary>
  <ul>
<li>Saves the aggregated compound properties to CSV files in the current
working directory. The files are named
'Data/Nodes/Compound_Properties/Chunk_{i}.csv', where <code>{i}</code> is
the chunk index.</li>
</ul>
</details>

    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>None</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>This method does not return a value. Instead, it saves the</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>fetched compound data directly to CSV files.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>Exception</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Logs an error and continues processing the next CID if</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = NodePropertiesExtractor(['CYP2D6', 'CYP3A4'])
extractor.create_data_directories()
extractor.extract_compound_properties('Data/AllDataConnected.csv')
This will read 'Data/AllDataConnected.csv', filter for compounds
associated with the top n genes, fetch their properties from PubChem,
and save the results into multiple CSV files for each chunk
of compounds processed.</p>
</blockquote>
</blockquote>
</blockquote>
</details>

<details class="note" open>
  <summary>Note</summary>
  <ul>
<li>Ensure that the 'main_data' CSV file exists and is accessible at
the specified path.</li>
<li>The method automatically handles NaN values in the 'CID' column
and excludes them from processing.</li>
<li>The <code>enzyme_count</code> attribute determines the number of top genes
for which compound properties will be fetched.</li>
<li>Internet access is required to fetch compound data from the PubChem API.</li>
<li>The method employs a <code>ThreadPoolExecutor</code> with a configurable
number of workers (default is len(enzyme_list)) to parallelize
requests, which can be adjusted based on system capabilities and
API rate limits.</li>
</ul>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_compound_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts and aggregates compound properties from PubChem for a list of</span>
<span class="sd">    compounds associated with specific genes.</span>

<span class="sd">    This method processes a CSV file specified by `main_data`, which contains</span>
<span class="sd">    gene identifiers and their associated compound IDs (CIDs). It selects</span>
<span class="sd">    compounds related to the top `n` most frequently occurring genes in the</span>
<span class="sd">    dataset, where `n` is determined by the instance&#39;s `_enzyme_count`</span>
<span class="sd">    attribute. The method then fetches detailed compound properties from</span>
<span class="sd">    PubChem in chunks, using concurrent requests to improve efficiency and</span>
<span class="sd">    manage the load on the PubChem API. The fetched compound properties are</span>
<span class="sd">    aggregated into a single DataFrame and saved to multiple CSV files,</span>
<span class="sd">    one for each chunk of compound IDs processed.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to a CSV file containing main data was which</span>
<span class="sd">        generated after running `extractor.run()`.</span>

<span class="sd">    Side Effects:</span>
<span class="sd">        - Saves the aggregated compound properties to CSV files in the current</span>
<span class="sd">        working directory. The files are named</span>
<span class="sd">        &#39;Data/Nodes/Compound_Properties/Chunk_{i}.csv&#39;, where `{i}` is</span>
<span class="sd">        the chunk index.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: This method does not return a value. Instead, it saves the</span>
<span class="sd">        fetched compound data directly to CSV files.</span>

<span class="sd">    Raises:</span>
<span class="sd">        Exception: Logs an error and continues processing the next CID if</span>
<span class="sd">        an error occurs while fetching data for a specific CID.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">        &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">        &gt;&gt;&gt; extractor.extract_compound_properties(&#39;Data/AllDataConnected.csv&#39;)</span>
<span class="sd">        This will read &#39;Data/AllDataConnected.csv&#39;, filter for compounds</span>
<span class="sd">        associated with the top n genes, fetch their properties from PubChem,</span>
<span class="sd">        and save the results into multiple CSV files for each chunk</span>
<span class="sd">        of compounds processed.</span>

<span class="sd">    Note:</span>
<span class="sd">        - Ensure that the &#39;main_data&#39; CSV file exists and is accessible at</span>
<span class="sd">        the specified path.</span>
<span class="sd">        - The method automatically handles NaN values in the &#39;CID&#39; column</span>
<span class="sd">        and excludes them from processing.</span>
<span class="sd">        - The `enzyme_count` attribute determines the number of top genes</span>
<span class="sd">        for which compound properties will be fetched.</span>
<span class="sd">        - Internet access is required to fetch compound data from the PubChem API.</span>
<span class="sd">        - The method employs a `ThreadPoolExecutor` with a configurable</span>
<span class="sd">        number of workers (default is len(enzyme_list)) to parallelize</span>
<span class="sd">        requests, which can be adjusted based on system capabilities and</span>
<span class="sd">        API rate limits.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_enzyme_count</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">])]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">])</span>
    <span class="n">IDs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="c1"># Define chunk size and calculate number of chunks</span>
    <span class="n">chunk_size</span> <span class="o">=</span> <span class="mi">10000</span>
    <span class="n">num_chunks</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">IDs</span><span class="p">)</span> <span class="o">/</span> <span class="n">chunk_size</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">start_chunk</span> <span class="o">&lt;</span> <span class="n">num_chunks</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start_chunk</span><span class="p">,</span> <span class="n">num_chunks</span><span class="p">):</span>
            <span class="c1"># Calculate start and end indices for each chunk</span>
            <span class="n">start_index</span> <span class="o">=</span> <span class="n">i</span> <span class="o">*</span> <span class="n">chunk_size</span>
            <span class="n">end_index</span> <span class="o">=</span> <span class="n">start_index</span> <span class="o">+</span> <span class="n">chunk_size</span>

            <span class="c1"># Extract chunk of CIDs</span>
            <span class="n">chunk_cids</span> <span class="o">=</span> <span class="n">IDs</span><span class="p">[</span><span class="n">start_index</span><span class="p">:</span><span class="n">end_index</span><span class="p">]</span>
            <span class="c1"># chunk_cids = [x for x in chunk_cids if not np.isnan(x)]</span>

            <span class="c1"># Use ThreadPoolExecutor to parallelize requests for the chunk</span>
            <span class="k">with</span> <span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
                <span class="n">future_to_cid</span> <span class="o">=</span> <span class="p">{</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fetch_data</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span> <span class="n">cid</span> <span class="k">for</span> <span class="n">cid</span> <span class="ow">in</span> <span class="n">chunk_cids</span><span class="p">}</span>
                <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>

                <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">as_completed</span><span class="p">(</span><span class="n">future_to_cid</span><span class="p">):</span>
                    <span class="n">cid</span> <span class="o">=</span> <span class="n">future_to_cid</span><span class="p">[</span><span class="n">future</span><span class="p">]</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">data</span> <span class="o">=</span> <span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span>
                        <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing CID </span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Concatenate results for the current chunk</span>
            <span class="n">chunk_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="c1"># Save the concatenated DataFrame to a CSV file for the chunk</span>
            <span class="n">chunk_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Nodes/Compound_Properties/Chunk_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span>
                            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No more chunks to process.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_gene_properties" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_gene_properties</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Extracts and processes gene properties from a given data source,
specifically targeting genes relevant to the study (e.g., CYP enzymes)
and records their details in a structured DataFrame.</p>
<p>This method reads gene data from a CSV file specified by <code>main_data</code>,
queries the PubChem database for additional properties of each unique
gene ID found in the file, and compiles these properties into a new
DataFrame. It focuses on fetching details like gene symbols, taxonomy,
taxonomy IDs, and synonyms for each gene. The final DataFrame is filtered
to include only genes of particular interest (e.g., certain CYP enzymes)
and saved to a separate CSV file for further analysis or use.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to a CSV file containing main data was which</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the compiled gene properties,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>including GeneID, Symbol, Taxonomy, Taxonomy ID, and Synonyms,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>filtered to include only specified genes of interest. This DataFrame</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>is also saved to 'Data/Nodes/Gene_Properties.csv'.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>Exception</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If there's an issue reading the initial CSV file or</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = NodePropertiesExtractor(['CYP2D6', 'CYP3A4'])
extractor.create_data_directories()
extractor.run()
gene_properties_df = extractor.extract_gene_properties('Data/AllDataConnected.csv')
print(gene_properties_df.head())</p>
</blockquote>
</blockquote>
</blockquote>
<p>This would read gene IDs from 'Data/AllDataConnected.csv', fetch
their properties from PubChem, and compile the details into a
DataFrame, filtering for specified genes of interest and saving
the results to 'Data/Nodes/Gene_Properties.csv'.</p>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>The method filters the resulting DataFrame to include only genes with
symbols in the predefined enzyme_list. Adjust this list as necessary
to match the focus of your study or application.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_gene_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts and processes gene properties from a given data source,</span>
<span class="sd">    specifically targeting genes relevant to the study (e.g., CYP enzymes)</span>
<span class="sd">    and records their details in a structured DataFrame.</span>

<span class="sd">    This method reads gene data from a CSV file specified by `main_data`,</span>
<span class="sd">    queries the PubChem database for additional properties of each unique</span>
<span class="sd">    gene ID found in the file, and compiles these properties into a new</span>
<span class="sd">    DataFrame. It focuses on fetching details like gene symbols, taxonomy,</span>
<span class="sd">    taxonomy IDs, and synonyms for each gene. The final DataFrame is filtered</span>
<span class="sd">    to include only genes of particular interest (e.g., certain CYP enzymes)</span>
<span class="sd">    and saved to a separate CSV file for further analysis or use.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to a CSV file containing main data was which</span>
<span class="sd">        generated after running `extractor.run()`.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the compiled gene properties,</span>
<span class="sd">        including GeneID, Symbol, Taxonomy, Taxonomy ID, and Synonyms,</span>
<span class="sd">        filtered to include only specified genes of interest. This DataFrame</span>
<span class="sd">        is also saved to &#39;Data/Nodes/Gene_Properties.csv&#39;.</span>

<span class="sd">    Raises:</span>
<span class="sd">        Exception: If there&#39;s an issue reading the initial CSV file or</span>
<span class="sd">        fetching gene details from PubChem, details of the exception are</span>
<span class="sd">        logged, and the method proceeds to process the next gene ID.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">        &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">        &gt;&gt;&gt; extractor.run()</span>
<span class="sd">        &gt;&gt;&gt; gene_properties_df = extractor.extract_gene_properties(&#39;Data/AllDataConnected.csv&#39;)</span>
<span class="sd">        &gt;&gt;&gt; print(gene_properties_df.head())</span>

<span class="sd">        This would read gene IDs from &#39;Data/AllDataConnected.csv&#39;, fetch</span>
<span class="sd">        their properties from PubChem, and compile the details into a</span>
<span class="sd">        DataFrame, filtering for specified genes of interest and saving</span>
<span class="sd">        the results to &#39;Data/Nodes/Gene_Properties.csv&#39;.</span>

<span class="sd">    Note:</span>
<span class="sd">        The method filters the resulting DataFrame to include only genes with</span>
<span class="sd">        symbols in the predefined enzyme_list. Adjust this list as necessary</span>
<span class="sd">        to match the focus of your study or application.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
    <span class="n">df_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Symbol&#39;</span><span class="p">,</span> <span class="s1">&#39;Taxonomy&#39;</span><span class="p">,</span>
                                    <span class="s1">&#39;Taxonomy ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Synonyms&#39;</span><span class="p">])</span>

    <span class="n">unique_gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="n">gene_details</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">unique_gene_ids</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">gene_id</span><span class="p">,</span> <span class="n">symbol</span><span class="p">,</span> <span class="n">taxonomy</span><span class="p">,</span> <span class="n">taxonomy_id</span><span class="p">,</span> <span class="n">synonyms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fetch_gene_details</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>
            <span class="n">gene_details</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;GeneID&#39;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">,</span>
                <span class="s1">&#39;Symbol&#39;</span><span class="p">:</span> <span class="n">symbol</span><span class="p">,</span>
                <span class="s1">&#39;Taxonomy&#39;</span><span class="p">:</span> <span class="n">taxonomy</span><span class="p">,</span>
                <span class="s1">&#39;Taxonomy ID&#39;</span><span class="p">:</span> <span class="n">taxonomy_id</span><span class="p">,</span>
                <span class="s1">&#39;Synonyms&#39;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">synonyms</span><span class="p">)</span>
            <span class="p">})</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Error occurred while processing gene_id </span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                          <span class="n">gene_id</span><span class="p">,</span> <span class="n">exc</span><span class="p">)</span>
            <span class="n">gene_details</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;GeneID&#39;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">,</span>
                <span class="s1">&#39;Symbol&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                <span class="s1">&#39;Taxonomy&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                <span class="s1">&#39;Taxonomy ID&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                <span class="s1">&#39;Synonyms&#39;</span><span class="p">:</span> <span class="kc">None</span>
            <span class="p">})</span>

    <span class="c1"># Now create the DataFrame from the list of dictionaries</span>
    <span class="n">df_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_details</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_enzyme_count</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">df_gene</span> <span class="o">=</span> <span class="n">df_gene</span><span class="p">[</span><span class="n">df_gene</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">])]</span>
    <span class="n">df_gene</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Data/Nodes/Gene_Properties.csv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df_gene</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.extract_protein_properties" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_protein_properties</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Extracts and compiles protein properties from the NCBI protein database
based on accession numbers.</p>
<p>Given a CSV file specified by <code>main_data</code>, this method reads protein
accession numbers and performs web scraping on the NCBI protein database
pages to extract protein titles. The method constructs a URL for
each accession number, sends a request to retrieve the page content,
and parses the HTML to find the protein title. The extracted titles,
along with their corresponding accession numbers and URLs, are
compiled into a DataFrame. This DataFrame is saved to a CSV file,
providing a structured summary of protein properties for further analysis or use.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to a CSV file containing main data was which</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame with columns 'RefSeq Accession', 'URL',</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>and 'Description', where 'Description' contains the title of the</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>protein extracted from its NCBI page. This DataFrame is saved to</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>'Data/Nodes/Protein_Properties.csv' in the current working directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>Exception</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If there's an issue reading the initial CSV file or</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <p>Assuming 'protein_data.csv' contains a column 'Target Accession'
with accession numbers:</p>
<blockquote>
<blockquote>
<blockquote>
<p>extractor = NodePropertiesExtractor(['CYP2D6', 'CYP3A4'])
extractor.create_data_directories()
extractor.run() # you need to run this only once
protein_properties_df = extractor.extract_protein_properties('Data/AllDataConnected.csv')
print(protein_properties_df.head())</p>
</blockquote>
</blockquote>
</blockquote>
<p>This would read accession numbers from 'Data/AllDataConnected.csv',
scrape their titles from the NCBI protein database, and compile the
results into a DataFrame, which is also saved to
'Data/Nodes/Protein_Properties.csv'.</p>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>This method requires internet access to query the NCBI protein
database. Ensure the input file path is correct and accessible to
avoid errors during processing. Web scraping is dependent on the
structure of the web page; changes to the NCBI protein database
pages may require updates to the scraping logic.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_protein_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts and compiles protein properties from the NCBI protein database</span>
<span class="sd">    based on accession numbers.</span>

<span class="sd">    Given a CSV file specified by `main_data`, this method reads protein</span>
<span class="sd">    accession numbers and performs web scraping on the NCBI protein database</span>
<span class="sd">    pages to extract protein titles. The method constructs a URL for</span>
<span class="sd">    each accession number, sends a request to retrieve the page content,</span>
<span class="sd">    and parses the HTML to find the protein title. The extracted titles,</span>
<span class="sd">    along with their corresponding accession numbers and URLs, are</span>
<span class="sd">    compiled into a DataFrame. This DataFrame is saved to a CSV file,</span>
<span class="sd">    providing a structured summary of protein properties for further analysis or use.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to a CSV file containing main data was which</span>
<span class="sd">        generated after running `extractor.run()`.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame with columns &#39;RefSeq Accession&#39;, &#39;URL&#39;,</span>
<span class="sd">        and &#39;Description&#39;, where &#39;Description&#39; contains the title of the</span>
<span class="sd">        protein extracted from its NCBI page. This DataFrame is saved to</span>
<span class="sd">        &#39;Data/Nodes/Protein_Properties.csv&#39; in the current working directory.</span>

<span class="sd">    Raises:</span>
<span class="sd">        Exception: If there&#39;s an issue reading the initial CSV file or</span>
<span class="sd">        querying the NCBI database, details of the exception are logged.</span>
<span class="sd">        The method continues processing the next accession number,</span>
<span class="sd">        ensuring robustness against individual failures.</span>

<span class="sd">    Example:</span>
<span class="sd">        Assuming &#39;protein_data.csv&#39; contains a column &#39;Target Accession&#39;</span>
<span class="sd">        with accession numbers:</span>

<span class="sd">        &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">        &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">        &gt;&gt;&gt; extractor.run() # you need to run this only once</span>
<span class="sd">        &gt;&gt;&gt; protein_properties_df = extractor.extract_protein_properties(&#39;Data/AllDataConnected.csv&#39;)</span>
<span class="sd">        &gt;&gt;&gt; print(protein_properties_df.head())</span>

<span class="sd">        This would read accession numbers from &#39;Data/AllDataConnected.csv&#39;,</span>
<span class="sd">        scrape their titles from the NCBI protein database, and compile the</span>
<span class="sd">        results into a DataFrame, which is also saved to</span>
<span class="sd">        &#39;Data/Nodes/Protein_Properties.csv&#39;.</span>

<span class="sd">    Note:</span>
<span class="sd">        This method requires internet access to query the NCBI protein</span>
<span class="sd">        database. Ensure the input file path is correct and accessible to</span>
<span class="sd">        avoid errors during processing. Web scraping is dependent on the</span>
<span class="sd">        structure of the web page; changes to the NCBI protein database</span>
<span class="sd">        pages may require updates to the scraping logic.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Initialize a list to store the extracted data</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_enzyme_count</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">])]</span>
    <span class="n">Accessions</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target Accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="c1"># Iterate over each protein accession number in the DataFrame</span>
    <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">Accessions</span><span class="p">:</span>
        <span class="c1"># Construct the URL to query the NCBI protein database</span>
        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ncbi.nlm.nih.gov/protein/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Send an HTTP request to the URL</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>

            <span class="c1"># Parse the HTML content of the response</span>
            <span class="n">soup</span> <span class="o">=</span> <span class="n">BeautifulSoup</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">,</span> <span class="s1">&#39;html.parser&#39;</span><span class="p">)</span>

            <span class="c1"># Extract the title from the parsed HTML</span>
            <span class="n">title</span> <span class="o">=</span> <span class="n">soup</span><span class="o">.</span><span class="n">title</span><span class="o">.</span><span class="n">string</span> <span class="k">if</span> <span class="n">soup</span><span class="o">.</span><span class="n">title</span> <span class="k">else</span> <span class="s1">&#39;Title Not Found&#39;</span>

            <span class="c1"># Append the extracted data to the list</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;RefSeq Accession&#39;</span><span class="p">:</span> <span class="n">accession</span><span class="p">,</span>
                         <span class="s1">&#39;URL&#39;</span><span class="p">:</span> <span class="n">url</span><span class="p">,</span> <span class="s1">&#39;Description&#39;</span><span class="p">:</span> <span class="n">title</span><span class="p">})</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="c1"># In case of an error, log the error message</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error fetching data for accession </span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;RefSeq Accession&#39;</span><span class="p">:</span> <span class="n">accession</span><span class="p">,</span> <span class="s1">&#39;URL&#39;</span><span class="p">:</span> <span class="n">url</span><span class="p">,</span>
                         <span class="s1">&#39;Description&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;Error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">})</span>

    <span class="c1"># Convert the list of data into a DataFrame</span>
    <span class="n">protein_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

    <span class="c1"># Save the DataFrame to a CSV file</span>
    <span class="n">protein_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Data/Nodes/Protein_Properties.csv&#39;</span><span class="p">,</span>
                      <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Return the DataFrame</span>
    <span class="k">return</span> <span class="n">protein_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.fetch_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">fetch_data</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Retrieves detailed chemical compound properties for a specified
Compound ID (CID) from the PubChem database.</p>
<p>This method constructs a query URL to fetch a wide range of properties
for the given CID from PubChem, including molecular formula,
molecular weight, canonical and isomeric SMILES, InChI codes,
physicochemical properties, and more. If the CID is valid and data is
available, it returns a pandas DataFrame containing these properties. This
method also generates a URL to retrieve the structure image of the
compound as a 2D PNG image, adding it as a column in the DataFrame.
In cases where the CID is NaN or an error occurs during data retrieval,
an empty DataFrame is returned.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>cid</code></td>
            <td>
                  <code>int or float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The Compound ID for which to fetch data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the fetched properties for the</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>given CID. The DataFrame includes columns for each property fetched</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>from PubChem, along with a 'StructureImage2DURL' column containing</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the URL to the compound's structure image. Returns an empty DataFrame</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>if the CID is NaN or if any error occurs during the fetch operation.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>Exception</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Logs an error message if the request to PubChem fails or</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = NodePropertiesExtractor(['CYP2D6', 'CYP3A4'])
extractor.create_data_directories()
compound_data_df = extractor.fetch_data(2244)
print(compound_data_df.head())</p>
</blockquote>
</blockquote>
</blockquote>
<p>This example fetches the properties for the compound with CID 2244
from PubChem and prints the first few rows
of the resulting DataFrame.</p>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>This method requires an active internet connection to access the
PubChem database. Ensure that the CID provided is valid and not NaN
to avoid fetching errors. The structure and availability of data
fields are subject to the current state of the PubChem database
and may vary.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fetch_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves detailed chemical compound properties for a specified</span>
<span class="sd">    Compound ID (CID) from the PubChem database.</span>

<span class="sd">    This method constructs a query URL to fetch a wide range of properties</span>
<span class="sd">    for the given CID from PubChem, including molecular formula,</span>
<span class="sd">    molecular weight, canonical and isomeric SMILES, InChI codes,</span>
<span class="sd">    physicochemical properties, and more. If the CID is valid and data is</span>
<span class="sd">    available, it returns a pandas DataFrame containing these properties. This</span>
<span class="sd">    method also generates a URL to retrieve the structure image of the</span>
<span class="sd">    compound as a 2D PNG image, adding it as a column in the DataFrame.</span>
<span class="sd">    In cases where the CID is NaN or an error occurs during data retrieval,</span>
<span class="sd">    an empty DataFrame is returned.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        cid (int or float): The Compound ID for which to fetch data.</span>
<span class="sd">        Can be an integer or NaN.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the fetched properties for the</span>
<span class="sd">        given CID. The DataFrame includes columns for each property fetched</span>
<span class="sd">        from PubChem, along with a &#39;StructureImage2DURL&#39; column containing</span>
<span class="sd">        the URL to the compound&#39;s structure image. Returns an empty DataFrame</span>
<span class="sd">        if the CID is NaN or if any error occurs during the fetch operation.</span>

<span class="sd">    Raises:</span>
<span class="sd">        Exception: Logs an error message if the request to PubChem fails or</span>
<span class="sd">        if the response cannot be processed into a DataFrame.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">        &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">        &gt;&gt;&gt; compound_data_df = extractor.fetch_data(2244)</span>
<span class="sd">        &gt;&gt;&gt; print(compound_data_df.head())</span>

<span class="sd">        This example fetches the properties for the compound with CID 2244</span>
<span class="sd">        from PubChem and prints the first few rows</span>
<span class="sd">        of the resulting DataFrame.</span>

<span class="sd">    Note:</span>
<span class="sd">        This method requires an active internet connection to access the</span>
<span class="sd">        PubChem database. Ensure that the CID provided is valid and not NaN</span>
<span class="sd">        to avoid fetching errors. The structure and availability of data</span>
<span class="sd">        fields are subject to the current state of the PubChem database</span>
<span class="sd">        and may vary.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">cid</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># Return an empty DataFrame for NaN CIDs</span>

    <span class="n">cid</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span>  <span class="c1"># Convert CID to integer</span>
    <span class="n">url</span> <span class="o">=</span> <span class="p">(</span><span class="sa">f</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">/property/&quot;</span>
           <span class="s2">&quot;MolecularFormula,MolecularWeight,CanonicalSMILES,IsomericSMILES,InChI,&quot;</span>
           <span class="s2">&quot;InChIKey,IUPACName,Title,XLogP,ExactMass,MonoisotopicMass,TPSA,Complexity,&quot;</span>
           <span class="s2">&quot;Charge,HBondDonorCount,HBondAcceptorCount,RotatableBondCount,HeavyAtomCount,&quot;</span>
           <span class="s2">&quot;IsotopeAtomCount,AtomStereoCount,DefinedAtomStereoCount,UndefinedAtomStereoCount,&quot;</span>
           <span class="s2">&quot;BondStereoCount,DefinedBondStereoCount,UndefinedBondStereoCount,CovalentUnitCount,&quot;</span>
           <span class="s2">&quot;PatentCount,PatentFamilyCount,LiteratureCount,Volume3D,XStericQuadrupole3D,&quot;</span>
           <span class="s2">&quot;YStericQuadrupole3D,ZStericQuadrupole3D,FeatureCount3D,FeatureAcceptorCount3D,&quot;</span>
           <span class="s2">&quot;FeatureDonorCount3D,FeatureAnionCount3D,FeatureCationCount3D,FeatureRingCount3D,&quot;</span>
           <span class="s2">&quot;FeatureHydrophobeCount3D,ConformerModelRMSD3D,EffectiveRotorCount3D,ConformerCount3D,&quot;</span>
           <span class="s2">&quot;Fingerprint2D/CSV&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
        <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
        <span class="n">compound_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">),</span>
                                    <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">compound_data</span><span class="p">[</span><span class="s1">&#39;StructureImage2DURL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">/PNG&quot;</span>
        <span class="k">return</span> <span class="n">compound_data</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing CID </span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># Return an empty DataFrame in case of error</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.get_enzyme_assays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_enzyme_assays</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Fetches assay data for a specified enzyme from the PubChem database and
returns it as a pandas DataFrame.</p>
<p>This method constructs a URL to query the PubChem database for concise
assay data related to the given enzyme. It processes the CSV response
into a DataFrame, which includes various assay data points provided by PubChem.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>enzyme</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the enzyme for which assay data is</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the assay data fetched from</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>PubChem for the specified enzyme. The DataFrame includes columns</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>based on the CSV response from PubChem, such as assay ID, results,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>and conditions. Returns None if no data is available or if an error</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>occurs during data fetching or processing.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="requests.RequestException">RequestException</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If an error occurs during the HTTP</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.errors.EmptyDataError">EmptyDataError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the response from PubChem contains no data.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = NodePropertiesExtractor(['enzyme'])
enzyme_assays_df = extractor.get_enzyme_assays('enzyme')
print(enzyme_assays_df.head())</p>
</blockquote>
</blockquote>
</blockquote>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_enzyme_assays</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches assay data for a specified enzyme from the PubChem database and</span>
<span class="sd">    returns it as a pandas DataFrame.</span>

<span class="sd">    This method constructs a URL to query the PubChem database for concise</span>
<span class="sd">    assay data related to the given enzyme. It processes the CSV response</span>
<span class="sd">    into a DataFrame, which includes various assay data points provided by PubChem.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        enzyme (str): The name of the enzyme for which assay data is</span>
<span class="sd">        requested. This name is used in the API query.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the assay data fetched from</span>
<span class="sd">        PubChem for the specified enzyme. The DataFrame includes columns</span>
<span class="sd">        based on the CSV response from PubChem, such as assay ID, results,</span>
<span class="sd">        and conditions. Returns None if no data is available or if an error</span>
<span class="sd">        occurs during data fetching or processing.</span>

<span class="sd">    Raises:</span>
<span class="sd">        requests.RequestException: If an error occurs during the HTTP</span>
<span class="sd">        request to the PubChem API.</span>
<span class="sd">        pd.errors.EmptyDataError: If the response from PubChem contains no data.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;enzyme&#39;])</span>
<span class="sd">        &gt;&gt;&gt; enzyme_assays_df = extractor.get_enzyme_assays(&#39;enzyme&#39;)</span>
<span class="sd">        &gt;&gt;&gt; print(enzyme_assays_df.head())</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">assays_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_base_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">enzyme</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="si">}</span><span class="s2">/concise/CSV&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fetching assays for enzyme: </span><span class="si">{</span><span class="n">enzyme</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_make_request</span><span class="p">(</span><span class="n">assays_url</span><span class="p">)</span>

    <span class="n">assays_csv_string</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">text</span>
    <span class="n">assays_csv_string_io</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">assays_csv_string</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">assays_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">assays_csv_string_io</span><span class="p">,</span>
                                <span class="n">sep</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_sep</span><span class="p">,</span>
                                <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Assays DataFrame for enzyme </span><span class="si">%s</span><span class="s2"> has shape: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                     <span class="n">enzyme</span><span class="p">,</span> <span class="n">assays_df</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">assays_df</span>
    <span class="k">except</span> <span class="n">pd</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">EmptyDataError</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No data available for enzyme </span><span class="si">%s</span><span class="s2">.&quot;</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_properties_extractor.NodePropertiesExtractor.run" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Orchestrates the process of fetching, filtering, and aggregating assay
data from PubChem for a predefined list of enzymes.</p>
<p>This method iteratively queries PubChem for assay data corresponding
to each enzyme specified in the <code>enzyme_list</code> attribute during class
initialization. It performs the following steps for each enzyme:
1. Constructs a query URL and fetches assay data from PubChem.
2. Filters the fetched data based on predefined criteria
(e.g., containing specific substrings in the assay name).
3. Aggregates the filtered data into a single pandas DataFrame.
4. Identifies enzymes for which data could not be fetched or were
excluded based on filtering criteria, logging their names.</p>
<p>The final aggregated DataFrame, containing assay data for all successfully
processed enzymes, is then saved to a CSV file. This method facilitates
the extraction and preprocessing of chemical assay data for further
analysis or integration into knowledge graphs.</p>


<details class="note" open>
  <summary>Note</summary>
  <ul>
<li>This method relies on the successful response from PubChem
for each enzyme query.</li>
<li>Enzymes with no available data or failing to meet the filtering
criteria are excluded from the final DataFrame.</li>
<li>The output CSV file is saved in the current working directory
with the name 'Data/AllDataConnected.csv'.</li>
</ul>
</details>

    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the aggregated and filtered</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>assay data for the specified enzymes. Columns in the DataFrame</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>correspond to the assay data fields returned by PubChem, subject to</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the filtering criteria applied within this method.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="requests.RequestException">RequestException</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If there is an issue with fetching data</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <p>Assuming <code>enzyme_list</code> was set to ['CYP2D6', 'CYP3A4'] during
class initialization:</p>
<blockquote>
<blockquote>
<blockquote>
<p>extractor = NodePropertiesExtractor(['CYP2D6', 'CYP3A4'])
extractor.create_data_directories()
result_df = extractor.run()
print(result_df.head())</p>
</blockquote>
</blockquote>
</blockquote>
<p>This will fetch and process assay data for 'CYP2D6' and 'CYP3A4',
returning a DataFrame with the processed data.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Orchestrates the process of fetching, filtering, and aggregating assay</span>
<span class="sd">    data from PubChem for a predefined list of enzymes.</span>

<span class="sd">    This method iteratively queries PubChem for assay data corresponding</span>
<span class="sd">    to each enzyme specified in the `enzyme_list` attribute during class</span>
<span class="sd">    initialization. It performs the following steps for each enzyme:</span>
<span class="sd">    1. Constructs a query URL and fetches assay data from PubChem.</span>
<span class="sd">    2. Filters the fetched data based on predefined criteria</span>
<span class="sd">    (e.g., containing specific substrings in the assay name).</span>
<span class="sd">    3. Aggregates the filtered data into a single pandas DataFrame.</span>
<span class="sd">    4. Identifies enzymes for which data could not be fetched or were</span>
<span class="sd">    excluded based on filtering criteria, logging their names.</span>

<span class="sd">    The final aggregated DataFrame, containing assay data for all successfully</span>
<span class="sd">    processed enzymes, is then saved to a CSV file. This method facilitates</span>
<span class="sd">    the extraction and preprocessing of chemical assay data for further</span>
<span class="sd">    analysis or integration into knowledge graphs.</span>

<span class="sd">    Note:</span>
<span class="sd">        - This method relies on the successful response from PubChem</span>
<span class="sd">        for each enzyme query.</span>
<span class="sd">        - Enzymes with no available data or failing to meet the filtering</span>
<span class="sd">        criteria are excluded from the final DataFrame.</span>
<span class="sd">        - The output CSV file is saved in the current working directory</span>
<span class="sd">        with the name &#39;Data/AllDataConnected.csv&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the aggregated and filtered</span>
<span class="sd">        assay data for the specified enzymes. Columns in the DataFrame</span>
<span class="sd">        correspond to the assay data fields returned by PubChem, subject to</span>
<span class="sd">        the filtering criteria applied within this method.</span>

<span class="sd">    Raises:</span>
<span class="sd">        requests.RequestException: If there is an issue with fetching data</span>
<span class="sd">        from PubChem, such as a network problem or an invalid response.</span>

<span class="sd">    Example:</span>
<span class="sd">        Assuming `enzyme_list` was set to [&#39;CYP2D6&#39;, &#39;CYP3A4&#39;] during</span>
<span class="sd">        class initialization:</span>

<span class="sd">        &gt;&gt;&gt; extractor = NodePropertiesExtractor([&#39;CYP2D6&#39;, &#39;CYP3A4&#39;])</span>
<span class="sd">        &gt;&gt;&gt; extractor.create_data_directories()</span>
<span class="sd">        &gt;&gt;&gt; result_df = extractor.run()</span>
<span class="sd">        &gt;&gt;&gt; print(result_df.head())</span>

<span class="sd">        This will fetch and process assay data for &#39;CYP2D6&#39; and &#39;CYP3A4&#39;,</span>
<span class="sd">        returning a DataFrame with the processed data.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Initialize an empty list to store enzymes with successful responses</span>
    <span class="n">enzymes_with_response</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Keep a copy of the original list to identify removed enzymes later</span>
    <span class="n">original_enzyme_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">enzyme</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span><span class="p">:</span>
        <span class="c1"># Formulate the URL</span>
        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_base_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">enzyme</span><span class="si">}</span><span class="s2">/concise/CSV&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
            <span class="c1"># Check for a successful response (status code 200)</span>
            <span class="k">if</span> <span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
                <span class="n">enzymes_with_response</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span>  <span class="c1"># Keep the enzyme in the new list</span>
        <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span><span class="p">:</span>
            <span class="c1"># If there&#39;s an exception, skip adding the enzyme to the new list</span>
            <span class="k">pass</span>

    <span class="c1"># Update the enzyme list with only the enzymes that had a successful response</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span> <span class="o">=</span> <span class="n">enzymes_with_response</span>

    <span class="c1"># Identify and print the removed enzymes</span>
    <span class="n">removed_enzymes</span> <span class="o">=</span> <span class="p">[</span><span class="n">enzyme</span> <span class="k">for</span> <span class="n">enzyme</span> <span class="ow">in</span> <span class="n">original_enzyme_list</span> <span class="k">if</span> <span class="n">enzyme</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">enzymes_with_response</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">removed_enzymes</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;These enzymes were removed because their names aren&#39;t correct: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">removed_enzymes</span><span class="p">))</span>

    <span class="n">df_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process_enzymes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">enzyme_list</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_concatenate_data</span><span class="p">(</span><span class="n">df_list</span><span class="p">)</span>
    <span class="n">substrings_to_filter</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CYP&#39;</span><span class="p">,</span> <span class="s1">&#39;Cytochrome&#39;</span><span class="p">]</span>
    <span class="n">pattern</span> <span class="o">=</span> <span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">substrings_to_filter</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Assay Name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">pattern</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">)]</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Data/AllDataConnected.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h2 id="4-node-data-processor">4. Node Data Processor</h2>


<div class="doc doc-object doc-module">



<a id="chemgraphbuilder.node_data_processor"></a>
    <div class="doc doc-contents first">

      <p>node_data_processor.py</p>
<p>This module provides the NodeDataProcessor class, which is responsible for
preprocessing various types of node data (assays, proteins, genes, and compounds)
for use in chemical knowledge graph construction. The preprocessing includes
renaming columns, consolidating multiple files, and saving the processed data
in a consistent format. This step ensures uniformity and ease of access for
subsequent data analysis and integration processes.</p>


<p><span class="doc-section-title">Classes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><a class="autorefs autorefs-internal" title="chemgraphbuilder.node_data_processor.NodeDataProcessor" href="#chemgraphbuilder.node_data_processor.NodeDataProcessor">NodeDataProcessor</a></code></td>
            <td>
              <div class="doc-md-description">
                <p>Handles preprocessing of assay, protein, gene, and compound data.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example-usage" open>
  <summary>Example Usage</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>processor = NodeDataProcessor(data_dir='path/to/data')
processor.preprocess_assays()
processor.preprocess_proteins()
processor.preprocess_genes()
processor.preprocess_compounds()</p>
</blockquote>
</blockquote>
</blockquote>
</details>


  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.node_data_processor.NodeDataProcessor" class="doc doc-heading">
            <code>NodeDataProcessor</code>


</h2>


    <div class="doc doc-contents ">


      <p>NodeDataProcessor is responsible for preprocessing various types of node data
(assays, proteins, genes, and compounds) by renaming columns, consolidating
multiple files, and saving the processed data. This preprocessing step is
crucial for ensuring uniformity and ease of access in subsequent analysis
and integration processes.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.node_data_processor.NodeDataProcessor.data_dir">data_dir</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the node data files are stored.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Methods:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_assays" href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_assays">preprocess_assays</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Processes and renames columns in assay data.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_proteins" href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_proteins">preprocess_proteins</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Processes and renames columns in protein data.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_genes" href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_genes">preprocess_genes</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Processes and renames columns in gene data.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_compounds" href="#chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_compounds">preprocess_compounds</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Consolidates and renames columns in compound data.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>

              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/node_data_processor.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">NodeDataProcessor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    NodeDataProcessor is responsible for preprocessing various types of node data</span>
<span class="sd">    (assays, proteins, genes, and compounds) by renaming columns, consolidating</span>
<span class="sd">    multiple files, and saving the processed data. This preprocessing step is</span>
<span class="sd">    crucial for ensuring uniformity and ease of access in subsequent analysis</span>
<span class="sd">    and integration processes.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data_dir (str): The directory where the node data files are stored.</span>

<span class="sd">    Methods:</span>
<span class="sd">        preprocess_assays(): Processes and renames columns in assay data.</span>
<span class="sd">        preprocess_proteins(): Processes and renames columns in protein data.</span>
<span class="sd">        preprocess_genes(): Processes and renames columns in gene data.</span>
<span class="sd">        preprocess_compounds(): Consolidates and renames columns in compound data.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes the NodeDataProcessor with a directory path to manage the data files.</span>

<span class="sd">        Args:</span>
<span class="sd">            data_dir (str): The directory where the node data files are stored.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span> <span class="o">=</span> <span class="n">data_dir</span>


    <span class="k">def</span> <span class="nf">preprocess_assays</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes the assay data by renaming columns and saving the modified data back to disk.</span>
<span class="sd">        This method also handles visualization of assay data distributions if necessary.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Assay_Properties.csv&#39;</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;AID&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayID&quot;</span><span class="p">,</span> <span class="s2">&quot;Assay Type&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayType&quot;</span><span class="p">,</span>
                           <span class="s2">&quot;Activity Name&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayActivityName&quot;</span><span class="p">,</span> <span class="s2">&quot;SourceID&quot;</span><span class="p">:</span> <span class="s2">&quot;AssaySourceID&quot;</span><span class="p">,</span>
                           <span class="s2">&quot;SourceName&quot;</span><span class="p">:</span> <span class="s2">&quot;AssaySourceName&quot;</span><span class="p">,</span> <span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayName&quot;</span><span class="p">,</span>
                           <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayDescription&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Assay_Properties_Processed.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">preprocess_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes the protein data by renaming columns and saving the processed data.</span>
<span class="sd">        This method simplifies access to protein data for downstream analysis.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Protein_Properties.csv&#39;</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="s2">&quot;ProteinID&quot;</span><span class="p">,</span> <span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="s2">&quot;ProteinName&quot;</span><span class="p">,</span>
                           <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;ProteinDescription&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Protein_Properties_Processed.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">preprocess_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes gene data by renaming columns and changing data types for specific fields.</span>
<span class="sd">        The processed data is saved for further use in gene-related analyses.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Gene_Properties.csv&#39;</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Symbol&quot;</span><span class="p">:</span> <span class="s2">&quot;GeneSymbol&quot;</span><span class="p">,</span> <span class="s2">&quot;Taxonomy ID&quot;</span><span class="p">:</span> <span class="s2">&quot;TaxonomyID&quot;</span><span class="p">,</span>
                           <span class="s2">&quot;Synonyms&quot;</span><span class="p">:</span> <span class="s2">&quot;GeneSynonyms&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;Int64&#39;</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Gene_Properties_Processed.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">preprocess_compounds</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Concatenates multiple CSV files containing compound data into a single file,</span>
<span class="sd">        renames columns for uniformity, and saves the consolidated data. This method</span>
<span class="sd">        facilitates easier management and analysis of compound data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">path</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Compound_Properties&#39;</span>
        <span class="n">all_csv_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="s2">&quot;/*.csv&quot;</span><span class="p">)</span>
        <span class="n">first_file</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">output_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Data/Nodes/Compound_Properties.csv&#39;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">all_csv_files</span><span class="p">:</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_in</span><span class="p">:</span>
                    <span class="n">header</span> <span class="o">=</span> <span class="n">f_in</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">first_file</span><span class="p">:</span>
                        <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
                        <span class="n">first_file</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f_in</span><span class="p">:</span>
                        <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;CID&quot;</span><span class="p">:</span> <span class="s2">&quot;CompoundID&quot;</span><span class="p">,</span> <span class="s2">&quot;Title&quot;</span><span class="p">:</span> <span class="s2">&quot;CompoundName&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.csv&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;_Processed&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_data_processor.NodeDataProcessor.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">data_dir</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Initializes the NodeDataProcessor with a directory path to manage the data files.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>data_dir</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the node data files are stored.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the NodeDataProcessor with a directory path to manage the data files.</span>

<span class="sd">    Args:</span>
<span class="sd">        data_dir (str): The directory where the node data files are stored.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span> <span class="o">=</span> <span class="n">data_dir</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_assays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">preprocess_assays</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Processes the assay data by renaming columns and saving the modified data back to disk.
This method also handles visualization of assay data distributions if necessary.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">preprocess_assays</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the assay data by renaming columns and saving the modified data back to disk.</span>
<span class="sd">    This method also handles visualization of assay data distributions if necessary.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Assay_Properties.csv&#39;</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;AID&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayID&quot;</span><span class="p">,</span> <span class="s2">&quot;Assay Type&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayType&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;Activity Name&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayActivityName&quot;</span><span class="p">,</span> <span class="s2">&quot;SourceID&quot;</span><span class="p">:</span> <span class="s2">&quot;AssaySourceID&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;SourceName&quot;</span><span class="p">:</span> <span class="s2">&quot;AssaySourceName&quot;</span><span class="p">,</span> <span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayName&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;AssayDescription&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Assay_Properties_Processed.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_compounds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">preprocess_compounds</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Concatenates multiple CSV files containing compound data into a single file,
renames columns for uniformity, and saves the consolidated data. This method
facilitates easier management and analysis of compound data.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">preprocess_compounds</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Concatenates multiple CSV files containing compound data into a single file,</span>
<span class="sd">    renames columns for uniformity, and saves the consolidated data. This method</span>
<span class="sd">    facilitates easier management and analysis of compound data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">path</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Compound_Properties&#39;</span>
    <span class="n">all_csv_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="s2">&quot;/*.csv&quot;</span><span class="p">)</span>
    <span class="n">first_file</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">output_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Data/Nodes/Compound_Properties.csv&#39;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">all_csv_files</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_in</span><span class="p">:</span>
                <span class="n">header</span> <span class="o">=</span> <span class="n">f_in</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">first_file</span><span class="p">:</span>
                    <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
                    <span class="n">first_file</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f_in</span><span class="p">:</span>
                    <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;CID&quot;</span><span class="p">:</span> <span class="s2">&quot;CompoundID&quot;</span><span class="p">,</span> <span class="s2">&quot;Title&quot;</span><span class="p">:</span> <span class="s2">&quot;CompoundName&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.csv&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;_Processed&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_genes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">preprocess_genes</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Processes gene data by renaming columns and changing data types for specific fields.
The processed data is saved for further use in gene-related analyses.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">preprocess_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes gene data by renaming columns and changing data types for specific fields.</span>
<span class="sd">    The processed data is saved for further use in gene-related analyses.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Gene_Properties.csv&#39;</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Symbol&quot;</span><span class="p">:</span> <span class="s2">&quot;GeneSymbol&quot;</span><span class="p">,</span> <span class="s2">&quot;Taxonomy ID&quot;</span><span class="p">:</span> <span class="s2">&quot;TaxonomyID&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;Synonyms&quot;</span><span class="p">:</span> <span class="s2">&quot;GeneSynonyms&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;Int64&#39;</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Gene_Properties_Processed.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.node_data_processor.NodeDataProcessor.preprocess_proteins" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">preprocess_proteins</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Processes the protein data by renaming columns and saving the processed data.
This method simplifies access to protein data for downstream analysis.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/node_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">preprocess_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the protein data by renaming columns and saving the processed data.</span>
<span class="sd">    This method simplifies access to protein data for downstream analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Protein_Properties.csv&#39;</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="s2">&quot;ProteinID&quot;</span><span class="p">,</span> <span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="s2">&quot;ProteinName&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;ProteinDescription&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/Nodes/Protein_Properties_Processed.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h2 id="5-add-graph-nodes">5. Add Graph Nodes</h2>


<div class="doc doc-object doc-module">



<a id="chemgraphbuilder.add_graph_nodes"></a>
    <div class="doc doc-contents first">

      <p>Module for adding node data from CSV files to a Neo4j database.</p>
<p>This module provides a class and methods to read node data from CSV files
and add them to a Neo4j database, including creating uniqueness constraints
and generating Cypher queries.</p>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes" class="doc doc-heading">
            <code>AddGraphNodes</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="chemgraphbuilder.neo4jdriver.Neo4jBase" href="#chemgraphbuilder.neo4jdriver.Neo4jBase">Neo4jBase</a></code></p>


      <p>A class used to add node data from a CSV file or a directory of CSV files to a Neo4j database.</p>
<h4 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes--methods">Methods:</h4>
<p>create_uniqueness_constraint(driver, label, unique_property):
    Create a uniqueness constraint for the unique property of nodes in Neo4j.
generate_cypher_queries(node_dict, label, unique_property):
    Generate Cypher queries to update nodes in Neo4j based on the data from the CSV file.
execute_queries(queries):
    Execute a list of provided Cypher queries against the Neo4j database.
read_csv_file(file_path, unique_property):
    Read data from a CSV file and extract node properties.
combine_csv_files(input_directory):
    Combine multiple CSV files with the same columns into a single DataFrame.
process_and_add_nodes(file_path, label, unique_property):
    Process the CSV file and add node data to the Neo4j database.
process_and_add_nodes_from_directory(directory_path, label, unique_property):
    Combine CSV files from a directory and add node data to the Neo4j database.</p>

              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">AddGraphNodes</span><span class="p">(</span><span class="n">Neo4jBase</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class used to add node data from a CSV file or a directory of CSV files to a Neo4j database.</span>

<span class="sd">    Methods:</span>
<span class="sd">    --------</span>
<span class="sd">    create_uniqueness_constraint(driver, label, unique_property):</span>
<span class="sd">        Create a uniqueness constraint for the unique property of nodes in Neo4j.</span>
<span class="sd">    generate_cypher_queries(node_dict, label, unique_property):</span>
<span class="sd">        Generate Cypher queries to update nodes in Neo4j based on the data from the CSV file.</span>
<span class="sd">    execute_queries(queries):</span>
<span class="sd">        Execute a list of provided Cypher queries against the Neo4j database.</span>
<span class="sd">    read_csv_file(file_path, unique_property):</span>
<span class="sd">        Read data from a CSV file and extract node properties.</span>
<span class="sd">    combine_csv_files(input_directory):</span>
<span class="sd">        Combine multiple CSV files with the same columns into a single DataFrame.</span>
<span class="sd">    process_and_add_nodes(file_path, label, unique_property):</span>
<span class="sd">        Process the CSV file and add node data to the Neo4j database.</span>
<span class="sd">    process_and_add_nodes_from_directory(directory_path, label, unique_property):</span>
<span class="sd">        Combine CSV files from a directory and add node data to the Neo4j database.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">driver</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes the AddGraphNodes class with a Neo4j driver.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        driver : neo4j.GraphDatabase.driver</span>
<span class="sd">            A driver instance to connect to the Neo4j database.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">driver</span> <span class="o">=</span> <span class="n">driver</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;AddGraphNodes class initialized.&quot;</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">create_uniqueness_constraint</span><span class="p">(</span><span class="n">driver</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a uniqueness constraint for the unique property of nodes in Neo4j.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        driver : neo4j.GraphDatabase.driver</span>
<span class="sd">            A driver instance to connect to the Neo4j database.</span>
<span class="sd">        label : str</span>
<span class="sd">            The label of the node.</span>
<span class="sd">        unique_property : str</span>
<span class="sd">            The unique property of the node.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">constraint_query</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;CREATE CONSTRAINT IF NOT EXISTS FOR (n:</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">) &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;REQUIRE n.</span><span class="si">{</span><span class="n">unique_property</span><span class="si">}</span><span class="s2"> IS UNIQUE&quot;</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">driver</span><span class="o">.</span><span class="n">session</span><span class="p">()</span> <span class="k">as</span> <span class="n">session</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">session</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">constraint_query</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="s2">&quot;Uniqueness constraint created successfully on </span><span class="si">%s</span><span class="s2"> property of </span><span class="si">%s</span><span class="s2"> nodes.&quot;</span><span class="p">,</span>
                    <span class="n">unique_property</span><span class="p">,</span> <span class="n">label</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed to create uniqueness constraint: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_generate_property_string</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)):</span>
            <span class="k">return</span> <span class="n">value</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">TypeError</span><span class="p">,</span> <span class="ne">ValueError</span><span class="p">):</span>
            <span class="n">escaped_value</span> <span class="o">=</span> <span class="s2">&quot;&#39;&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">n&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;&#39;&quot;</span>
            <span class="k">return</span> <span class="n">escaped_value</span>

    <span class="k">def</span> <span class="nf">generate_cypher_queries</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_dict</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate Cypher queries for updating Neo4j based on the provided node data dictionary.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        node_dict : dict</span>
<span class="sd">            A dictionary with unique identifiers as keys and node data as values.</span>
<span class="sd">        label : str</span>
<span class="sd">            The label of the node.</span>
<span class="sd">        unique_property : str</span>
<span class="sd">            The unique property of the node.</span>

<span class="sd">        Yields:</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            A Cypher query string.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Create an index for the unique_property</span>
        <span class="n">create_index_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;CREATE INDEX IF NOT EXISTS FOR (n:</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">) ON (n.</span><span class="si">{</span><span class="n">unique_property</span><span class="si">}</span><span class="s2">)&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">create_index_query</span><span class="p">)</span>
        <span class="k">yield</span> <span class="n">create_index_query</span>

        <span class="k">for</span> <span class="n">unique_id</span><span class="p">,</span> <span class="n">properties</span> <span class="ow">in</span> <span class="n">node_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">unique_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;&quot;</span><span class="si">{</span><span class="n">unique_id</span><span class="si">}</span><span class="s1">&quot;&#39;</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">unique_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">unique_id</span>
            <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;MERGE (n:</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2"> </span><span class="se">{{</span><span class="si">{</span><span class="n">unique_property</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">unique_id</span><span class="si">}</span><span class="se">}}</span><span class="s2">)&quot;</span>
            <span class="n">set_clauses</span> <span class="o">=</span> <span class="p">[</span>
                <span class="sa">f</span><span class="s2">&quot;n.</span><span class="si">{</span><span class="n">prop</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_generate_property_string</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">for</span> <span class="n">prop</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">properties</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">]</span>
            <span class="k">if</span> <span class="n">set_clauses</span><span class="p">:</span>
                <span class="n">query</span> <span class="o">+=</span> <span class="s2">&quot; SET &quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">set_clauses</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">query</span> <span class="o">+=</span> <span class="s2">&quot;;&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
            <span class="k">yield</span> <span class="n">query</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cypher queries generated successfully.&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">execute_queries</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">queries</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute the provided list of Cypher queries against the Neo4j database.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        queries : list</span>
<span class="sd">            A list of Cypher query strings to execute.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Executing Cypher queries...&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">driver</span><span class="o">.</span><span class="n">session</span><span class="p">()</span> <span class="k">as</span> <span class="n">session</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Executing Cypher queries Started....&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">query</span> <span class="ow">in</span> <span class="n">queries</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">session</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed to execute query: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;All queries executed.&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">read_csv_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read data from a CSV file and extract node properties.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        file_path : str</span>
<span class="sd">            The path to the CSV file.</span>
<span class="sd">        unique_property : str</span>
<span class="sd">            The column name that serves as the unique identifier for the nodes.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">        dict</span>
<span class="sd">            A dictionary with unique identifiers as keys and extracted data as values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading data from CSV file: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="n">unique_property</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;any&#39;</span><span class="p">)</span>
        <span class="n">node_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">row</span><span class="p">[</span><span class="n">unique_property</span><span class="p">]:</span> <span class="n">row</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="n">unique_property</span><span class="p">])</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully read data for </span><span class="si">%d</span><span class="s2"> nodes from CSV.&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">node_dict</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">node_dict</span>

    <span class="k">def</span> <span class="nf">combine_csv_files</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_directory</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Combine multiple CSV files with the same columns into a single DataFrame.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        input_directory : str</span>
<span class="sd">            The directory containing the CSV files to be combined.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">        DataFrame</span>
<span class="sd">            A combined DataFrame containing data from all the CSV files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Combining CSV files from directory: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">input_directory</span><span class="p">)</span>
        <span class="n">dfs</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">input_directory</span><span class="p">,</span> <span class="n">file</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">input_directory</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.csv&quot;</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="n">combined_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully combined </span><span class="si">%d</span><span class="s2"> CSV files.&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">dfs</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">combined_df</span>

    <span class="k">def</span> <span class="nf">process_and_add_nodes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Process the CSV file and add node data to the Neo4j database.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        file_path : str</span>
<span class="sd">            The path to the CSV file.</span>
<span class="sd">        label : str</span>
<span class="sd">            The label of the node.</span>
<span class="sd">        unique_property : str</span>
<span class="sd">            The unique property of the node.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Processing and adding nodes from file: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
        <span class="n">node_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_csv_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span>
        <span class="n">queries</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_cypher_queries</span><span class="p">(</span><span class="n">node_dict</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">execute_queries</span><span class="p">(</span><span class="n">queries</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully processed and added nodes from file: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">process_and_add_nodes_from_directory</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Combine CSV files from a directory and add node data to the Neo4j database.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        directory_path : str</span>
<span class="sd">            The path to the directory containing the CSV files.</span>
<span class="sd">        label : str</span>
<span class="sd">            The label of the node.</span>
<span class="sd">        unique_property : str</span>
<span class="sd">            The unique property of the node.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Processing and adding nodes from directory: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">)</span>
        <span class="n">combined_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">combine_csv_files</span><span class="p">(</span><span class="n">directory_path</span><span class="p">)</span>
        <span class="n">temp_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory_path</span><span class="p">,</span> <span class="s2">&quot;combined_temp.csv&quot;</span><span class="p">)</span>
        <span class="n">combined_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">temp_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">process_and_add_nodes</span><span class="p">(</span><span class="n">temp_file</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">temp_file</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully processed and added nodes from directory: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                         <span class="n">directory_path</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">public_generate_property_string</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Public method to access the protected _generate_property_string method for testing.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        value : Any</span>
<span class="sd">            The value to be formatted.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">        str</span>
<span class="sd">            The formatted property string.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generate_property_string</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">driver</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Initializes the AddGraphNodes class with a Neo4j driver.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.__init__--parameters">Parameters:</h5>
<p>driver : neo4j.GraphDatabase.driver
    A driver instance to connect to the Neo4j database.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">driver</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the AddGraphNodes class with a Neo4j driver.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    driver : neo4j.GraphDatabase.driver</span>
<span class="sd">        A driver instance to connect to the Neo4j database.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">driver</span> <span class="o">=</span> <span class="n">driver</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;AddGraphNodes class initialized.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">combine_csv_files</span><span class="p">(</span><span class="n">input_directory</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Combine multiple CSV files with the same columns into a single DataFrame.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files--parameters">Parameters:</h5>
<p>input_directory : str
    The directory containing the CSV files to be combined.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.combine_csv_files--returns">Returns:</h5>
<p>DataFrame
    A combined DataFrame containing data from all the CSV files.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">combine_csv_files</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_directory</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Combine multiple CSV files with the same columns into a single DataFrame.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    input_directory : str</span>
<span class="sd">        The directory containing the CSV files to be combined.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    DataFrame</span>
<span class="sd">        A combined DataFrame containing data from all the CSV files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Combining CSV files from directory: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">input_directory</span><span class="p">)</span>
    <span class="n">dfs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">input_directory</span><span class="p">,</span> <span class="n">file</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">input_directory</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.csv&quot;</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="n">combined_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully combined </span><span class="si">%d</span><span class="s2"> CSV files.&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">dfs</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">combined_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.create_uniqueness_constraint" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_uniqueness_constraint</span><span class="p">(</span><span class="n">driver</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

      <p>Create a uniqueness constraint for the unique property of nodes in Neo4j.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.create_uniqueness_constraint--parameters">Parameters:</h5>
<p>driver : neo4j.GraphDatabase.driver
    A driver instance to connect to the Neo4j database.
label : str
    The label of the node.
unique_property : str
    The unique property of the node.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">create_uniqueness_constraint</span><span class="p">(</span><span class="n">driver</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a uniqueness constraint for the unique property of nodes in Neo4j.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    driver : neo4j.GraphDatabase.driver</span>
<span class="sd">        A driver instance to connect to the Neo4j database.</span>
<span class="sd">    label : str</span>
<span class="sd">        The label of the node.</span>
<span class="sd">    unique_property : str</span>
<span class="sd">        The unique property of the node.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">constraint_query</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;CREATE CONSTRAINT IF NOT EXISTS FOR (n:</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">) &quot;</span>
        <span class="sa">f</span><span class="s2">&quot;REQUIRE n.</span><span class="si">{</span><span class="n">unique_property</span><span class="si">}</span><span class="s2"> IS UNIQUE&quot;</span>
    <span class="p">)</span>
    <span class="k">with</span> <span class="n">driver</span><span class="o">.</span><span class="n">session</span><span class="p">()</span> <span class="k">as</span> <span class="n">session</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">session</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">constraint_query</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Uniqueness constraint created successfully on </span><span class="si">%s</span><span class="s2"> property of </span><span class="si">%s</span><span class="s2"> nodes.&quot;</span><span class="p">,</span>
                <span class="n">unique_property</span><span class="p">,</span> <span class="n">label</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed to create uniqueness constraint: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.execute_queries" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">execute_queries</span><span class="p">(</span><span class="n">queries</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Execute the provided list of Cypher queries against the Neo4j database.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.execute_queries--parameters">Parameters:</h5>
<p>queries : list
    A list of Cypher query strings to execute.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">execute_queries</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">queries</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Execute the provided list of Cypher queries against the Neo4j database.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    queries : list</span>
<span class="sd">        A list of Cypher query strings to execute.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Executing Cypher queries...&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">driver</span><span class="o">.</span><span class="n">session</span><span class="p">()</span> <span class="k">as</span> <span class="n">session</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Executing Cypher queries Started....&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">query</span> <span class="ow">in</span> <span class="n">queries</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">session</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed to execute query: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;All queries executed.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_cypher_queries</span><span class="p">(</span><span class="n">node_dict</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Generate Cypher queries for updating Neo4j based on the provided node data dictionary.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries--parameters">Parameters:</h5>
<p>node_dict : dict
    A dictionary with unique identifiers as keys and node data as values.
label : str
    The label of the node.
unique_property : str
    The unique property of the node.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.generate_cypher_queries--yields">Yields:</h5>
<p>str
    A Cypher query string.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_cypher_queries</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_dict</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate Cypher queries for updating Neo4j based on the provided node data dictionary.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    node_dict : dict</span>
<span class="sd">        A dictionary with unique identifiers as keys and node data as values.</span>
<span class="sd">    label : str</span>
<span class="sd">        The label of the node.</span>
<span class="sd">    unique_property : str</span>
<span class="sd">        The unique property of the node.</span>

<span class="sd">    Yields:</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        A Cypher query string.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create an index for the unique_property</span>
    <span class="n">create_index_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;CREATE INDEX IF NOT EXISTS FOR (n:</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">) ON (n.</span><span class="si">{</span><span class="n">unique_property</span><span class="si">}</span><span class="s2">)&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">create_index_query</span><span class="p">)</span>
    <span class="k">yield</span> <span class="n">create_index_query</span>

    <span class="k">for</span> <span class="n">unique_id</span><span class="p">,</span> <span class="n">properties</span> <span class="ow">in</span> <span class="n">node_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">unique_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;&quot;</span><span class="si">{</span><span class="n">unique_id</span><span class="si">}</span><span class="s1">&quot;&#39;</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">unique_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">unique_id</span>
        <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;MERGE (n:</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2"> </span><span class="se">{{</span><span class="si">{</span><span class="n">unique_property</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">unique_id</span><span class="si">}</span><span class="se">}}</span><span class="s2">)&quot;</span>
        <span class="n">set_clauses</span> <span class="o">=</span> <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;n.</span><span class="si">{</span><span class="n">prop</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_generate_property_string</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="n">prop</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">properties</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">]</span>
        <span class="k">if</span> <span class="n">set_clauses</span><span class="p">:</span>
            <span class="n">query</span> <span class="o">+=</span> <span class="s2">&quot; SET &quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">set_clauses</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">query</span> <span class="o">+=</span> <span class="s2">&quot;;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
        <span class="k">yield</span> <span class="n">query</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cypher queries generated successfully.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_and_add_nodes</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Process the CSV file and add node data to the Neo4j database.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes--parameters">Parameters:</h5>
<p>file_path : str
    The path to the CSV file.
label : str
    The label of the node.
unique_property : str
    The unique property of the node.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">process_and_add_nodes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Process the CSV file and add node data to the Neo4j database.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    file_path : str</span>
<span class="sd">        The path to the CSV file.</span>
<span class="sd">    label : str</span>
<span class="sd">        The label of the node.</span>
<span class="sd">    unique_property : str</span>
<span class="sd">        The unique property of the node.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Processing and adding nodes from file: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
    <span class="n">node_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_csv_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span>
    <span class="n">queries</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">generate_cypher_queries</span><span class="p">(</span><span class="n">node_dict</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">))</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">execute_queries</span><span class="p">(</span><span class="n">queries</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully processed and added nodes from file: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes_from_directory" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_and_add_nodes_from_directory</span><span class="p">(</span><span class="n">directory_path</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Combine CSV files from a directory and add node data to the Neo4j database.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.process_and_add_nodes_from_directory--parameters">Parameters:</h5>
<p>directory_path : str
    The path to the directory containing the CSV files.
label : str
    The label of the node.
unique_property : str
    The unique property of the node.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">process_and_add_nodes_from_directory</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Combine CSV files from a directory and add node data to the Neo4j database.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    directory_path : str</span>
<span class="sd">        The path to the directory containing the CSV files.</span>
<span class="sd">    label : str</span>
<span class="sd">        The label of the node.</span>
<span class="sd">    unique_property : str</span>
<span class="sd">        The unique property of the node.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Processing and adding nodes from directory: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">)</span>
    <span class="n">combined_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">combine_csv_files</span><span class="p">(</span><span class="n">directory_path</span><span class="p">)</span>
    <span class="n">temp_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory_path</span><span class="p">,</span> <span class="s2">&quot;combined_temp.csv&quot;</span><span class="p">)</span>
    <span class="n">combined_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">temp_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">process_and_add_nodes</span><span class="p">(</span><span class="n">temp_file</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">temp_file</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully processed and added nodes from directory: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                     <span class="n">directory_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">public_generate_property_string</span><span class="p">(</span><span class="n">value</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Public method to access the protected _generate_property_string method for testing.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string--parameters">Parameters:</h5>
<p>value : Any
    The value to be formatted.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.public_generate_property_string--returns">Returns:</h5>
<p>str
    The formatted property string.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">public_generate_property_string</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Public method to access the protected _generate_property_string method for testing.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    value : Any</span>
<span class="sd">        The value to be formatted.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    str</span>
<span class="sd">        The formatted property string.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generate_property_string</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read_csv_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Read data from a CSV file and extract node properties.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file--parameters">Parameters:</h5>
<p>file_path : str
    The path to the CSV file.
unique_property : str
    The column name that serves as the unique identifier for the nodes.</p>
<h5 id="chemgraphbuilder.add_graph_nodes.AddGraphNodes.read_csv_file--returns">Returns:</h5>
<p>dict
    A dictionary with unique identifiers as keys and extracted data as values.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/add_graph_nodes.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">read_csv_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">,</span> <span class="n">unique_property</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read data from a CSV file and extract node properties.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    file_path : str</span>
<span class="sd">        The path to the CSV file.</span>
<span class="sd">    unique_property : str</span>
<span class="sd">        The column name that serves as the unique identifier for the nodes.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    dict</span>
<span class="sd">        A dictionary with unique identifiers as keys and extracted data as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading data from CSV file: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="n">unique_property</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;any&#39;</span><span class="p">)</span>
    <span class="n">node_dict</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">row</span><span class="p">[</span><span class="n">unique_property</span><span class="p">]:</span> <span class="n">row</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="n">unique_property</span><span class="p">])</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()</span>
    <span class="p">}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Successfully read data for </span><span class="si">%d</span><span class="s2"> nodes from CSV.&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">node_dict</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">node_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h2 id="6-relationship-properties-extractor">6. Relationship Properties Extractor</h2>


<div class="doc doc-object doc-module">



<a id="chemgraphbuilder.relationship_properties_extractor"></a>
    <div class="doc doc-contents first">

      <p>This module defines the <code>RelationshipPropertiesExtractor</code> class, which is responsible for extracting and analyzing
relationship properties among compounds, genes, and assays from the PubChem database.</p>
<p>The class facilitates the retrieval of complex relational data between chemical entities, enabling detailed analysis
of biochemical interactions and properties. The extracted data is ideal for constructing knowledge graphs, supporting
drug discovery, and understanding genetic influences on compound behavior.</p>


<p><span class="doc-section-title">Classes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.relationship_properties_extractor.- RelationshipPropertiesExtractor">- RelationshipPropertiesExtractor</span></code></td>
            <td>
              <div class="doc-md-description">
                <p>A class to extract and analyze relationship properties from PubChem.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="usage-example" open>
  <summary>Usage Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = RelationshipPropertiesExtractor()
extractor.assay_compound_relationship("Data/AllDataCollected.csv")
This example fetches assay-compound relationship data for specified assays and saves the data to CSV files.</p>
</blockquote>
</blockquote>
</blockquote>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>Ensure network access to the PubChem API for data retrieval.</p>
</details>


  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor" class="doc doc-heading">
            <code>RelationshipPropertiesExtractor</code>


</h2>


    <div class="doc doc-contents ">


      <p>Extracts and analyzes relationship properties among compounds, genes, and
assays from the PubChem database.</p>
<p>This class facilitates the retrieval of complex relational data between
chemical entities, enabling detailed analysis of biochemical interactions
and properties. The extracted data is ideal for constructing knowledge
graphs, supporting drug discovery, and understanding genetic influences
on compound behavior.</p>
<p>Methods within the class are tailored to query specific relationship types
from PubChem, including compound-assay relationships, compound co-occurrences,
and compound transformations influenced by genes. Data fetched from PubChem
is processed and saved in structured formats (CSV files), ready for further
analysis or database integration.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.session">session</span></code></td>
            <td>
                  <code><span title="requests.Session">Session</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Session object to persist certain parameters</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="usage" open>
  <summary>Usage</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = RelationshipPropertiesExtractor()
extractor.assay_compound_relationship("Data/AllDataCollected.csv")
This example fetches assay-compound relationship data for specified
assays and saves the data to CSV files.</p>
</blockquote>
</blockquote>
</blockquote>
</details>
              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">RelationshipPropertiesExtractor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts and analyzes relationship properties among compounds, genes, and</span>
<span class="sd">    assays from the PubChem database.</span>

<span class="sd">    This class facilitates the retrieval of complex relational data between</span>
<span class="sd">    chemical entities, enabling detailed analysis of biochemical interactions</span>
<span class="sd">    and properties. The extracted data is ideal for constructing knowledge</span>
<span class="sd">    graphs, supporting drug discovery, and understanding genetic influences</span>
<span class="sd">    on compound behavior.</span>

<span class="sd">    Methods within the class are tailored to query specific relationship types</span>
<span class="sd">    from PubChem, including compound-assay relationships, compound co-occurrences,</span>
<span class="sd">    and compound transformations influenced by genes. Data fetched from PubChem</span>
<span class="sd">    is processed and saved in structured formats (CSV files), ready for further</span>
<span class="sd">    analysis or database integration.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        session (requests.Session): Session object to persist certain parameters</span>
<span class="sd">        across requests.</span>

<span class="sd">    Usage:</span>
<span class="sd">        &gt;&gt;&gt; extractor = RelationshipPropertiesExtractor()</span>
<span class="sd">        &gt;&gt;&gt; extractor.assay_compound_relationship(&quot;Data/AllDataCollected.csv&quot;)</span>
<span class="sd">        This example fetches assay-compound relationship data for specified</span>
<span class="sd">        assays and saves the data to CSV files.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initializes a RelationshipPropertiesExtractor with a Requests session</span>
<span class="sd">         for efficient network calls.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">session</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">Session</span><span class="p">()</span>


    <span class="k">def</span> <span class="nf">_send_request</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">url</span><span class="p">,</span> <span class="n">max_retries</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">initial_wait</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">attempt</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_retries</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">session</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
                <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
                <span class="k">return</span> <span class="n">response</span>
            <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">HTTPError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">503</span><span class="p">:</span>
                    <span class="n">wait</span> <span class="o">=</span> <span class="n">initial_wait</span> <span class="o">*</span> <span class="p">(</span><span class="mi">2</span> <span class="o">**</span> <span class="n">attempt</span><span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Server busy or under maintenance. Retrying in </span><span class="si">{</span><span class="n">wait</span><span class="si">}</span><span class="s2"> seconds...&quot;</span><span class="p">)</span>
                    <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">wait</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HTTP Error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">break</span>  <span class="c1"># Break the loop for non-503 HTTP errors</span>
            <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Request Exception: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">wait</span> <span class="o">=</span> <span class="n">initial_wait</span> <span class="o">*</span> <span class="p">(</span><span class="mi">2</span> <span class="o">**</span> <span class="n">attempt</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Network error. Retrying in </span><span class="si">{</span><span class="n">wait</span><span class="si">}</span><span class="s2"> seconds...&quot;</span><span class="p">)</span>
                <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">wait</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>  <span class="c1"># Return None to indicate failure after all retries</span>


    <span class="k">def</span> <span class="nf">fetch_data_for_aid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">aid</span><span class="p">,</span> <span class="n">columns_to_remove</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches and processes assay data for a specified Assay ID (AID) from the</span>
<span class="sd">        PubChem database, preparing it for analysis or further processing.</span>

<span class="sd">        This method queries the PubChem database for assay data associated with</span>
<span class="sd">        a given AID. It constructs the query URL, sends the request using a</span>
<span class="sd">        previously established session, and processes the response. The response</span>
<span class="sd">        is expected to be in CSV format, which this method reads into a pandas</span>
<span class="sd">        DataFrame. Specific columns can be removed from this DataFrame based on</span>
<span class="sd">        the requirements for analysis. This allows for the customization of</span>
<span class="sd">        the fetched data, making it easier to work with specific datasets.</span>

<span class="sd">        If the request is successful and the data is fetched without issues,</span>
<span class="sd">        it undergoes initial processing to remove unwanted columns as specified</span>
<span class="sd">        by the &#39;columns_to_remove&#39; parameter. In case of an error during the</span>
<span class="sd">        data fetching or processing (e.g., issues with parsing the CSV data),</span>
<span class="sd">        appropriate error messages are logged, and an empty DataFrame is</span>
<span class="sd">        returned as a fallback.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            aid (int): The assay ID for which data is to be fetched. This ID is</span>
<span class="sd">            used to construct the query URL to the PubChem database.</span>
<span class="sd">            columns_to_remove (list of str): A list of column names that should</span>
<span class="sd">            be removed from the fetched DataFrame. This allows for the exclusion</span>
<span class="sd">            of data that might not be relevant to the subsequent analysis or</span>
<span class="sd">            processing steps.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pandas.DataFrame: A DataFrame containing the processed data</span>
<span class="sd">            associated with the given AID. The DataFrame will exclude columns</span>
<span class="sd">            listed in &#39;columns_to_remove&#39;. If the data fetching fails or if</span>
<span class="sd">            an error occurs during processing, an empty DataFrame is returned.</span>

<span class="sd">        Raises:</span>
<span class="sd">            requests.RequestException: If an error occurs during the HTTP request</span>
<span class="sd">            to the PubChem API. This includes scenarios such as timeout issues,</span>
<span class="sd">            non-200 status codes, or network-related errors. The exception is</span>
<span class="sd">            handled internally with logging, but it&#39;s important to be aware of</span>
<span class="sd">            its possibility.</span>
<span class="sd">            pd.errors.ParserError: If an error occurs while parsing the CSV</span>
<span class="sd">            response from PubChem into a DataFrame. This could happen due to</span>
<span class="sd">            malformed data or unexpected changes in the response format.</span>
<span class="sd">            Like with RequestException, this error is logged and results in</span>
<span class="sd">            the return of an empty DataFrame.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; extractor = RelationshipPropertiesExtractor()</span>
<span class="sd">            &gt;&gt;&gt; processed_data_df = extractor.fetch_data_for_aid(12345, [&#39;UnwantedColumn1&#39;, &#39;UnwantedColumn2&#39;])</span>
<span class="sd">            &gt;&gt;&gt; print(processed_data_df.head())</span>
<span class="sd">            This example demonstrates how to fetch and process assay data for</span>
<span class="sd">            the assay with ID 12345, removing &#39;UnwantedColumn1&#39; and</span>
<span class="sd">            &#39;UnwantedColumn2&#39; from the resulting DataFrame. The first few rows</span>
<span class="sd">            of the processed DataFrame are printed as an output.</span>

<span class="sd">        Note:</span>
<span class="sd">            - This method is part of a class that requires a valid session with</span>
<span class="sd">            the PubChem API. Ensure that the class is properly initialized and that</span>
<span class="sd">            the session is active.</span>
<span class="sd">            - The removal of columns is an optional step and can be customized</span>
<span class="sd">            based on the analysis needs. If no columns need to be removed, pass an</span>
<span class="sd">            empty list as &#39;columns_to_remove&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">url</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/assay/pcget.cgi?&quot;</span>
            <span class="s2">&quot;query=download&amp;record_type=datatable&amp;actvty=&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;all&amp;response_type=display&amp;aid=</span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_send_request</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">response</span> <span class="ow">and</span> <span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">compound_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
                <span class="c1"># Drop specified columns and process column names in-place for memory efficiency</span>
                <span class="n">columns_to_remove_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">columns_to_remove</span><span class="p">)</span>
                <span class="n">existing_columns_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">compound_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
                <span class="n">columns_to_actually_remove</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">columns_to_remove_set</span> <span class="o">&amp;</span> <span class="n">existing_columns_set</span><span class="p">)</span>
                <span class="n">compound_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">columns_to_actually_remove</span><span class="p">,</span>
                                 <span class="n">errors</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">compound_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;PUBCHEM_&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;PUBCHEM_&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

                <span class="c1"># compound_df.drop(columns=[col for col in columns_to_remove if col in compound_df.columns], errors=&#39;ignore&#39;, inplace=True)</span>
                <span class="c1"># compound_df.columns = [col.replace(&#39;PUBCHEM_&#39;, &#39;&#39;) if col.startswith(&#39;PUBCHEM_&#39;) else col for col in compound_df.columns]</span>
                <span class="n">compound_df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aid</span>
                <span class="k">return</span> <span class="n">compound_df</span>
            <span class="k">except</span> <span class="n">pd</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">ParserError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CSV parsing failed for AID </span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to fetch data for AID </span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">. Status code: </span><span class="si">{</span><span class="n">response</span><span class="o">.</span><span class="n">status_code</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">response</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;No Response&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># Return an empty DataFrame in case of failure</span>



    <span class="k">def</span> <span class="nf">_process_dataframe</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">aid</span><span class="p">,</span> <span class="n">columns_to_remove</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes the DataFrame by removing specified columns and renaming others.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            df (pandas.DataFrame): The DataFrame to be processed.</span>
<span class="sd">            aid (int): The assay ID associated with the DataFrame.</span>
<span class="sd">            columns_to_remove (list of str): Columns to be removed from the DataFrame.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Drop unnecessary columns efficiently</span>
        <span class="n">columns_to_remove_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">columns_to_remove</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">columns_to_remove_set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)),</span> <span class="n">errors</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

        <span class="c1"># Efficiently rename columns that start with &#39;PUBCHEM_&#39;</span>
        <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;PUBCHEM_&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;PUBCHEM_&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aid</span>


    <span class="k">def</span> <span class="nf">assay_compound_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assays_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes and stores relationships between assays and compounds based</span>
<span class="sd">        on assay data from PubChem.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            assays_data (str): Path to a CSV file containing assay IDs (AIDs).</span>
<span class="sd">            start_chunk (int): The starting index for processing chunks.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">chunk_idx</span><span class="p">,</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">assays_data</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">100</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">chunk_idx</span> <span class="o">&gt;=</span> <span class="n">start_chunk</span><span class="p">:</span>
                <span class="n">columns_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PUBCHEM_RESULT_TAG&#39;</span><span class="p">,</span> <span class="s1">&#39;PUBCHEM_SID&#39;</span><span class="p">,</span> <span class="s1">&#39;PUBCHEM_EXT_DATASOURCE_SMILES&#39;</span><span class="p">]</span>
                <span class="n">output_dir</span> <span class="o">=</span> <span class="s1">&#39;Data/Relationships/Assay_Compound_Relationship&#39;</span>

                <span class="k">for</span> <span class="n">aid</span> <span class="ow">in</span> <span class="n">chunk</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s1">/AID_</span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">):</span>
                        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fetch_data_for_aid</span><span class="p">(</span><span class="n">aid</span><span class="p">,</span> <span class="n">columns_to_remove</span><span class="p">)</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output_dir</span><span class="p">):</span>
                                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
                            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s1">/AID_</span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed chunk </span><span class="si">{</span><span class="n">chunk_idx</span><span class="si">}</span><span class="s2"> for assay-compound relationships.&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No More Chunck to Process.&quot;</span><span class="p">)</span>



    <span class="k">def</span> <span class="nf">_write_to_csv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes a DataFrame to a CSV file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">assay_enzyme_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts and saves relationships between assays and enzymes from the</span>
<span class="sd">        specified dataset.</span>

<span class="sd">        This method processes assay data to identify relationships between</span>
<span class="sd">        assays and their target enzymes. It selects relevant columns from the</span>
<span class="sd">        input data, removes duplicates to ensure unique relationships, and saves</span>
<span class="sd">        the cleaned data to a CSV file for further analysis or integration into</span>
<span class="sd">        knowledge graphs.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to the CSV file containing the main data. The</span>
<span class="sd">            file should include columns for &#39;AID&#39; (Assay ID), &#39;Target GeneID&#39;,</span>
<span class="sd">            and &#39;Activity Name&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pandas.DataFrame: A DataFrame containing the unique relationships</span>
<span class="sd">            between assays and enzymes, including the assay ID, target gene ID,</span>
<span class="sd">            and activity name.</span>

<span class="sd">        Side Effects:</span>
<span class="sd">            - Writes a CSV file to &#39;Data/Assay_Enzyme_Relationship.csv&#39;,</span>
<span class="sd">            containing the processed relationships data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
        <span class="n">columns_to_select</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Activity Name&#39;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">columns_to_select</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Assay_Enzyme_Relationship.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>


    <span class="k">def</span> <span class="nf">gene_enzyme_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts and saves relationships between genes and enzymes based on</span>
<span class="sd">        the provided dataset.</span>

<span class="sd">        This method selects relevant columns to highlight the relationships</span>
<span class="sd">        between genes and their corresponding enzymes.</span>
<span class="sd">        It removes duplicate entries to ensure that each relationship is</span>
<span class="sd">        represented uniquely and saves the resultant data to</span>
<span class="sd">        a CSV file. This facilitates easy integration of genetic data into</span>
<span class="sd">        knowledge bases or further analysis.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to the CSV file containing gene and enzyme data.</span>
<span class="sd">            Expected columns include &#39;Target GeneID&#39; and &#39;Target Accession&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pandas.DataFrame: A DataFrame of unique gene-enzyme relationships,</span>
<span class="sd">            including gene ID and enzyme accession numbers.</span>

<span class="sd">        Side Effects:</span>
<span class="sd">            - Writes the processed data to &#39;Data/Gene_Enzyme_Relationship.csv&#39;</span>
<span class="sd">            in a structured CSV format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
        <span class="n">columns_to_select</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Accession&#39;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">columns_to_select</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Gene_Enzyme_Relationship.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>


    <span class="k">def</span> <span class="nf">compound_gene_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Identifies and records relationships between compounds and enzymes from</span>
<span class="sd">        the input data.</span>

<span class="sd">        This method focuses on extracting compound-enzyme interaction data,</span>
<span class="sd">        including activity outcomes and values. It selects</span>
<span class="sd">        pertinent columns, removes duplicate records, and sorts the data by</span>
<span class="sd">        Compound ID and Target Accession for clarity. The cleaned dataset is</span>
<span class="sd">        then saved to a CSV file, providing a structured view  of how compounds</span>
<span class="sd">        interact with various enzymes, which can be critical for drug discovery</span>
<span class="sd">        and pharmacological research.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to the CSV file with compound and enzyme data.</span>
<span class="sd">            This file should contain columns for &#39;CID&#39; (Compound ID),</span>
<span class="sd">            &#39;Target Accession&#39;, &#39;Activity Outcome&#39;, &#39;Activity Name&#39;, and</span>
<span class="sd">            &#39;Activity Value [uM]&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pandas.DataFrame: A DataFrame with processed compound-enzyme</span>
<span class="sd">            relationships, sorted and cleaned for direct analysis or database</span>
<span class="sd">            insertion.</span>

<span class="sd">        Side Effects:</span>
<span class="sd">            - Saves the processed relationships data to</span>
<span class="sd">            &#39;Data/Relationships/Compound_Gene_Relationship.csv&#39;,</span>
<span class="sd">            facilitating easy access and integration.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
        <span class="n">columns_to_select</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Accession&#39;</span><span class="p">,</span>
                             <span class="s1">&#39;Activity Outcome&#39;</span><span class="p">,</span> <span class="s1">&#39;Activity Name&#39;</span><span class="p">,</span>
                             <span class="s1">&#39;Activity Value [uM]&#39;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">columns_to_select</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;CID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Accession&#39;</span><span class="p">])</span>
        <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span> <span class="p">,</span> <span class="n">thresh</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1">###</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Relationships/Compound_Gene_Relationship.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>


    <span class="k">def</span> <span class="nf">fetch_similar_cids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches similar compound IDs (CIDs) from the PubChem database for a</span>
<span class="sd">        given compound ID (CID) using 2D similarity.</span>

<span class="sd">        This method queries the PubChem database to find compounds that are</span>
<span class="sd">        similar to the given CID based on 2D structural similarity.</span>
<span class="sd">        The similarity threshold is set to 95%, and a maximum of 100 similar</span>
<span class="sd">        CIDs are fetched. The response is parsed from XML format to extract</span>
<span class="sd">        the similar CIDs.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            cid (int): The compound ID for which similar CIDs are to be fetched.</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: A tuple containing the original CID and a list of similar</span>
<span class="sd">            CIDs. If an error occurs, the list of similar CIDs will be empty.</span>

<span class="sd">        Raises:</span>
<span class="sd">            Exception: Logs an error message with the original CID and the</span>
<span class="sd">            exception if the request to PubChem fails or if parsing the XML</span>
<span class="sd">            response encounters an error.</span>

<span class="sd">        Note:</span>
<span class="sd">            - The method utilizes the `requests` library for HTTP requests and</span>
<span class="sd">            `xml.etree.ElementTree` for XML parsing.</span>
<span class="sd">            - In case of a request failure or parsing error, the method logs</span>
<span class="sd">            the error and returns the original CID with an empty list,</span>
<span class="sd">            allowing the calling function to handle the exception as needed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">url</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/&quot;</span>
               <span class="sa">f</span><span class="s2">&quot;fastsimilarity_2d/cid/</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span><span class="si">}</span><span class="s2">/cids/XML?Threshold=95&amp;MaxRecords=100&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
            <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
            <span class="n">xml_data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">text</span>

            <span class="c1"># Parse XML data</span>
            <span class="n">tree</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">xml_data</span><span class="p">))</span>
            <span class="n">root</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">getroot</span><span class="p">()</span>

            <span class="c1"># Extracting CID values</span>
            <span class="n">similar_cids</span> <span class="o">=</span> <span class="p">[</span><span class="n">element</span><span class="o">.</span><span class="n">text</span> <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;{http://pubchem.ncbi.nlm.nih.gov/pug_rest}CID&#39;</span><span class="p">)]</span>
            <span class="k">return</span> <span class="n">cid</span><span class="p">,</span> <span class="n">similar_cids</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing CID </span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">cid</span><span class="p">,</span> <span class="p">[]</span>


    <span class="k">def</span> <span class="nf">process_chunk</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chunk</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes a chunk of CIDs in parallel to fetch similar CIDs for each CID</span>
<span class="sd">        in the chunk.</span>

<span class="sd">        This method uses a ThreadPoolExecutor to send out concurrent requests for</span>
<span class="sd">        fetching similar CIDs for a list of CIDs.</span>
<span class="sd">        The number of worker threads is set to 5. Each CID&#39;s request is handled</span>
<span class="sd">        by `fetch_similar_cids` method.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            chunk (list of int): A list of compound IDs (CIDs) to process in</span>
<span class="sd">            parallel.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list of tuples: A list of tuples, each containing a CID and its</span>
<span class="sd">            corresponding list of similar CIDs.</span>

<span class="sd">        Side Effects:</span>
<span class="sd">            - Utilizes concurrent threads to speed up the fetching process.</span>
<span class="sd">            - May log errors if any occur during the fetching of similar CIDs</span>
<span class="sd">            for individual CIDs.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
            <span class="n">futures</span> <span class="o">=</span> <span class="p">[</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fetch_similar_cids</span><span class="p">,</span> <span class="n">cid</span><span class="p">)</span> <span class="k">for</span> <span class="n">cid</span> <span class="ow">in</span> <span class="n">chunk</span><span class="p">]</span>
            <span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span> <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">as_completed</span><span class="p">(</span><span class="n">futures</span><span class="p">)]</span>
        <span class="k">return</span> <span class="n">results</span>


    <span class="k">def</span> <span class="nf">compound_similarity_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Identifies and records the similarity relationships between compounds</span>
<span class="sd">        based on a list of CIDs. The similarity is detrmined by the Tanimoto</span>
<span class="sd">        similarity coefficient with threshold 95% to ensure highe structural </span>
<span class="sd">        similarity.</span>

<span class="sd">        This method reads a CSV file containing compound data, filters compounds</span>
<span class="sd">        based on specific &#39;Target GeneID&#39; values,</span>
<span class="sd">        and fetches similar CIDs for each compound. The compounds are processed</span>
<span class="sd">        in chunks to manage memory usage and improve efficiency. The results are</span>
<span class="sd">        saved into separate CSV files for each chunk.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to the CSV file containing the main compound data.</span>
<span class="sd">            start_chunk (int): The starting index for processing chunks.</span>
<span class="sd">        Note:</span>
<span class="sd">            - The method filters the main data for compounds associated with</span>
<span class="sd">            specific &#39;Target GeneID&#39; values before fetching similar CIDs,</span>
<span class="sd">            optimizing the process for relevant compounds only.</span>
<span class="sd">            - The division of CIDs into chunks and concurrent processing helps</span>
<span class="sd">            in managing large datasets and utilizes parallelism for faster</span>
<span class="sd">            execution.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="mi">1576</span><span class="p">,</span> <span class="mi">1544</span><span class="p">,</span> <span class="mi">1557</span><span class="p">,</span> <span class="mi">1559</span><span class="p">,</span> <span class="mi">1565</span><span class="p">])]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">])</span>
        <span class="n">IDs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="n">chunk_size</span><span class="o">=</span><span class="mi">10000</span>
        <span class="n">chunks</span> <span class="o">=</span> <span class="p">[</span><span class="n">IDs</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="n">chunk_size</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">IDs</span><span class="p">),</span> <span class="n">chunk_size</span><span class="p">)]</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">chunks</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="n">start_chunk</span><span class="p">:</span>
                <span class="n">chunk_results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_chunk</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
                <span class="n">chunk_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">chunk_results</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">,</span> <span class="s1">&#39;Similar CIDs&#39;</span><span class="p">])</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;Data/Relationships/Compound_Similarities&#39;</span><span class="p">):</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s1">&#39;Data/Relationships/Compound_Similarities&#39;</span><span class="p">)</span>
                <span class="n">chunk_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Relationships/Compound_Similarities/Chunk_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed chunk </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> for compound similarity relationships.&quot;</span><span class="p">)</span>



    <span class="k">def</span> <span class="nf">_fetch_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches chemical-chemical and chemical-gene relationship data for a given</span>
<span class="sd">        compound ID (CID). Checks if each data file exists before fetching.</span>

<span class="sd">        Args:</span>
<span class="sd">            cid (int): The compound ID for which data is to be fetched.</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: A tuple containing the CID, and two lists of data</span>
<span class="sd">            (chemical-chemical and chemical-gene relationships).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cpd_cpd_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Data/Relationships/Cpd_Cpd_CoOcuurence/CID_</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s1">.csv&#39;</span>
        <span class="n">cpd_gene_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Data/Relationships/Cpd_gene_CoOcuurence/CID_</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s1">.csv&#39;</span>

        <span class="n">cpd_cpd_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fetch_chemical_neighbor_data</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">cpd_cpd_file</span><span class="p">)</span> <span class="k">else</span> <span class="p">[]</span>
        <span class="n">cpd_gene_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fetch_chemical_gene_data</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">cpd_gene_file</span><span class="p">)</span> <span class="k">else</span> <span class="p">[]</span>

        <span class="k">return</span> <span class="n">cid</span><span class="p">,</span> <span class="n">cpd_cpd_data</span><span class="p">,</span> <span class="n">cpd_gene_data</span>


    <span class="k">def</span> <span class="nf">_fetch_chemical_neighbor_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches chemical-chemical relationship data for a given CID.</span>

<span class="sd">        Args:</span>
<span class="sd">            cid (int): The compound ID for which data is to be fetched.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of chemical-chemical relationship data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cpd_cpd_url</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/link_db/link_db_server.cgi?format=JSON&amp;type=&quot;</span>
                       <span class="sa">f</span><span class="s2">&quot;ChemicalNeighbor&amp;operation=GetAllLinks&amp;id_1=</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">&amp;response_type=display&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_send_request</span><span class="p">(</span><span class="n">cpd_cpd_url</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
            <span class="k">return</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;LinkDataSet&#39;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;LinkData&#39;</span><span class="p">,</span> <span class="p">[])</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to fetch chemical-chemical data for CID </span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">[]</span>


    <span class="k">def</span> <span class="nf">_fetch_chemical_gene_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches chemical-gene relationship data for a given CID.</span>

<span class="sd">        Args:</span>
<span class="sd">            cid (int): The compound ID for which data is to be fetched.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of chemical-gene relationship data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cpd_gene_url</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/link_db/link_db_server.cgi?format=JSON&amp;type=&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;ChemicalGeneSymbolNeighbor&amp;operation=GetAllLinks&amp;id_1=</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">&amp;response_type=display&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_send_request</span><span class="p">(</span><span class="n">cpd_gene_url</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
            <span class="k">return</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;LinkDataSet&#39;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;LinkData&#39;</span><span class="p">,</span> <span class="p">[])</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to fetch chemical-gene data for CID </span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">[]</span>


    <span class="k">def</span> <span class="nf">_write_data_to_csv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="n">filter_condition</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes given data to a CSV file, with optional filtering before saving.</span>

<span class="sd">        This method takes a list of dictionaries (data), converts it into a</span>
<span class="sd">        pandas DataFrame, and optionally filters the DataFrame based on</span>
<span class="sd">        specified conditions before writing the result to a CSV file. The</span>
<span class="sd">        filtering is performed on specified columns with their expected</span>
<span class="sd">        values provided in &#39;filter_condition&#39;. This allows for selective</span>
<span class="sd">        data saving, especially useful when dealing with large datasets</span>
<span class="sd">        or when only a subset of data is needed for further processing</span>
<span class="sd">        or analysis.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            data (list of dict): Data to be written to a CSV file. Each</span>
<span class="sd">            dictionary in the list represents a row in the DataFrame, with keys</span>
<span class="sd">            as column names and values as row values.</span>
<span class="sd">            filename (str): Path to the CSV file where the data will be saved.</span>
<span class="sd">            If the file exists, it will be overwritten.</span>
<span class="sd">            filter_condition (dict, optional): A dictionary specifying the</span>
<span class="sd">            columns to filter by and the values to include. Keys in the</span>
<span class="sd">            dictionary are column names, and values are lists of acceptable</span>
<span class="sd">            values for that column. Rows not meeting the filter condition are</span>
<span class="sd">            excluded from the final DataFrame to be saved.</span>

<span class="sd">        Side Effects:</span>
<span class="sd">            - Writes a CSV file to the given filename path. The file is overwritten</span>
<span class="sd">            if it already exists.</span>
<span class="sd">            - Logs a warning if a specified column for filtering is not found in</span>
<span class="sd">            the DataFrame.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">filter_condition</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">column</span><span class="p">,</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">filter_condition</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">values</span><span class="p">)]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Column </span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2"> not found in DataFrame.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">compound_cooccurrence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">,</span> <span class="n">rate_limit</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Analyzes compound co-occurrence relationships from the specified main</span>
<span class="sd">        data file and saves the results into structured CSV files.</span>

<span class="sd">        This method takes a path to a CSV file containing compound data and</span>
<span class="sd">        performs batch processing to extract relationships between compounds</span>
<span class="sd">        and genes from the PubChem database. It filters compounds based on their</span>
<span class="sd">        association with specific genes of interest, then fetches co-occurrence</span>
<span class="sd">        data for each compound using parallel requests. The data fetched</span>
<span class="sd">        includes both compound-compound and compound-gene co-occurrence</span>
<span class="sd">        relationships. Results are saved in separate CSV files within specific</span>
<span class="sd">        directories for later analysis.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            main_data (str): Path to the CSV file containing the main data. This</span>
<span class="sd">            file should include &#39;CID&#39; (Compound ID) and &#39;Target GeneID&#39; columns.</span>
<span class="sd">            rate_limit (int): Controls the rate of API requests to avoid</span>
<span class="sd">            exceeding PubChem&#39;s request limits. Specifies the maximum number of</span>
<span class="sd">            requests that can be made per second.</span>
<span class="sd">            start_chunk (int): The starting index for processing chunks.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: A message indicating the successful completion of data</span>
<span class="sd">            processing and saving.</span>

<span class="sd">        Raises:</span>
<span class="sd">            FileNotFoundError: If the specified &#39;main_data&#39; file does not exist</span>
<span class="sd">            or cannot be read.</span>
<span class="sd">            ValueError: If &#39;main_data&#39; does not contain the required columns</span>
<span class="sd">            (&#39;CID&#39; and &#39;Target GeneID&#39;).</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; extractor = RelationshipPropertiesExtractor()</span>
<span class="sd">            &gt;&gt;&gt; completion_message = extractor.compound_cooccurrence(&#39;Data/AllDataConnected.csv&#39;, rate_limit=5)</span>
<span class="sd">            &gt;&gt;&gt; print(completion_message)</span>
<span class="sd">            This would process the compound data, fetch co-occurrence data from</span>
<span class="sd">            PubChem, and save the results into CSV files.</span>
<span class="sd">            The completion message would indicate successful processing.</span>

<span class="sd">        Note:</span>
<span class="sd">            The &#39;main_data&#39; file must be properly formatted, with at least &#39;CID&#39;</span>
<span class="sd">            and &#39;Target GeneID&#39; columns present. The method assumes the existence</span>
<span class="sd">            of &#39;Data/Relationships/Cpd_Cpd_CoOcuurence&#39; and</span>
<span class="sd">            &#39;Data/Relationships/Cpd_gene_CoOcuurence&#39; directories for saving</span>
<span class="sd">            the output CSV files. It is recommended to check and adhere to</span>
<span class="sd">            PubChem&#39;s current rate limits when setting the &#39;rate_limit&#39;</span>
<span class="sd">            parameter to avoid potential blocks or restrictions on your</span>
<span class="sd">            IP address due to excessive requests.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">3000</span><span class="p">)</span>  <span class="c1"># Reading in chunks for large files</span>
        <span class="k">for</span> <span class="n">chunk_idx</span><span class="p">,</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">chunk_idx</span> <span class="o">&gt;=</span> <span class="n">start_chunk</span><span class="p">:</span>
                <span class="n">chunk</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="n">chunk</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="mi">1576</span><span class="p">,</span> <span class="mi">1544</span><span class="p">,</span> <span class="mi">1557</span><span class="p">,</span> <span class="mi">1559</span><span class="p">,</span> <span class="mi">1565</span><span class="p">])]</span>
                <span class="n">chunk</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">IDs</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

                <span class="n">start_time</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
                <span class="k">with</span> <span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="n">rate_limit</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
                    <span class="n">futures</span> <span class="o">=</span> <span class="p">{</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_fetch_data</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">cid</span><span class="p">)):</span> <span class="n">cid</span> <span class="k">for</span> <span class="n">cid</span> <span class="ow">in</span> <span class="n">IDs</span><span class="p">}</span>
                    <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">as_completed</span><span class="p">(</span><span class="n">futures</span><span class="p">):</span>
                        <span class="n">cid</span><span class="p">,</span> <span class="n">cpd_cpd_data</span><span class="p">,</span> <span class="n">cpd_gene_data</span> <span class="o">=</span> <span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_write_data_to_csv</span><span class="p">(</span><span class="n">cpd_cpd_data</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Data/Relationships/Cpd_Cpd_CoOcuurence/CID_</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_write_data_to_csv</span><span class="p">(</span><span class="n">cpd_gene_data</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Data/Relationships/Cpd_gene_CoOcuurence/CID_</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span>
                                                <span class="n">filter_condition</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;ID_2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp3a4&#39;}&quot;</span><span class="p">,</span> <span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp1a2&#39;}&quot;</span><span class="p">,</span>
                                                                           <span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp2c9&#39;}&quot;</span><span class="p">,</span> <span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp2c19&#39;}&quot;</span><span class="p">,</span>
                                                                           <span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp2d6&#39;}&quot;</span><span class="p">]})</span>
                        <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">rate_limit</span><span class="p">)</span>  <span class="c1"># Ensuring we don&#39;t exceed rate limit</span>
                <span class="n">elapsed</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed chunk </span><span class="si">{</span><span class="n">chunk_idx</span><span class="si">}</span><span class="s2"> in </span><span class="si">{</span><span class="n">elapsed</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> seconds&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="s2">&quot;Data fetching and saving completed.&quot;</span>


    <span class="k">def</span> <span class="nf">compound_transformation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_properties</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Analyzes compound transformation data based on gene properties, focusing</span>
<span class="sd">        on metabolic transformations involving specified genes. This method</span>
<span class="sd">        queries the PubChem database for transformation data related</span>
<span class="sd">        to compounds associated with the genes identified in the provided CSV file.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            gene_properties (str): Path to the CSV file containing gene properties</span>
<span class="sd">            generated by the NodePropertiesExtractor class, which should include</span>
<span class="sd">            &#39;GeneID&#39; as one of its columns. This file is used to identify genes</span>
<span class="sd">            of interest for which compound transformation data will be fetched.</span>

<span class="sd">        Processing Steps:</span>
<span class="sd">            1. Reads the provided CSV file to extract unique gene identifiers.</span>
<span class="sd">            2. For each gene identifier, constructs a query to fetch relevant</span>
<span class="sd">            compound transformation data from PubChem, focusing on metabolic</span>
<span class="sd">            transformations where the gene plays a role.</span>
<span class="sd">            3. Processes and aggregates the fetched data into a structured</span>
<span class="sd">            pandas DataFrame.</span>
<span class="sd">            4. Filters the aggregated data to retain specific columns relevant</span>
<span class="sd">            to compound transformations, including substrate and metabolite</span>
<span class="sd">            Compound IDs (CIDs), the type of metabolic conversion, gene</span>
<span class="sd">            identifiers, PubMed IDs, and DOIs for related publications.</span>
<span class="sd">            5. Saves the aggregated and filtered DataFrame to a CSV file for</span>
<span class="sd">            further analysis or integration into knowledge graphs or other</span>
<span class="sd">            data models.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pandas.DataFrame: A DataFrame containing processed compound</span>
<span class="sd">            transformation data, including substrate and metabolite CIDs,</span>
<span class="sd">            metabolic conversion types, gene identifiers, PubMed IDs, and DOIs.</span>
<span class="sd">            The DataFrame structure facilitates further analysis or use in</span>
<span class="sd">            constructing detailed views of metabolic pathways involving the</span>
<span class="sd">            specified genes.</span>

<span class="sd">        Side Effects:</span>
<span class="sd">            - Saves the aggregated compound transformation data to</span>
<span class="sd">            &#39;Data/Relationships/Compound_Transformation.csv&#39;</span>
<span class="sd">            in the current working directory. This file captures the relationship</span>
<span class="sd">            between substrates, metabolites, and genes based on the input gene</span>
<span class="sd">            properties.</span>

<span class="sd">        Raises:</span>
<span class="sd">            FileNotFoundError: If the specified &#39;gene_properties&#39; file does not</span>
<span class="sd">            exist or cannot be read.</span>
<span class="sd">            ValueError: If &#39;gene_properties&#39; does not contain the required</span>
<span class="sd">            &#39;GeneID&#39; column.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; extractor = RelationshipPropertiesExtractor()</span>
<span class="sd">            &gt;&gt;&gt; transformation_df = extractor.compound_transformation(&#39;Data/Nodes/gene_properties.csv&#39;)</span>
<span class="sd">            &gt;&gt;&gt; print(transformation_df.head())</span>
<span class="sd">            This example processes gene properties from</span>
<span class="sd">            &#39;path/to/gene_properties.csv&#39;, queries PubChem for</span>
<span class="sd">            compound transformation data related to the genes,</span>
<span class="sd">            and compiles the results into a DataFrame.</span>

<span class="sd">        Note:</span>
<span class="sd">            The method assumes that the input &#39;gene_properties&#39; file is</span>
<span class="sd">            accessible and correctly formatted.</span>
<span class="sd">            The availability and structure of the PubChem database may affect</span>
<span class="sd">            the completeness and accuracy of the fetched transformation data.</span>
<span class="sd">            Users should verify the existence of the &#39;Data/Relationships&#39;</span>
<span class="sd">            directory and have appropriate permissions to write files to it.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_properties</span><span class="p">)</span>
        <span class="n">IDs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="n">transformation_dfs</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">gid</span> <span class="ow">in</span> <span class="n">IDs</span><span class="p">:</span>
            <span class="n">gid</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gid</span><span class="p">)</span>
            <span class="n">url</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/sdq/sdqagent.cgi?infmt=json&amp;outfmt=csv&quot;</span>
                   <span class="s2">&quot;&amp;query={{</span><span class="se">\&quot;</span><span class="s2">download</span><span class="se">\&quot;</span><span class="s2">:</span><span class="se">\&quot;</span><span class="s2">*</span><span class="se">\&quot;</span><span class="s2">,</span><span class="se">\&quot;</span><span class="s2">collection</span><span class="se">\&quot;</span><span class="s2">:</span><span class="se">\&quot;</span><span class="s2">chemblmetabolism</span><span class="se">\&quot;</span><span class="s2">,</span><span class="se">\&quot;</span><span class="s2">where</span><span class="se">\&quot;</span><span class="s2">:&quot;</span>
                   <span class="sa">f</span><span class="s2">&quot;</span><span class="se">{{\&quot;</span><span class="s2">ands</span><span class="se">\&quot;</span><span class="s2">:[</span><span class="se">{{\&quot;</span><span class="s2">geneid</span><span class="se">\&quot;</span><span class="s2">:</span><span class="se">\&quot;</span><span class="si">{</span><span class="n">gid</span><span class="si">}</span><span class="se">\&quot;}}</span><span class="s2">]</span><span class="se">}}</span><span class="s2">,</span><span class="se">\&quot;</span><span class="s2">order</span><span class="se">\&quot;</span><span class="s2">:[</span><span class="se">\&quot;</span><span class="s2">relevancescore,desc</span><span class="se">\&quot;</span><span class="s2">]&quot;</span>
                   <span class="sa">f</span><span class="s2">&quot;,</span><span class="se">\&quot;</span><span class="s2">start</span><span class="se">\&quot;</span><span class="s2">:1,</span><span class="se">\&quot;</span><span class="s2">limit</span><span class="se">\&quot;</span><span class="s2">:10000000,</span><span class="se">\&quot;</span><span class="s2">downloadfilename</span><span class="se">\&quot;</span><span class="s2">:</span><span class="se">\&quot;</span><span class="s2">pubchem_geneid_</span><span class="si">{</span><span class="n">gid</span><span class="si">}</span><span class="s2">_chemblmetabolism</span><span class="se">\&quot;}}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_send_request</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">response</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">transformation_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">),</span>
                                                    <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                                    <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">transformation_df</span> <span class="o">=</span> <span class="n">transformation_df</span><span class="p">[[</span><span class="s1">&#39;substratecid&#39;</span><span class="p">,</span>
                                                           <span class="s1">&#39;metabolitecid&#39;</span><span class="p">,</span>
                                                           <span class="s1">&#39;metconversion&#39;</span><span class="p">,</span>
                                                           <span class="s1">&#39;geneids&#39;</span><span class="p">,</span>
                                                           <span class="s1">&#39;pmids&#39;</span><span class="p">,</span>
                                                           <span class="s1">&#39;dois&#39;</span><span class="p">]]</span>
                    <span class="n">transformation_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transformation_df</span><span class="p">)</span>
                <span class="k">except</span> <span class="n">pd</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">ParserError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error parsing CSV for gene ID </span><span class="si">{</span><span class="n">gid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">continue</span>  <span class="c1"># Skip this gene ID and continue with others</span>

        <span class="n">transformation_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">transformation_dfs</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">if</span> <span class="n">transformation_dfs</span> <span class="k">else</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_write_to_csv</span><span class="p">(</span><span class="n">transformation_df</span><span class="p">,</span>
                           <span class="s1">&#39;Data/Relationships/Compound_Transformation.csv&#39;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">transformation_df</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Initializes a RelationshipPropertiesExtractor with a Requests session
for efficient network calls.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initializes a RelationshipPropertiesExtractor with a Requests session</span>
<span class="sd">     for efficient network calls.&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">session</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">Session</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.assay_compound_relationship" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">assay_compound_relationship</span><span class="p">(</span><span class="n">assays_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Processes and stores relationships between assays and compounds based
on assay data from PubChem.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>assays_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to a CSV file containing assay IDs (AIDs).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>start_chunk</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The starting index for processing chunks.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">assay_compound_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assays_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes and stores relationships between assays and compounds based</span>
<span class="sd">    on assay data from PubChem.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        assays_data (str): Path to a CSV file containing assay IDs (AIDs).</span>
<span class="sd">        start_chunk (int): The starting index for processing chunks.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">chunk_idx</span><span class="p">,</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">assays_data</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">100</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">chunk_idx</span> <span class="o">&gt;=</span> <span class="n">start_chunk</span><span class="p">:</span>
            <span class="n">columns_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PUBCHEM_RESULT_TAG&#39;</span><span class="p">,</span> <span class="s1">&#39;PUBCHEM_SID&#39;</span><span class="p">,</span> <span class="s1">&#39;PUBCHEM_EXT_DATASOURCE_SMILES&#39;</span><span class="p">]</span>
            <span class="n">output_dir</span> <span class="o">=</span> <span class="s1">&#39;Data/Relationships/Assay_Compound_Relationship&#39;</span>

            <span class="k">for</span> <span class="n">aid</span> <span class="ow">in</span> <span class="n">chunk</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s1">/AID_</span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">):</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fetch_data_for_aid</span><span class="p">(</span><span class="n">aid</span><span class="p">,</span> <span class="n">columns_to_remove</span><span class="p">)</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output_dir</span><span class="p">):</span>
                            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
                        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s1">/AID_</span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed chunk </span><span class="si">{</span><span class="n">chunk_idx</span><span class="si">}</span><span class="s2"> for assay-compound relationships.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No More Chunck to Process.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.assay_enzyme_relationship" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">assay_enzyme_relationship</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Extracts and saves relationships between assays and enzymes from the
specified dataset.</p>
<p>This method processes assay data to identify relationships between
assays and their target enzymes. It selects relevant columns from the
input data, removes duplicates to ensure unique relationships, and saves
the cleaned data to a CSV file for further analysis or integration into
knowledge graphs.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the CSV file containing the main data. The</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pandas.DataFrame: A DataFrame containing the unique relationships</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>between assays and enzymes, including the assay ID, target gene ID,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>and activity name.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="side-effects" open>
  <summary>Side Effects</summary>
  <ul>
<li>Writes a CSV file to 'Data/Assay_Enzyme_Relationship.csv',
containing the processed relationships data.</li>
</ul>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">assay_enzyme_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts and saves relationships between assays and enzymes from the</span>
<span class="sd">    specified dataset.</span>

<span class="sd">    This method processes assay data to identify relationships between</span>
<span class="sd">    assays and their target enzymes. It selects relevant columns from the</span>
<span class="sd">    input data, removes duplicates to ensure unique relationships, and saves</span>
<span class="sd">    the cleaned data to a CSV file for further analysis or integration into</span>
<span class="sd">    knowledge graphs.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to the CSV file containing the main data. The</span>
<span class="sd">        file should include columns for &#39;AID&#39; (Assay ID), &#39;Target GeneID&#39;,</span>
<span class="sd">        and &#39;Activity Name&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: A DataFrame containing the unique relationships</span>
<span class="sd">        between assays and enzymes, including the assay ID, target gene ID,</span>
<span class="sd">        and activity name.</span>

<span class="sd">    Side Effects:</span>
<span class="sd">        - Writes a CSV file to &#39;Data/Assay_Enzyme_Relationship.csv&#39;,</span>
<span class="sd">        containing the processed relationships data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
    <span class="n">columns_to_select</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Activity Name&#39;</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">columns_to_select</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Assay_Enzyme_Relationship.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_cooccurrence" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compound_cooccurrence</span><span class="p">(</span><span class="n">main_data</span><span class="p">,</span> <span class="n">rate_limit</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Analyzes compound co-occurrence relationships from the specified main
data file and saves the results into structured CSV files.</p>
<p>This method takes a path to a CSV file containing compound data and
performs batch processing to extract relationships between compounds
and genes from the PubChem database. It filters compounds based on their
association with specific genes of interest, then fetches co-occurrence
data for each compound using parallel requests. The data fetched
includes both compound-compound and compound-gene co-occurrence
relationships. Results are saved in separate CSV files within specific
directories for later analysis.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the CSV file containing the main data. This</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>rate_limit</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Controls the rate of API requests to avoid</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>start_chunk</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The starting index for processing chunks.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A message indicating the successful completion of data</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>processing and saving.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>FileNotFoundError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the specified 'main_data' file does not exist</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If 'main_data' does not contain the required columns</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = RelationshipPropertiesExtractor()
completion_message = extractor.compound_cooccurrence('Data/AllDataConnected.csv', rate_limit=5)
print(completion_message)
This would process the compound data, fetch co-occurrence data from
PubChem, and save the results into CSV files.
The completion message would indicate successful processing.</p>
</blockquote>
</blockquote>
</blockquote>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>The 'main_data' file must be properly formatted, with at least 'CID'
and 'Target GeneID' columns present. The method assumes the existence
of 'Data/Relationships/Cpd_Cpd_CoOcuurence' and
'Data/Relationships/Cpd_gene_CoOcuurence' directories for saving
the output CSV files. It is recommended to check and adhere to
PubChem's current rate limits when setting the 'rate_limit'
parameter to avoid potential blocks or restrictions on your
IP address due to excessive requests.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compound_cooccurrence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">,</span> <span class="n">rate_limit</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analyzes compound co-occurrence relationships from the specified main</span>
<span class="sd">    data file and saves the results into structured CSV files.</span>

<span class="sd">    This method takes a path to a CSV file containing compound data and</span>
<span class="sd">    performs batch processing to extract relationships between compounds</span>
<span class="sd">    and genes from the PubChem database. It filters compounds based on their</span>
<span class="sd">    association with specific genes of interest, then fetches co-occurrence</span>
<span class="sd">    data for each compound using parallel requests. The data fetched</span>
<span class="sd">    includes both compound-compound and compound-gene co-occurrence</span>
<span class="sd">    relationships. Results are saved in separate CSV files within specific</span>
<span class="sd">    directories for later analysis.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to the CSV file containing the main data. This</span>
<span class="sd">        file should include &#39;CID&#39; (Compound ID) and &#39;Target GeneID&#39; columns.</span>
<span class="sd">        rate_limit (int): Controls the rate of API requests to avoid</span>
<span class="sd">        exceeding PubChem&#39;s request limits. Specifies the maximum number of</span>
<span class="sd">        requests that can be made per second.</span>
<span class="sd">        start_chunk (int): The starting index for processing chunks.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: A message indicating the successful completion of data</span>
<span class="sd">        processing and saving.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the specified &#39;main_data&#39; file does not exist</span>
<span class="sd">        or cannot be read.</span>
<span class="sd">        ValueError: If &#39;main_data&#39; does not contain the required columns</span>
<span class="sd">        (&#39;CID&#39; and &#39;Target GeneID&#39;).</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; extractor = RelationshipPropertiesExtractor()</span>
<span class="sd">        &gt;&gt;&gt; completion_message = extractor.compound_cooccurrence(&#39;Data/AllDataConnected.csv&#39;, rate_limit=5)</span>
<span class="sd">        &gt;&gt;&gt; print(completion_message)</span>
<span class="sd">        This would process the compound data, fetch co-occurrence data from</span>
<span class="sd">        PubChem, and save the results into CSV files.</span>
<span class="sd">        The completion message would indicate successful processing.</span>

<span class="sd">    Note:</span>
<span class="sd">        The &#39;main_data&#39; file must be properly formatted, with at least &#39;CID&#39;</span>
<span class="sd">        and &#39;Target GeneID&#39; columns present. The method assumes the existence</span>
<span class="sd">        of &#39;Data/Relationships/Cpd_Cpd_CoOcuurence&#39; and</span>
<span class="sd">        &#39;Data/Relationships/Cpd_gene_CoOcuurence&#39; directories for saving</span>
<span class="sd">        the output CSV files. It is recommended to check and adhere to</span>
<span class="sd">        PubChem&#39;s current rate limits when setting the &#39;rate_limit&#39;</span>
<span class="sd">        parameter to avoid potential blocks or restrictions on your</span>
<span class="sd">        IP address due to excessive requests.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">3000</span><span class="p">)</span>  <span class="c1"># Reading in chunks for large files</span>
    <span class="k">for</span> <span class="n">chunk_idx</span><span class="p">,</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">chunk_idx</span> <span class="o">&gt;=</span> <span class="n">start_chunk</span><span class="p">:</span>
            <span class="n">chunk</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="n">chunk</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="mi">1576</span><span class="p">,</span> <span class="mi">1544</span><span class="p">,</span> <span class="mi">1557</span><span class="p">,</span> <span class="mi">1559</span><span class="p">,</span> <span class="mi">1565</span><span class="p">])]</span>
            <span class="n">chunk</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">IDs</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

            <span class="n">start_time</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
            <span class="k">with</span> <span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="n">rate_limit</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
                <span class="n">futures</span> <span class="o">=</span> <span class="p">{</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_fetch_data</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">cid</span><span class="p">)):</span> <span class="n">cid</span> <span class="k">for</span> <span class="n">cid</span> <span class="ow">in</span> <span class="n">IDs</span><span class="p">}</span>
                <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">as_completed</span><span class="p">(</span><span class="n">futures</span><span class="p">):</span>
                    <span class="n">cid</span><span class="p">,</span> <span class="n">cpd_cpd_data</span><span class="p">,</span> <span class="n">cpd_gene_data</span> <span class="o">=</span> <span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_write_data_to_csv</span><span class="p">(</span><span class="n">cpd_cpd_data</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Data/Relationships/Cpd_Cpd_CoOcuurence/CID_</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_write_data_to_csv</span><span class="p">(</span><span class="n">cpd_gene_data</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Data/Relationships/Cpd_gene_CoOcuurence/CID_</span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span>
                                            <span class="n">filter_condition</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;ID_2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp3a4&#39;}&quot;</span><span class="p">,</span> <span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp1a2&#39;}&quot;</span><span class="p">,</span>
                                                                       <span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp2c9&#39;}&quot;</span><span class="p">,</span> <span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp2c19&#39;}&quot;</span><span class="p">,</span>
                                                                       <span class="s2">&quot;{&#39;GeneSymbol&#39;: &#39;cyp2d6&#39;}&quot;</span><span class="p">]})</span>
                    <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">rate_limit</span><span class="p">)</span>  <span class="c1"># Ensuring we don&#39;t exceed rate limit</span>
            <span class="n">elapsed</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed chunk </span><span class="si">{</span><span class="n">chunk_idx</span><span class="si">}</span><span class="s2"> in </span><span class="si">{</span><span class="n">elapsed</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> seconds&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="s2">&quot;Data fetching and saving completed.&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_gene_relationship" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compound_gene_relationship</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Identifies and records relationships between compounds and enzymes from
the input data.</p>
<p>This method focuses on extracting compound-enzyme interaction data,
including activity outcomes and values. It selects
pertinent columns, removes duplicate records, and sorts the data by
Compound ID and Target Accession for clarity. The cleaned dataset is
then saved to a CSV file, providing a structured view  of how compounds
interact with various enzymes, which can be critical for drug discovery
and pharmacological research.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the CSV file with compound and enzyme data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pandas.DataFrame: A DataFrame with processed compound-enzyme</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>relationships, sorted and cleaned for direct analysis or database</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>insertion.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="side-effects" open>
  <summary>Side Effects</summary>
  <ul>
<li>Saves the processed relationships data to
'Data/Relationships/Compound_Gene_Relationship.csv',
facilitating easy access and integration.</li>
</ul>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compound_gene_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Identifies and records relationships between compounds and enzymes from</span>
<span class="sd">    the input data.</span>

<span class="sd">    This method focuses on extracting compound-enzyme interaction data,</span>
<span class="sd">    including activity outcomes and values. It selects</span>
<span class="sd">    pertinent columns, removes duplicate records, and sorts the data by</span>
<span class="sd">    Compound ID and Target Accession for clarity. The cleaned dataset is</span>
<span class="sd">    then saved to a CSV file, providing a structured view  of how compounds</span>
<span class="sd">    interact with various enzymes, which can be critical for drug discovery</span>
<span class="sd">    and pharmacological research.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to the CSV file with compound and enzyme data.</span>
<span class="sd">        This file should contain columns for &#39;CID&#39; (Compound ID),</span>
<span class="sd">        &#39;Target Accession&#39;, &#39;Activity Outcome&#39;, &#39;Activity Name&#39;, and</span>
<span class="sd">        &#39;Activity Value [uM]&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: A DataFrame with processed compound-enzyme</span>
<span class="sd">        relationships, sorted and cleaned for direct analysis or database</span>
<span class="sd">        insertion.</span>

<span class="sd">    Side Effects:</span>
<span class="sd">        - Saves the processed relationships data to</span>
<span class="sd">        &#39;Data/Relationships/Compound_Gene_Relationship.csv&#39;,</span>
<span class="sd">        facilitating easy access and integration.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
    <span class="n">columns_to_select</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Accession&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;Activity Outcome&#39;</span><span class="p">,</span> <span class="s1">&#39;Activity Name&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;Activity Value [uM]&#39;</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">columns_to_select</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;CID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Accession&#39;</span><span class="p">])</span>
    <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span> <span class="p">,</span> <span class="n">thresh</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1">###</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Relationships/Compound_Gene_Relationship.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_similarity_relationship" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compound_similarity_relationship</span><span class="p">(</span><span class="n">main_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Identifies and records the similarity relationships between compounds
based on a list of CIDs. The similarity is detrmined by the Tanimoto
similarity coefficient with threshold 95% to ensure highe structural 
similarity.</p>
<p>This method reads a CSV file containing compound data, filters compounds
based on specific 'Target GeneID' values,
and fetches similar CIDs for each compound. The compounds are processed
in chunks to manage memory usage and improve efficiency. The results are
saved into separate CSV files for each chunk.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the CSV file containing the main compound data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>start_chunk</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The starting index for processing chunks.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
      </tbody>
    </table>
      <p>Note:
    - The method filters the main data for compounds associated with
    specific 'Target GeneID' values before fetching similar CIDs,
    optimizing the process for relevant compounds only.
    - The division of CIDs into chunks and concurrent processing helps
    in managing large datasets and utilizes parallelism for faster
    execution.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compound_similarity_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">,</span> <span class="n">start_chunk</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Identifies and records the similarity relationships between compounds</span>
<span class="sd">    based on a list of CIDs. The similarity is detrmined by the Tanimoto</span>
<span class="sd">    similarity coefficient with threshold 95% to ensure highe structural </span>
<span class="sd">    similarity.</span>

<span class="sd">    This method reads a CSV file containing compound data, filters compounds</span>
<span class="sd">    based on specific &#39;Target GeneID&#39; values,</span>
<span class="sd">    and fetches similar CIDs for each compound. The compounds are processed</span>
<span class="sd">    in chunks to manage memory usage and improve efficiency. The results are</span>
<span class="sd">    saved into separate CSV files for each chunk.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to the CSV file containing the main compound data.</span>
<span class="sd">        start_chunk (int): The starting index for processing chunks.</span>
<span class="sd">    Note:</span>
<span class="sd">        - The method filters the main data for compounds associated with</span>
<span class="sd">        specific &#39;Target GeneID&#39; values before fetching similar CIDs,</span>
<span class="sd">        optimizing the process for relevant compounds only.</span>
<span class="sd">        - The division of CIDs into chunks and concurrent processing helps</span>
<span class="sd">        in managing large datasets and utilizes parallelism for faster</span>
<span class="sd">        execution.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="mi">1576</span><span class="p">,</span> <span class="mi">1544</span><span class="p">,</span> <span class="mi">1557</span><span class="p">,</span> <span class="mi">1559</span><span class="p">,</span> <span class="mi">1565</span><span class="p">])]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">])</span>
    <span class="n">IDs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="n">chunk_size</span><span class="o">=</span><span class="mi">10000</span>
    <span class="n">chunks</span> <span class="o">=</span> <span class="p">[</span><span class="n">IDs</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="n">chunk_size</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">IDs</span><span class="p">),</span> <span class="n">chunk_size</span><span class="p">)]</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">chunks</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="n">start_chunk</span><span class="p">:</span>
            <span class="n">chunk_results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_chunk</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
            <span class="n">chunk_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">chunk_results</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CID&#39;</span><span class="p">,</span> <span class="s1">&#39;Similar CIDs&#39;</span><span class="p">])</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;Data/Relationships/Compound_Similarities&#39;</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s1">&#39;Data/Relationships/Compound_Similarities&#39;</span><span class="p">)</span>
            <span class="n">chunk_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Relationships/Compound_Similarities/Chunk_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed chunk </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> for compound similarity relationships.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.compound_transformation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compound_transformation</span><span class="p">(</span><span class="n">gene_properties</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Analyzes compound transformation data based on gene properties, focusing
on metabolic transformations involving specified genes. This method
queries the PubChem database for transformation data related
to compounds associated with the genes identified in the provided CSV file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>gene_properties</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the CSV file containing gene properties</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<details class="processing-steps" open>
  <summary>Processing Steps</summary>
  <ol>
<li>Reads the provided CSV file to extract unique gene identifiers.</li>
<li>For each gene identifier, constructs a query to fetch relevant
compound transformation data from PubChem, focusing on metabolic
transformations where the gene plays a role.</li>
<li>Processes and aggregates the fetched data into a structured
pandas DataFrame.</li>
<li>Filters the aggregated data to retain specific columns relevant
to compound transformations, including substrate and metabolite
Compound IDs (CIDs), the type of metabolic conversion, gene
identifiers, PubMed IDs, and DOIs for related publications.</li>
<li>Saves the aggregated and filtered DataFrame to a CSV file for
further analysis or integration into knowledge graphs or other
data models.</li>
</ol>
</details>

    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pandas.DataFrame: A DataFrame containing processed compound</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>transformation data, including substrate and metabolite CIDs,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>metabolic conversion types, gene identifiers, PubMed IDs, and DOIs.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The DataFrame structure facilitates further analysis or use in</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>constructing detailed views of metabolic pathways involving the</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>specified genes.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="side-effects" open>
  <summary>Side Effects</summary>
  <ul>
<li>Saves the aggregated compound transformation data to
'Data/Relationships/Compound_Transformation.csv'
in the current working directory. This file captures the relationship
between substrates, metabolites, and genes based on the input gene
properties.</li>
</ul>
</details>

<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>FileNotFoundError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the specified 'gene_properties' file does not</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If 'gene_properties' does not contain the required</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = RelationshipPropertiesExtractor()
transformation_df = extractor.compound_transformation('Data/Nodes/gene_properties.csv')
print(transformation_df.head())
This example processes gene properties from
'path/to/gene_properties.csv', queries PubChem for
compound transformation data related to the genes,
and compiles the results into a DataFrame.</p>
</blockquote>
</blockquote>
</blockquote>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>The method assumes that the input 'gene_properties' file is
accessible and correctly formatted.
The availability and structure of the PubChem database may affect
the completeness and accuracy of the fetched transformation data.
Users should verify the existence of the 'Data/Relationships'
directory and have appropriate permissions to write files to it.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compound_transformation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_properties</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analyzes compound transformation data based on gene properties, focusing</span>
<span class="sd">    on metabolic transformations involving specified genes. This method</span>
<span class="sd">    queries the PubChem database for transformation data related</span>
<span class="sd">    to compounds associated with the genes identified in the provided CSV file.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        gene_properties (str): Path to the CSV file containing gene properties</span>
<span class="sd">        generated by the NodePropertiesExtractor class, which should include</span>
<span class="sd">        &#39;GeneID&#39; as one of its columns. This file is used to identify genes</span>
<span class="sd">        of interest for which compound transformation data will be fetched.</span>

<span class="sd">    Processing Steps:</span>
<span class="sd">        1. Reads the provided CSV file to extract unique gene identifiers.</span>
<span class="sd">        2. For each gene identifier, constructs a query to fetch relevant</span>
<span class="sd">        compound transformation data from PubChem, focusing on metabolic</span>
<span class="sd">        transformations where the gene plays a role.</span>
<span class="sd">        3. Processes and aggregates the fetched data into a structured</span>
<span class="sd">        pandas DataFrame.</span>
<span class="sd">        4. Filters the aggregated data to retain specific columns relevant</span>
<span class="sd">        to compound transformations, including substrate and metabolite</span>
<span class="sd">        Compound IDs (CIDs), the type of metabolic conversion, gene</span>
<span class="sd">        identifiers, PubMed IDs, and DOIs for related publications.</span>
<span class="sd">        5. Saves the aggregated and filtered DataFrame to a CSV file for</span>
<span class="sd">        further analysis or integration into knowledge graphs or other</span>
<span class="sd">        data models.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: A DataFrame containing processed compound</span>
<span class="sd">        transformation data, including substrate and metabolite CIDs,</span>
<span class="sd">        metabolic conversion types, gene identifiers, PubMed IDs, and DOIs.</span>
<span class="sd">        The DataFrame structure facilitates further analysis or use in</span>
<span class="sd">        constructing detailed views of metabolic pathways involving the</span>
<span class="sd">        specified genes.</span>

<span class="sd">    Side Effects:</span>
<span class="sd">        - Saves the aggregated compound transformation data to</span>
<span class="sd">        &#39;Data/Relationships/Compound_Transformation.csv&#39;</span>
<span class="sd">        in the current working directory. This file captures the relationship</span>
<span class="sd">        between substrates, metabolites, and genes based on the input gene</span>
<span class="sd">        properties.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the specified &#39;gene_properties&#39; file does not</span>
<span class="sd">        exist or cannot be read.</span>
<span class="sd">        ValueError: If &#39;gene_properties&#39; does not contain the required</span>
<span class="sd">        &#39;GeneID&#39; column.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; extractor = RelationshipPropertiesExtractor()</span>
<span class="sd">        &gt;&gt;&gt; transformation_df = extractor.compound_transformation(&#39;Data/Nodes/gene_properties.csv&#39;)</span>
<span class="sd">        &gt;&gt;&gt; print(transformation_df.head())</span>
<span class="sd">        This example processes gene properties from</span>
<span class="sd">        &#39;path/to/gene_properties.csv&#39;, queries PubChem for</span>
<span class="sd">        compound transformation data related to the genes,</span>
<span class="sd">        and compiles the results into a DataFrame.</span>

<span class="sd">    Note:</span>
<span class="sd">        The method assumes that the input &#39;gene_properties&#39; file is</span>
<span class="sd">        accessible and correctly formatted.</span>
<span class="sd">        The availability and structure of the PubChem database may affect</span>
<span class="sd">        the completeness and accuracy of the fetched transformation data.</span>
<span class="sd">        Users should verify the existence of the &#39;Data/Relationships&#39;</span>
<span class="sd">        directory and have appropriate permissions to write files to it.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_properties</span><span class="p">)</span>
    <span class="n">IDs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="n">transformation_dfs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">gid</span> <span class="ow">in</span> <span class="n">IDs</span><span class="p">:</span>
        <span class="n">gid</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gid</span><span class="p">)</span>
        <span class="n">url</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/sdq/sdqagent.cgi?infmt=json&amp;outfmt=csv&quot;</span>
               <span class="s2">&quot;&amp;query={{</span><span class="se">\&quot;</span><span class="s2">download</span><span class="se">\&quot;</span><span class="s2">:</span><span class="se">\&quot;</span><span class="s2">*</span><span class="se">\&quot;</span><span class="s2">,</span><span class="se">\&quot;</span><span class="s2">collection</span><span class="se">\&quot;</span><span class="s2">:</span><span class="se">\&quot;</span><span class="s2">chemblmetabolism</span><span class="se">\&quot;</span><span class="s2">,</span><span class="se">\&quot;</span><span class="s2">where</span><span class="se">\&quot;</span><span class="s2">:&quot;</span>
               <span class="sa">f</span><span class="s2">&quot;</span><span class="se">{{\&quot;</span><span class="s2">ands</span><span class="se">\&quot;</span><span class="s2">:[</span><span class="se">{{\&quot;</span><span class="s2">geneid</span><span class="se">\&quot;</span><span class="s2">:</span><span class="se">\&quot;</span><span class="si">{</span><span class="n">gid</span><span class="si">}</span><span class="se">\&quot;}}</span><span class="s2">]</span><span class="se">}}</span><span class="s2">,</span><span class="se">\&quot;</span><span class="s2">order</span><span class="se">\&quot;</span><span class="s2">:[</span><span class="se">\&quot;</span><span class="s2">relevancescore,desc</span><span class="se">\&quot;</span><span class="s2">]&quot;</span>
               <span class="sa">f</span><span class="s2">&quot;,</span><span class="se">\&quot;</span><span class="s2">start</span><span class="se">\&quot;</span><span class="s2">:1,</span><span class="se">\&quot;</span><span class="s2">limit</span><span class="se">\&quot;</span><span class="s2">:10000000,</span><span class="se">\&quot;</span><span class="s2">downloadfilename</span><span class="se">\&quot;</span><span class="s2">:</span><span class="se">\&quot;</span><span class="s2">pubchem_geneid_</span><span class="si">{</span><span class="n">gid</span><span class="si">}</span><span class="s2">_chemblmetabolism</span><span class="se">\&quot;}}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_send_request</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">response</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">transformation_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">),</span>
                                                <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                                <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">transformation_df</span> <span class="o">=</span> <span class="n">transformation_df</span><span class="p">[[</span><span class="s1">&#39;substratecid&#39;</span><span class="p">,</span>
                                                       <span class="s1">&#39;metabolitecid&#39;</span><span class="p">,</span>
                                                       <span class="s1">&#39;metconversion&#39;</span><span class="p">,</span>
                                                       <span class="s1">&#39;geneids&#39;</span><span class="p">,</span>
                                                       <span class="s1">&#39;pmids&#39;</span><span class="p">,</span>
                                                       <span class="s1">&#39;dois&#39;</span><span class="p">]]</span>
                <span class="n">transformation_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transformation_df</span><span class="p">)</span>
            <span class="k">except</span> <span class="n">pd</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">ParserError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error parsing CSV for gene ID </span><span class="si">{</span><span class="n">gid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>  <span class="c1"># Skip this gene ID and continue with others</span>

    <span class="n">transformation_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">transformation_dfs</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">if</span> <span class="n">transformation_dfs</span> <span class="k">else</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_write_to_csv</span><span class="p">(</span><span class="n">transformation_df</span><span class="p">,</span>
                       <span class="s1">&#39;Data/Relationships/Compound_Transformation.csv&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">transformation_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.fetch_data_for_aid" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">fetch_data_for_aid</span><span class="p">(</span><span class="n">aid</span><span class="p">,</span> <span class="n">columns_to_remove</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Fetches and processes assay data for a specified Assay ID (AID) from the
PubChem database, preparing it for analysis or further processing.</p>
<p>This method queries the PubChem database for assay data associated with
a given AID. It constructs the query URL, sends the request using a
previously established session, and processes the response. The response
is expected to be in CSV format, which this method reads into a pandas
DataFrame. Specific columns can be removed from this DataFrame based on
the requirements for analysis. This allows for the customization of
the fetched data, making it easier to work with specific datasets.</p>
<p>If the request is successful and the data is fetched without issues,
it undergoes initial processing to remove unwanted columns as specified
by the 'columns_to_remove' parameter. In case of an error during the
data fetching or processing (e.g., issues with parsing the CSV data),
appropriate error messages are logged, and an empty DataFrame is
returned as a fallback.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>aid</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The assay ID for which data is to be fetched. This ID is</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>columns_to_remove</code></td>
            <td>
                  <code>list of str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of column names that should</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pandas.DataFrame: A DataFrame containing the processed data</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>associated with the given AID. The DataFrame will exclude columns</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>listed in 'columns_to_remove'. If the data fetching fails or if</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>an error occurs during processing, an empty DataFrame is returned.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="requests.RequestException">RequestException</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If an error occurs during the HTTP request</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.errors.ParserError">ParserError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If an error occurs while parsing the CSV</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>extractor = RelationshipPropertiesExtractor()
processed_data_df = extractor.fetch_data_for_aid(12345, ['UnwantedColumn1', 'UnwantedColumn2'])
print(processed_data_df.head())
This example demonstrates how to fetch and process assay data for
the assay with ID 12345, removing 'UnwantedColumn1' and
'UnwantedColumn2' from the resulting DataFrame. The first few rows
of the processed DataFrame are printed as an output.</p>
</blockquote>
</blockquote>
</blockquote>
</details>

<details class="note" open>
  <summary>Note</summary>
  <ul>
<li>This method is part of a class that requires a valid session with
the PubChem API. Ensure that the class is properly initialized and that
the session is active.</li>
<li>The removal of columns is an optional step and can be customized
based on the analysis needs. If no columns need to be removed, pass an
empty list as 'columns_to_remove'.</li>
</ul>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fetch_data_for_aid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">aid</span><span class="p">,</span> <span class="n">columns_to_remove</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches and processes assay data for a specified Assay ID (AID) from the</span>
<span class="sd">    PubChem database, preparing it for analysis or further processing.</span>

<span class="sd">    This method queries the PubChem database for assay data associated with</span>
<span class="sd">    a given AID. It constructs the query URL, sends the request using a</span>
<span class="sd">    previously established session, and processes the response. The response</span>
<span class="sd">    is expected to be in CSV format, which this method reads into a pandas</span>
<span class="sd">    DataFrame. Specific columns can be removed from this DataFrame based on</span>
<span class="sd">    the requirements for analysis. This allows for the customization of</span>
<span class="sd">    the fetched data, making it easier to work with specific datasets.</span>

<span class="sd">    If the request is successful and the data is fetched without issues,</span>
<span class="sd">    it undergoes initial processing to remove unwanted columns as specified</span>
<span class="sd">    by the &#39;columns_to_remove&#39; parameter. In case of an error during the</span>
<span class="sd">    data fetching or processing (e.g., issues with parsing the CSV data),</span>
<span class="sd">    appropriate error messages are logged, and an empty DataFrame is</span>
<span class="sd">    returned as a fallback.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        aid (int): The assay ID for which data is to be fetched. This ID is</span>
<span class="sd">        used to construct the query URL to the PubChem database.</span>
<span class="sd">        columns_to_remove (list of str): A list of column names that should</span>
<span class="sd">        be removed from the fetched DataFrame. This allows for the exclusion</span>
<span class="sd">        of data that might not be relevant to the subsequent analysis or</span>
<span class="sd">        processing steps.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: A DataFrame containing the processed data</span>
<span class="sd">        associated with the given AID. The DataFrame will exclude columns</span>
<span class="sd">        listed in &#39;columns_to_remove&#39;. If the data fetching fails or if</span>
<span class="sd">        an error occurs during processing, an empty DataFrame is returned.</span>

<span class="sd">    Raises:</span>
<span class="sd">        requests.RequestException: If an error occurs during the HTTP request</span>
<span class="sd">        to the PubChem API. This includes scenarios such as timeout issues,</span>
<span class="sd">        non-200 status codes, or network-related errors. The exception is</span>
<span class="sd">        handled internally with logging, but it&#39;s important to be aware of</span>
<span class="sd">        its possibility.</span>
<span class="sd">        pd.errors.ParserError: If an error occurs while parsing the CSV</span>
<span class="sd">        response from PubChem into a DataFrame. This could happen due to</span>
<span class="sd">        malformed data or unexpected changes in the response format.</span>
<span class="sd">        Like with RequestException, this error is logged and results in</span>
<span class="sd">        the return of an empty DataFrame.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; extractor = RelationshipPropertiesExtractor()</span>
<span class="sd">        &gt;&gt;&gt; processed_data_df = extractor.fetch_data_for_aid(12345, [&#39;UnwantedColumn1&#39;, &#39;UnwantedColumn2&#39;])</span>
<span class="sd">        &gt;&gt;&gt; print(processed_data_df.head())</span>
<span class="sd">        This example demonstrates how to fetch and process assay data for</span>
<span class="sd">        the assay with ID 12345, removing &#39;UnwantedColumn1&#39; and</span>
<span class="sd">        &#39;UnwantedColumn2&#39; from the resulting DataFrame. The first few rows</span>
<span class="sd">        of the processed DataFrame are printed as an output.</span>

<span class="sd">    Note:</span>
<span class="sd">        - This method is part of a class that requires a valid session with</span>
<span class="sd">        the PubChem API. Ensure that the class is properly initialized and that</span>
<span class="sd">        the session is active.</span>
<span class="sd">        - The removal of columns is an optional step and can be customized</span>
<span class="sd">        based on the analysis needs. If no columns need to be removed, pass an</span>
<span class="sd">        empty list as &#39;columns_to_remove&#39;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">url</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/assay/pcget.cgi?&quot;</span>
        <span class="s2">&quot;query=download&amp;record_type=datatable&amp;actvty=&quot;</span>
        <span class="sa">f</span><span class="s2">&quot;all&amp;response_type=display&amp;aid=</span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="n">response</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_send_request</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">response</span> <span class="ow">and</span> <span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">compound_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
            <span class="c1"># Drop specified columns and process column names in-place for memory efficiency</span>
            <span class="n">columns_to_remove_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">columns_to_remove</span><span class="p">)</span>
            <span class="n">existing_columns_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">compound_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
            <span class="n">columns_to_actually_remove</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">columns_to_remove_set</span> <span class="o">&amp;</span> <span class="n">existing_columns_set</span><span class="p">)</span>
            <span class="n">compound_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">columns_to_actually_remove</span><span class="p">,</span>
                             <span class="n">errors</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">compound_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;PUBCHEM_&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;PUBCHEM_&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="c1"># compound_df.drop(columns=[col for col in columns_to_remove if col in compound_df.columns], errors=&#39;ignore&#39;, inplace=True)</span>
            <span class="c1"># compound_df.columns = [col.replace(&#39;PUBCHEM_&#39;, &#39;&#39;) if col.startswith(&#39;PUBCHEM_&#39;) else col for col in compound_df.columns]</span>
            <span class="n">compound_df</span><span class="p">[</span><span class="s1">&#39;AID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aid</span>
            <span class="k">return</span> <span class="n">compound_df</span>
        <span class="k">except</span> <span class="n">pd</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">ParserError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CSV parsing failed for AID </span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to fetch data for AID </span><span class="si">{</span><span class="n">aid</span><span class="si">}</span><span class="s2">. Status code: </span><span class="si">{</span><span class="n">response</span><span class="o">.</span><span class="n">status_code</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">response</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;No Response&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># Return an empty DataFrame in case of failure</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.fetch_similar_cids" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">fetch_similar_cids</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Fetches similar compound IDs (CIDs) from the PubChem database for a
given compound ID (CID) using 2D similarity.</p>
<p>This method queries the PubChem database to find compounds that are
similar to the given CID based on 2D structural similarity.
The similarity threshold is set to 95%, and a maximum of 100 similar
CIDs are fetched. The response is parsed from XML format to extract
the similar CIDs.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>cid</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The compound ID for which similar CIDs are to be fetched.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>tuple</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A tuple containing the original CID and a list of similar</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>CIDs. If an error occurs, the list of similar CIDs will be empty.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>Exception</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Logs an error message with the original CID and the</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Note</summary>
  <ul>
<li>The method utilizes the <code>requests</code> library for HTTP requests and
<code>xml.etree.ElementTree</code> for XML parsing.</li>
<li>In case of a request failure or parsing error, the method logs
the error and returns the original CID with an empty list,
allowing the calling function to handle the exception as needed.</li>
</ul>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fetch_similar_cids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cid</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches similar compound IDs (CIDs) from the PubChem database for a</span>
<span class="sd">    given compound ID (CID) using 2D similarity.</span>

<span class="sd">    This method queries the PubChem database to find compounds that are</span>
<span class="sd">    similar to the given CID based on 2D structural similarity.</span>
<span class="sd">    The similarity threshold is set to 95%, and a maximum of 100 similar</span>
<span class="sd">    CIDs are fetched. The response is parsed from XML format to extract</span>
<span class="sd">    the similar CIDs.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        cid (int): The compound ID for which similar CIDs are to be fetched.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing the original CID and a list of similar</span>
<span class="sd">        CIDs. If an error occurs, the list of similar CIDs will be empty.</span>

<span class="sd">    Raises:</span>
<span class="sd">        Exception: Logs an error message with the original CID and the</span>
<span class="sd">        exception if the request to PubChem fails or if parsing the XML</span>
<span class="sd">        response encounters an error.</span>

<span class="sd">    Note:</span>
<span class="sd">        - The method utilizes the `requests` library for HTTP requests and</span>
<span class="sd">        `xml.etree.ElementTree` for XML parsing.</span>
<span class="sd">        - In case of a request failure or parsing error, the method logs</span>
<span class="sd">        the error and returns the original CID with an empty list,</span>
<span class="sd">        allowing the calling function to handle the exception as needed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">url</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/&quot;</span>
           <span class="sa">f</span><span class="s2">&quot;fastsimilarity_2d/cid/</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span><span class="si">}</span><span class="s2">/cids/XML?Threshold=95&amp;MaxRecords=100&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
        <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
        <span class="n">xml_data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">text</span>

        <span class="c1"># Parse XML data</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">xml_data</span><span class="p">))</span>
        <span class="n">root</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">getroot</span><span class="p">()</span>

        <span class="c1"># Extracting CID values</span>
        <span class="n">similar_cids</span> <span class="o">=</span> <span class="p">[</span><span class="n">element</span><span class="o">.</span><span class="n">text</span> <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;{http://pubchem.ncbi.nlm.nih.gov/pug_rest}CID&#39;</span><span class="p">)]</span>
        <span class="k">return</span> <span class="n">cid</span><span class="p">,</span> <span class="n">similar_cids</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing CID </span><span class="si">{</span><span class="n">cid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cid</span><span class="p">,</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.gene_enzyme_relationship" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">gene_enzyme_relationship</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Extracts and saves relationships between genes and enzymes based on
the provided dataset.</p>
<p>This method selects relevant columns to highlight the relationships
between genes and their corresponding enzymes.
It removes duplicate entries to ensure that each relationship is
represented uniquely and saves the resultant data to
a CSV file. This facilitates easy integration of genetic data into
knowledge bases or further analysis.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>main_data</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the CSV file containing gene and enzyme data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pandas.DataFrame: A DataFrame of unique gene-enzyme relationships,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>including gene ID and enzyme accession numbers.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="side-effects" open>
  <summary>Side Effects</summary>
  <ul>
<li>Writes the processed data to 'Data/Gene_Enzyme_Relationship.csv'
in a structured CSV format.</li>
</ul>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">gene_enzyme_relationship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">main_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts and saves relationships between genes and enzymes based on</span>
<span class="sd">    the provided dataset.</span>

<span class="sd">    This method selects relevant columns to highlight the relationships</span>
<span class="sd">    between genes and their corresponding enzymes.</span>
<span class="sd">    It removes duplicate entries to ensure that each relationship is</span>
<span class="sd">    represented uniquely and saves the resultant data to</span>
<span class="sd">    a CSV file. This facilitates easy integration of genetic data into</span>
<span class="sd">    knowledge bases or further analysis.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        main_data (str): Path to the CSV file containing gene and enzyme data.</span>
<span class="sd">        Expected columns include &#39;Target GeneID&#39; and &#39;Target Accession&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: A DataFrame of unique gene-enzyme relationships,</span>
<span class="sd">        including gene ID and enzyme accession numbers.</span>

<span class="sd">    Side Effects:</span>
<span class="sd">        - Writes the processed data to &#39;Data/Gene_Enzyme_Relationship.csv&#39;</span>
<span class="sd">        in a structured CSV format.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">main_data</span><span class="p">)</span>
    <span class="n">columns_to_select</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Target GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Accession&#39;</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">columns_to_select</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data/Gene_Enzyme_Relationship.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_properties_extractor.RelationshipPropertiesExtractor.process_chunk" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_chunk</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Processes a chunk of CIDs in parallel to fetch similar CIDs for each CID
in the chunk.</p>
<p>This method uses a ThreadPoolExecutor to send out concurrent requests for
fetching similar CIDs for a list of CIDs.
The number of worker threads is set to 5. Each CID's request is handled
by <code>fetch_similar_cids</code> method.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>chunk</code></td>
            <td>
                  <code>list of int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of compound IDs (CIDs) to process in</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of tuples: A list of tuples, each containing a CID and its</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>corresponding list of similar CIDs.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="side-effects" open>
  <summary>Side Effects</summary>
  <ul>
<li>Utilizes concurrent threads to speed up the fetching process.</li>
<li>May log errors if any occur during the fetching of similar CIDs
for individual CIDs.</li>
</ul>
</details>
            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_properties_extractor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">process_chunk</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chunk</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes a chunk of CIDs in parallel to fetch similar CIDs for each CID</span>
<span class="sd">    in the chunk.</span>

<span class="sd">    This method uses a ThreadPoolExecutor to send out concurrent requests for</span>
<span class="sd">    fetching similar CIDs for a list of CIDs.</span>
<span class="sd">    The number of worker threads is set to 5. Each CID&#39;s request is handled</span>
<span class="sd">    by `fetch_similar_cids` method.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        chunk (list of int): A list of compound IDs (CIDs) to process in</span>
<span class="sd">        parallel.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list of tuples: A list of tuples, each containing a CID and its</span>
<span class="sd">        corresponding list of similar CIDs.</span>

<span class="sd">    Side Effects:</span>
<span class="sd">        - Utilizes concurrent threads to speed up the fetching process.</span>
<span class="sd">        - May log errors if any occur during the fetching of similar CIDs</span>
<span class="sd">        for individual CIDs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
        <span class="n">futures</span> <span class="o">=</span> <span class="p">[</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fetch_similar_cids</span><span class="p">,</span> <span class="n">cid</span><span class="p">)</span> <span class="k">for</span> <span class="n">cid</span> <span class="ow">in</span> <span class="n">chunk</span><span class="p">]</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span> <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">as_completed</span><span class="p">(</span><span class="n">futures</span><span class="p">)]</span>
    <span class="k">return</span> <span class="n">results</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h2 id="7-relationship-data-processor">7. Relationship Data Processor</h2>


<div class="doc doc-object doc-module">



<a id="chemgraphbuilder.relationship_data_processor"></a>
    <div class="doc doc-contents first">

      <p>Module for processing relationship data files using Dask and pandas.</p>
<p>This module includes the <code>RelationshipDataProcessor</code> class, which is used to
load, filter, clean, and process data files related to assay-compound relationships.</p>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor" class="doc doc-heading">
            <code>RelationshipDataProcessor</code>


</h2>


    <div class="doc doc-contents ">


      <p>A class to process relationship data files, filtering and augmenting the data.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.path">path</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory path where the data files are stored.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.csv_files">csv_files</span></code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of CSV files matching the pattern 'AID_*.csv'.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.all_data_connected">all_data_connected</span></code></td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary containing additional data connected to assays.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/relationship_data_processor.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">RelationshipDataProcessor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class to process relationship data files, filtering and augmenting the data.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        path (str): The directory path where the data files are stored.</span>
<span class="sd">        csv_files (list): List of CSV files matching the pattern &#39;AID_*.csv&#39;.</span>
<span class="sd">        all_data_connected (dict): A dictionary containing additional data connected to assays.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes the RelationshipDataProcessor with the specified path.</span>

<span class="sd">        Args:</span>
<span class="sd">            path (str): The directory path containing the CSV files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">path</span> <span class="o">=</span> <span class="n">path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;AID_*.csv&quot;</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">all_data_connected</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_all_data_connected</span><span class="p">(</span><span class="s1">&#39;Data/AllDataConnected.csv&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_load_all_data_connected</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads additional data from a specified file and organizes it into a dictionary.</span>

<span class="sd">        Args:</span>
<span class="sd">            file_path (str): The path to the file containing additional data.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary with keys as tuples of (aid, cid, activity_outcome)</span>
<span class="sd">                  and values as dictionaries of additional information.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_data_connected</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">ddf</span> <span class="o">=</span> <span class="n">dd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">blocksize</span><span class="o">=</span><span class="mf">20e6</span><span class="p">)</span>
        <span class="n">ddf</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ddf</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="n">ddf</span> <span class="o">=</span> <span class="n">ddf</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;aid&#39;</span><span class="p">,</span> <span class="s1">&#39;cid&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;any&#39;</span><span class="p">)</span>
        <span class="n">partitions</span> <span class="o">=</span> <span class="n">ddf</span><span class="o">.</span><span class="n">to_delayed</span><span class="p">()</span>
        <span class="n">ddf</span> <span class="o">=</span> <span class="n">ddf</span><span class="o">.</span><span class="n">repartition</span><span class="p">(</span><span class="n">partition_size</span><span class="o">=</span><span class="mf">10e5</span><span class="p">)</span>

        <span class="nd">@dask</span><span class="o">.</span><span class="n">delayed</span>
        <span class="k">def</span> <span class="nf">process_partition</span><span class="p">(</span><span class="n">partition</span><span class="p">):</span>
            <span class="n">result</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">partition</span> <span class="o">=</span> <span class="n">partition</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;aid&#39;</span><span class="p">,</span> <span class="s1">&#39;cid&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;any&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">partition</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                <span class="n">key</span> <span class="o">=</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;aid&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]),</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">])</span>
                <span class="n">result</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
            <span class="k">return</span> <span class="n">result</span>

        <span class="n">results</span> <span class="o">=</span> <span class="n">dask</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">process_partition</span><span class="p">(</span><span class="n">part</span><span class="p">)</span> <span class="k">for</span> <span class="n">part</span> <span class="ow">in</span> <span class="n">partitions</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">all_data_connected</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

        <span class="c1"># Optionally save the dictionary to a file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Data/Relationships/all_data_connected_dict.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">all_data_connected</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">all_data_connected</span>

    <span class="k">def</span> <span class="nf">_add_all_data_connected_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">row</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adds additional information from all_data_connected to a row.</span>

<span class="sd">        Args:</span>
<span class="sd">            row (pd.Series): A row from a DataFrame.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.Series: The updated row with additional data if available.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">key</span> <span class="o">=</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;aid&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]),</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">all_data_connected</span><span class="p">:</span>
            <span class="n">additional_info</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">all_data_connected</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">col</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">additional_info</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">row</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">val</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Key </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> not found in all_data_connected.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">row</span>

    <span class="k">def</span> <span class="nf">_get_filtered_columns</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts unique column names from the CSV files and additional data.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: A list of unique column names.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_columns</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="c1"># Extract additional columns from the all_data_connected dictionary</span>
        <span class="n">additional_columns</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">all_data_connected</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">additional_columns</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">value</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="k">def</span> <span class="nf">read_columns</span><span class="p">(</span><span class="n">file</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># Read only column names from the CSV file</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">return</span> <span class="nb">set</span><span class="p">([</span><span class="n">col</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error reading </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="nb">set</span><span class="p">()</span>

        <span class="c1"># Use ThreadPoolExecutor for concurrent reading of columns from multiple files</span>
        <span class="k">with</span> <span class="n">concurrent</span><span class="o">.</span><span class="n">futures</span><span class="o">.</span><span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
            <span class="n">results</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">executor</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">read_columns</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">csv_files</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">columns</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">all_columns</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">columns</span><span class="p">)</span>

        <span class="n">all_columns</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">additional_columns</span><span class="p">)</span>

        <span class="c1"># Save the combined columns to a file for reference</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Data/Relationships/all_columns.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">all_columns</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">item</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_columns</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">process_files</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes the CSV files by filtering, cleaning, and augmenting data.</span>

<span class="sd">        The processed data is saved to output files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_filter_and_clean_data</span><span class="p">()</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Data filtered, cleaned, and combined successfully.&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_filter_and_clean_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Filters and cleans data from CSV files, then saves to output files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">output_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s1">&#39;Data/Relationships/Assay_Compound_Relationship.csv&#39;</span><span class="p">)</span>
        <span class="n">compound_gene_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s1">&#39;Data/Relationships/Compound_Gene_Relationship.csv&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output_file</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">compound_gene_file</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">compound_gene_file</span><span class="p">)</span>

        <span class="n">unique_column_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_filtered_columns</span><span class="p">()</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;activity&#39;</span><span class="p">]</span>

        <span class="c1"># Initialize output files with headers</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">unique_column_names</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span> <span class="s1">&#39;target_geneid&#39;</span><span class="p">,</span> <span class="s1">&#39;activity&#39;</span><span class="p">,</span> <span class="s1">&#39;aid&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">compound_gene_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">csv_files</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">100</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing file </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">csv_files</span><span class="p">)</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dask</span><span class="o">.</span><span class="n">delayed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_process_file</span><span class="p">)(</span><span class="n">file</span><span class="p">,</span> <span class="n">unique_column_names</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="n">compound_gene_file</span><span class="p">))</span>

        <span class="n">dask</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="o">*</span><span class="n">tasks</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">csv_files</span><span class="p">)</span><span class="si">}</span><span class="s2"> files&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_process_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file</span><span class="p">,</span> <span class="n">unique_column_names</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="n">compound_gene_file</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes a single CSV file, applying filtering, cleaning, and adding data.</span>

<span class="sd">        Args:</span>
<span class="sd">            file (str): The file path to the CSV file.</span>
<span class="sd">            unique_column_names (list): The list of unique column names to use.</span>
<span class="sd">            output_file (str): The path to the output file for combined data.</span>
<span class="sd">            compound_gene_file (str): The path to the output file for compound-gene relationships.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ddf</span> <span class="o">=</span> <span class="n">dd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">blocksize</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;ASSAYDATA_COMMENT&#39;</span><span class="p">:</span> <span class="s1">&#39;object&#39;</span><span class="p">})</span>
        <span class="n">ddf</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ddf</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="n">ddf</span> <span class="o">=</span> <span class="n">ddf</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;any&#39;</span><span class="p">)</span>

        <span class="c1"># Repartition to balance memory usage and performance</span>
        <span class="n">ddf</span> <span class="o">=</span> <span class="n">ddf</span><span class="o">.</span><span class="n">repartition</span><span class="p">(</span><span class="n">partition_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>

        <span class="n">phenotype_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ddf</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;phenotype&#39;</span><span class="p">)]</span>

        <span class="k">def</span> <span class="nf">process_partition</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">to_frame</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>  <span class="c1"># Convert to DataFrame if a Series is encountered</span>
                <span class="k">if</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()]</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Duplicated columns removed from partition.&quot;</span><span class="p">)</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">unique_column_names</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">NA</span><span class="p">)</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;aid&#39;</span><span class="p">,</span> <span class="s1">&#39;cid&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;any&#39;</span><span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;measured_activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">phenotype_cols</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">mode</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">row</span><span class="o">.</span><span class="n">mode</span><span class="p">()</span><span class="o">.</span><span class="n">empty</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

                    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_add_all_data_connected_info</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

                    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">phenotype_cols</span><span class="p">)</span> <span class="ow">and</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">,</span> <span class="s1">&#39;assay_name&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">propagate_phenotype</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

                    <span class="k">if</span> <span class="s1">&#39;target_geneid&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;target_geneid&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">NA</span>

                    <span class="k">if</span> <span class="s1">&#39;sid&#39;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;activity_url&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/bioassay/</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;aid&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">#sid=</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;sid&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;activity_url&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">NA</span>

                    <span class="c1"># Drop rows where both aid and cid are 1</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;aid&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">)]</span>

                    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_determine_labels_and_activity</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

                    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed partition with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="si">}</span><span class="s2"> rows.&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                        <span class="c1"># Write the processed data to the output files</span>
                        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                        <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span> <span class="s1">&#39;target_geneid&#39;</span><span class="p">,</span> <span class="s1">&#39;activity&#39;</span><span class="p">,</span> <span class="s1">&#39;aid&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">compound_gene_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">compound_gene_file</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No data to process after filtering.&quot;</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing partition: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">ddf</span><span class="o">.</span><span class="n">map_partitions</span><span class="p">(</span><span class="n">process_partition</span><span class="p">)</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">most_frequent</span><span class="p">(</span><span class="n">row</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Finds the most frequent value in a row, excluding NaN values.</span>

<span class="sd">        Args:</span>
<span class="sd">            row (pd.Series): A row from a DataFrame.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: The most frequent value in the row.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
        <span class="n">string_values</span> <span class="o">=</span> <span class="n">values</span><span class="p">[</span><span class="n">values</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">))]</span>
        <span class="k">return</span> <span class="n">string_values</span><span class="o">.</span><span class="n">mode</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">string_values</span><span class="o">.</span><span class="n">empty</span> <span class="k">else</span> <span class="kc">None</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">propagate_phenotype</span><span class="p">(</span><span class="n">group</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Propagates the phenotype information within a group.</span>

<span class="sd">        Args:</span>
<span class="sd">            group (pd.DataFrame): A DataFrame group.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: The updated group with propagated phenotype information.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">phenotype_value</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_value</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">group</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">phenotype_value</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">group</span>

    <span class="k">def</span> <span class="nf">_determine_labels_and_activity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_df</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determines the activity labels for the data based on predefined keywords.</span>

<span class="sd">        Args:</span>
<span class="sd">            merged_df (pd.DataFrame): The DataFrame containing merged data.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: The DataFrame with determined activity labels.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">inhibitor_keywords</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;reversible inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;time dependent inhibition&#39;</span><span class="p">,</span>
            <span class="s1">&#39;inhibitory activity&#39;</span><span class="p">,</span> <span class="s1">&#39;time-dependent inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;time dependent irreversible inhibition&#39;</span><span class="p">,</span>
            <span class="s1">&#39;inhibitory concentration&#39;</span><span class="p">,</span> <span class="s1">&#39;inhibitory effect&#39;</span><span class="p">,</span> <span class="s1">&#39;inhibitory potency&#39;</span><span class="p">,</span>
            <span class="s1">&#39;concentration required to inhibit&#39;</span><span class="p">,</span> <span class="s1">&#39;competitive inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;cyp inhibition&#39;</span><span class="p">,</span>
            <span class="s1">&#39;irreversible inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;mechanism based inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;mixed inhibition&#39;</span><span class="p">,</span>
            <span class="s1">&#39;mixed type inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;inhibitory constant&#39;</span><span class="p">,</span> <span class="s1">&#39;antagonistic activity&#39;</span><span class="p">,</span> <span class="s1">&#39;selectivity&#39;</span><span class="p">,</span>
            <span class="s1">&#39;s1p4 agonists&#39;</span><span class="p">,</span> <span class="s1">&#39;small molecule antagonists&#39;</span><span class="p">,</span> <span class="s1">&#39;displacement&#39;</span><span class="p">,</span> <span class="s1">&#39;mediated midazolam 1-hydroxylation&#39;</span><span class="p">,</span>
            <span class="s1">&#39;time/nadph-dependent inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;reversal inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;mechanism-based inhibition&#39;</span><span class="p">,</span>
            <span class="s1">&#39;mechanism based time dependent inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;reversible competitive inhibition&#39;</span><span class="p">,</span>
            <span class="s1">&#39;predictive competitive inhibition&#39;</span><span class="p">,</span><span class="s1">&#39;noncompetitive inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;in vitro inhibitory&#39;</span><span class="p">,</span>
            <span class="s1">&#39;in vitro inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;inhibition of&#39;</span><span class="p">,</span> <span class="s1">&#39;direct inhibition&#39;</span><span class="p">,</span><span class="s1">&#39;enzyme inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;dndi&#39;</span><span class="p">,</span>
            <span class="s1">&#39;inhibition assay&#39;</span>
        <span class="p">]</span>

        <span class="n">ligand_keywords</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;binding affinity&#39;</span><span class="p">,</span> <span class="s1">&#39;spectral binding&#39;</span><span class="p">,</span> <span class="s1">&#39;interaction with&#39;</span><span class="p">,</span> <span class="s1">&#39;bind&#39;</span><span class="p">,</span>
            <span class="s1">&#39;covalent binding affinity&#39;</span><span class="p">,</span> <span class="s1">&#39;apparent binding affinity&#39;</span>
        <span class="p">]</span>

        <span class="n">inhibitor_substrate_keywords</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;inhibitors and substrates&#39;</span>
        <span class="p">]</span>

        <span class="n">inhibitor_activator_modulator_keywords</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;apoprotein formation&#39;</span><span class="p">,</span> <span class="s1">&#39;panel assay&#39;</span><span class="p">,</span> <span class="s1">&#39;eurofins-panlabs enzyme assay&#39;</span>
        <span class="p">]</span>

        <span class="n">substrate_keywords</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;drug metabolism&#39;</span><span class="p">,</span> <span class="s1">&#39;prodrug&#39;</span><span class="p">,</span> <span class="s1">&#39;metabolic&#39;</span><span class="p">,</span> <span class="s1">&#39;oxidation&#39;</span><span class="p">,</span> <span class="s1">&#39;substrate activity&#39;</span><span class="p">,</span>
            <span class="s1">&#39;michaelis-menten&#39;</span><span class="p">,</span> <span class="s1">&#39;metabolic stability&#39;</span><span class="p">,</span> <span class="s1">&#39;bioactivation&#39;</span><span class="p">,</span> <span class="s1">&#39;drug level&#39;</span><span class="p">,</span>
            <span class="s1">&#39;enzyme-mediated drug depletion&#39;</span><span class="p">,</span> <span class="s1">&#39;enzyme-mediated compound formation&#39;</span><span class="p">,</span>
            <span class="s1">&#39;phenotyping&#39;</span><span class="p">,</span> <span class="s1">&#39;activity of human recombinant cyp&#39;</span><span class="p">,</span> <span class="s1">&#39;activity of recombinant cyp&#39;</span><span class="p">,</span>
            <span class="s1">&#39;activity at cyp&#39;</span><span class="p">,</span> <span class="s1">&#39;enzyme-mediated drug metabolism&#39;</span>
        <span class="p">]</span>

        <span class="n">inactivator_keywords</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;inactivator&#39;</span><span class="p">,</span> <span class="s1">&#39;inactivation of&#39;</span><span class="p">,</span> <span class="s1">&#39;mechanism based inactivation of&#39;</span><span class="p">,</span> <span class="s1">&#39;inactivators&#39;</span><span class="p">,</span>
            <span class="s1">&#39;metabolism dependent inactivation&#39;</span>
        <span class="p">]</span>

        <span class="n">activator_keywords</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;assay for activators&#39;</span><span class="p">,</span> <span class="s1">&#39;activation of&#39;</span><span class="p">,</span> <span class="s1">&#39;activators of&#39;</span>
        <span class="p">]</span>

        <span class="n">inducer_keywords</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;induction of&#39;</span><span class="p">,</span> <span class="s1">&#39;inducer&#39;</span><span class="p">,</span> <span class="s1">&#39;inducers&#39;</span><span class="p">,</span> <span class="s1">&#39;time-dependant induction&#39;</span>
        <span class="p">]</span>

        <span class="n">all_keywords</span> <span class="o">=</span> <span class="p">(</span><span class="n">inhibitor_keywords</span> <span class="o">+</span> <span class="n">ligand_keywords</span> <span class="o">+</span> <span class="n">inhibitor_substrate_keywords</span> <span class="o">+</span>
                        <span class="n">inhibitor_activator_modulator_keywords</span> <span class="o">+</span> <span class="n">substrate_keywords</span> <span class="o">+</span>
                        <span class="n">inactivator_keywords</span> <span class="o">+</span> <span class="n">activator_keywords</span> <span class="o">+</span> <span class="n">inducer_keywords</span><span class="p">)</span>

        <span class="n">keyword_to_label</span> <span class="o">=</span> <span class="p">{</span>
            <span class="o">**</span><span class="p">{</span><span class="n">keyword</span><span class="p">:</span> <span class="s1">&#39;Inhibitor&#39;</span> <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">inhibitor_keywords</span><span class="p">},</span>
            <span class="o">**</span><span class="p">{</span><span class="n">keyword</span><span class="p">:</span> <span class="s1">&#39;Inhibitor/Substrate&#39;</span> <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">inhibitor_substrate_keywords</span><span class="p">},</span>
            <span class="o">**</span><span class="p">{</span><span class="n">keyword</span><span class="p">:</span> <span class="s1">&#39;Inhibitor/Inducer/Modulator&#39;</span> <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">inhibitor_activator_modulator_keywords</span><span class="p">},</span>
            <span class="o">**</span><span class="p">{</span><span class="n">keyword</span><span class="p">:</span> <span class="s1">&#39;Substrate&#39;</span> <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">substrate_keywords</span><span class="p">},</span>
            <span class="o">**</span><span class="p">{</span><span class="n">keyword</span><span class="p">:</span> <span class="s1">&#39;Inactivator&#39;</span> <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">inactivator_keywords</span><span class="p">},</span>
            <span class="o">**</span><span class="p">{</span><span class="n">keyword</span><span class="p">:</span> <span class="s1">&#39;Activator&#39;</span> <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">activator_keywords</span><span class="p">},</span>
            <span class="o">**</span><span class="p">{</span><span class="n">keyword</span><span class="p">:</span> <span class="s1">&#39;Inducer&#39;</span> <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">inducer_keywords</span><span class="p">},</span>
            <span class="o">**</span><span class="p">{</span><span class="n">keyword</span><span class="p">:</span> <span class="s1">&#39;Ligand&#39;</span> <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">ligand_keywords</span><span class="p">},</span>
        <span class="p">}</span>

        <span class="k">def</span> <span class="nf">determine_active_label</span><span class="p">(</span><span class="n">assay_name</span><span class="p">):</span>
            <span class="c1"># Determine the appropriate label based on the first keyword found in the assay name</span>
            <span class="n">assay_name_lower</span> <span class="o">=</span> <span class="n">assay_name</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
            <span class="n">first_keyword</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">first_position</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">assay_name_lower</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">all_keywords</span><span class="p">:</span>
                <span class="n">position</span> <span class="o">=</span> <span class="n">assay_name_lower</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">keyword</span><span class="p">)</span>
                <span class="k">if</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="n">first_position</span><span class="p">:</span>
                    <span class="n">first_keyword</span> <span class="o">=</span> <span class="n">keyword</span>
                    <span class="n">first_position</span> <span class="o">=</span> <span class="n">position</span>

            <span class="k">if</span> <span class="n">first_keyword</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">keyword_to_label</span><span class="p">[</span><span class="n">first_keyword</span><span class="p">]</span>
            <span class="k">return</span> <span class="s1">&#39;Inhibitor/Inducer/Modulator&#39;</span>

        <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Assign the &#39;Inactive&#39; label where the activity outcome is inactive</span>
        <span class="n">inactive_mask</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Inactive&#39;</span>
        <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">inactive_mask</span><span class="p">,</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Inactive&#39;</span>

        <span class="c1"># Assign labels based on assay name keywords for active outcomes</span>
        <span class="n">active_mask</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Active&#39;</span>
        <span class="k">if</span> <span class="n">active_mask</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
            <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span><span class="p">,</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span><span class="p">,</span> <span class="s1">&#39;assay_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">determine_active_label</span><span class="p">)</span>
            <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span> <span class="o">&amp;</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;activity_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s1">&#39;Km&#39;</span><span class="p">,</span> <span class="s1">&#39;Drug metabolism&#39;</span><span class="p">]),</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Substrate&#39;</span>

            <span class="n">substrate_pattern</span> <span class="o">=</span> <span class="sa">r</span><span class="s1">&#39;(activity of.*oxidation)|(activity at cyp.*phenotyping)|(activity at human recombinant cyp.*formation)|(activity at recombinant cyp.*formation)&#39;</span>
            <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span> <span class="o">&amp;</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;assay_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">substrate_pattern</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Substrate&#39;</span>

            <span class="n">ActIndMod_pattern</span> <span class="o">=</span> <span class="sa">r</span><span class="s1">&#39;(effect on cyp)|(effect on human recombinant cyp)|(effect on recombinant cyp)|(effect on human cyp)&#39;</span>
            <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span> <span class="o">&amp;</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;assay_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">ActIndMod_pattern</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Inhibitor/Inducer/Modulator&#39;</span>

            <span class="n">inducer_pattern</span> <span class="o">=</span> <span class="sa">r</span><span class="s1">&#39;(effect on cyp.*induction)|(induction of.*)&#39;</span>
            <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span> <span class="o">&amp;</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;assay_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">inducer_pattern</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Inducer&#39;</span>

            <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span> <span class="o">&amp;</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;activity_direction&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;decreasing&#39;</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Inhibitor&#39;</span>
            <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span> <span class="o">&amp;</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;activity_direction&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;increasing&#39;</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Activator&#39;</span>
            <span class="n">merged_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">active_mask</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;aid&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1215398</span><span class="p">),</span> <span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Inactivator&#39;</span>

        <span class="k">return</span> <span class="n">merged_df</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">path</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Initializes the RelationshipDataProcessor with the specified path.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>path</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory path containing the CSV files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the RelationshipDataProcessor with the specified path.</span>

<span class="sd">    Args:</span>
<span class="sd">        path (str): The directory path containing the CSV files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">path</span> <span class="o">=</span> <span class="n">path</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">csv_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;AID_*.csv&quot;</span><span class="p">))</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">all_data_connected</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_all_data_connected</span><span class="p">(</span><span class="s1">&#39;Data/AllDataConnected.csv&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.most_frequent" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">most_frequent</span><span class="p">(</span><span class="n">row</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

      <p>Finds the most frequent value in a row, excluding NaN values.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>row</code></td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A row from a DataFrame.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The most frequent value in the row.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">most_frequent</span><span class="p">(</span><span class="n">row</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Finds the most frequent value in a row, excluding NaN values.</span>

<span class="sd">    Args:</span>
<span class="sd">        row (pd.Series): A row from a DataFrame.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The most frequent value in the row.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">values</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
    <span class="n">string_values</span> <span class="o">=</span> <span class="n">values</span><span class="p">[</span><span class="n">values</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">))]</span>
    <span class="k">return</span> <span class="n">string_values</span><span class="o">.</span><span class="n">mode</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">string_values</span><span class="o">.</span><span class="n">empty</span> <span class="k">else</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.process_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_files</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Processes the CSV files by filtering, cleaning, and augmenting data.</p>
<p>The processed data is saved to output files.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">process_files</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the CSV files by filtering, cleaning, and augmenting data.</span>

<span class="sd">    The processed data is saved to output files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_filter_and_clean_data</span><span class="p">()</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Data filtered, cleaned, and combined successfully.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.relationship_data_processor.RelationshipDataProcessor.propagate_phenotype" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">propagate_phenotype</span><span class="p">(</span><span class="n">group</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

      <p>Propagates the phenotype information within a group.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>group</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A DataFrame group.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: The updated group with propagated phenotype information.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/relationship_data_processor.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">propagate_phenotype</span><span class="p">(</span><span class="n">group</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Propagates the phenotype information within a group.</span>

<span class="sd">    Args:</span>
<span class="sd">        group (pd.DataFrame): A DataFrame group.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: The updated group with propagated phenotype information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">phenotype_value</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_value</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">group</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">phenotype_value</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">group</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h2 id="8-add-graph-relationships">8. Add Graph Relationships</h2>


<div class="doc doc-object doc-module">



<a id="chemgraphbuilder.setup_data_folder"></a>
    <div class="doc doc-contents first">

      <p>Module to set up a data directory with a predefined structure.</p>
<p>This module provides the DataFolderSetup class, which creates a directory
structure for a data folder. The structure includes nodes and
relationships folders with specified subfolders.</p>


<p><span class="doc-section-title">Classes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.setup_data_folder.DataFolderSetup">DataFolderSetup</span></code></td>
            <td>
              <div class="doc-md-description">
                <p>Class to set up a data directory with a predefined structure.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Functions:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="chemgraphbuilder.setup_data_folder.main" href="#chemgraphbuilder.setup_data_folder.main">main</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Main function to set up the data directory.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="chemgraphbuilder.setup_data_folder.SetupDataFolder" class="doc doc-heading">
            <code>SetupDataFolder</code>


</h2>


    <div class="doc doc-contents ">


      <p>Class to set up a data directory with a predefined structure.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.setup_data_folder.SetupDataFolder.data_folder">data_folder</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the data folder.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.setup_data_folder.SetupDataFolder.base_path">base_path</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The base path for the data directory.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="chemgraphbuilder.setup_data_folder.SetupDataFolder.structure">structure</span></code></td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The structure of directories to create.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

              <details class="quote">
                <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">SetupDataFolder</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class to set up a data directory with a predefined structure.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data_folder (str): The name of the data folder.</span>
<span class="sd">        base_path (str): The base path for the data directory.</span>
<span class="sd">        structure (dict): The structure of directories to create.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes the DataFolderSetup with the data folder name and directory structure.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_folder</span> <span class="o">=</span> <span class="s2">&quot;Data&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">base_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_folder</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Nodes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Compound_Properties&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Relationships&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;Assay_Compound_Relationship&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Compound_Similarities&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Cpd_Cpd_CoOccurrence&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Cpd_Gene_CoOccurrence&quot;</span>
            <span class="p">]</span>
        <span class="p">}</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">create_folder</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a folder if it does not already exist.</span>

<span class="sd">        Args:</span>
<span class="sd">            path (str): The path of the folder to create.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Created folder: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Folder already exists: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">setup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets up the data directory structure based on the predefined structure.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Create the base data directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">)</span>

        <span class="c1"># Create the &#39;Nodes&#39; directory and its subdirectories</span>
        <span class="n">nodes_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">,</span> <span class="s2">&quot;Nodes&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">nodes_path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="s2">&quot;Nodes&quot;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nodes_path</span><span class="p">,</span> <span class="n">folder</span><span class="p">))</span>

        <span class="c1"># Create the &#39;Relationships&#39; directory and its subdirectories</span>
        <span class="n">relationships_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">,</span> <span class="s2">&quot;Relationships&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">relationships_path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="s2">&quot;Relationships&quot;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">relationships_path</span><span class="p">,</span> <span class="n">folder</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.setup_data_folder.SetupDataFolder.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Initializes the DataFolderSetup with the data folder name and directory structure.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the DataFolderSetup with the data folder name and directory structure.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data_folder</span> <span class="o">=</span> <span class="s2">&quot;Data&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">base_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_folder</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;Nodes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Compound_Properties&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Relationships&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;Assay_Compound_Relationship&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Compound_Similarities&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Cpd_Cpd_CoOccurrence&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Cpd_Gene_CoOccurrence&quot;</span>
        <span class="p">]</span>
    <span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.setup_data_folder.SetupDataFolder.create_folder" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_folder</span><span class="p">(</span><span class="n">path</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

      <p>Creates a folder if it does not already exist.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>path</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path of the folder to create.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">create_folder</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a folder if it does not already exist.</span>

<span class="sd">    Args:</span>
<span class="sd">        path (str): The path of the folder to create.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Created folder: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Folder already exists: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="chemgraphbuilder.setup_data_folder.SetupDataFolder.setup" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">setup</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Sets up the data directory structure based on the predefined structure.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">setup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sets up the data directory structure based on the predefined structure.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create the base data directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">)</span>

    <span class="c1"># Create the &#39;Nodes&#39; directory and its subdirectories</span>
    <span class="n">nodes_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">,</span> <span class="s2">&quot;Nodes&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">nodes_path</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="s2">&quot;Nodes&quot;</span><span class="p">]:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nodes_path</span><span class="p">,</span> <span class="n">folder</span><span class="p">))</span>

    <span class="c1"># Create the &#39;Relationships&#39; directory and its subdirectories</span>
    <span class="n">relationships_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_path</span><span class="p">,</span> <span class="s2">&quot;Relationships&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">relationships_path</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="s2">&quot;Relationships&quot;</span><span class="p">]:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">relationships_path</span><span class="p">,</span> <span class="n">folder</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h2 id="chemgraphbuilder.setup_data_folder.main" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Main function to set up the data directory.</p>

            <details class="quote">
              <summary>Source code in <code>chemgraphbuilder/setup_data_folder.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function to set up the data directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_folder_setup</span> <span class="o">=</span> <span class="n">SetupDataFolder</span><span class="p">()</span>
    <span class="n">data_folder_setup</span><span class="o">.</span><span class="n">setup</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
    <a href="https://github.com/your-username" target="_blank" rel="noopener" title="GitHub" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
    
    
    
    
    <a href="https://www.linkedin.com/in/your-profile/" target="_blank" rel="noopener" title="LinkedIn" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.sections", "navigation.left", "navigation.tracking", "navigation.instant", "navigation.expand", "navigation.indexes", "header.social"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
        <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js"></script>
      
        <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>
      
    
  </body>
</html>